// @matter/types/dist/esm/protocol/definitions/general.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var GeneralStatusCode = /* @__PURE__ */ ((GeneralStatusCode2) => {
  GeneralStatusCode2[GeneralStatusCode2["Success"] = 0] = "Success";
  GeneralStatusCode2[GeneralStatusCode2["Failure"] = 1] = "Failure";
  GeneralStatusCode2[GeneralStatusCode2["BadPrecondition"] = 2] = "BadPrecondition";
  GeneralStatusCode2[GeneralStatusCode2["OutOfRange"] = 3] = "OutOfRange";
  GeneralStatusCode2[GeneralStatusCode2["BadRequest"] = 4] = "BadRequest";
  GeneralStatusCode2[GeneralStatusCode2["Unsupported"] = 5] = "Unsupported";
  GeneralStatusCode2[GeneralStatusCode2["Unexpected"] = 6] = "Unexpected";
  GeneralStatusCode2[GeneralStatusCode2["ResourceExhausted"] = 7] = "ResourceExhausted";
  GeneralStatusCode2[GeneralStatusCode2["Busy"] = 8] = "Busy";
  GeneralStatusCode2[GeneralStatusCode2["Timeout"] = 9] = "Timeout";
  GeneralStatusCode2[GeneralStatusCode2["Continue"] = 10] = "Continue";
  GeneralStatusCode2[GeneralStatusCode2["Aborted"] = 11] = "Aborted";
  GeneralStatusCode2[GeneralStatusCode2["InvalidArgument"] = 12] = "InvalidArgument";
  GeneralStatusCode2[GeneralStatusCode2["NotFound"] = 13] = "NotFound";
  GeneralStatusCode2[GeneralStatusCode2["AlreadyExists"] = 14] = "AlreadyExists";
  GeneralStatusCode2[GeneralStatusCode2["PermissionDenied"] = 15] = "PermissionDenied";
  GeneralStatusCode2[GeneralStatusCode2["DataLoss"] = 16] = "DataLoss";
  GeneralStatusCode2[GeneralStatusCode2["MessageTooLarge"] = 17] = "MessageTooLarge";
  return GeneralStatusCode2;
})(GeneralStatusCode || {});
export {
  GeneralStatusCode
};
