// @matter/protocol/dist/esm/certificate/kinds/Icac.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Bytes, Diagnostic, PublicKey } from "#general";
import { FabricId } from "#types";
import { CertificateError } from "./common.js";
import { ExtensionKeyUsageSchema } from "./definitions/base.js";
import { OperationalCertificate } from "./definitions/operational.js";
import { OperationalBase } from "./OperationalBase.js";
class Icac extends OperationalBase {
  /** Construct the class from a Tlv version of the certificate */
  static fromTlv(tlv) {
    return new Icac(OperationalCertificate.TlvIcac.decode(tlv));
  }
  /** Validates all basic certificate fields on construction. */
  validateFields() {
    const {
      extensions: {
        basicConstraints: { isCa }
      }
    } = this.cert;
    if (!isCa) {
      throw new CertificateError("Intermediate certificate must be a CA.");
    }
  }
  /**
   * Encodes the certificate with the signature as Matter Tlv.
   * If the certificate is not signed, it throws a CertificateError.
   */
  asSignedTlv() {
    return OperationalCertificate.TlvIcac.encode({ ...this.cert, signature: this.signature });
  }
  /**
   * Verify requirements a Matter Intermediate CA certificate must fulfill.
   * Rules for this are listed in @see {@link MatterSpecification.v12.Core} ยง6.5.x
   */
  async verify(crypto, root) {
    this.generalVerify();
    const {
      subject,
      issuer: { rcacId },
      extensions
    } = this.cert;
    const { fabricId, icacId } = subject;
    const { basicConstraints, extendedKeyUsage, subjectKeyIdentifier, authorityKeyIdentifier } = extensions;
    const { fabricId: rootFabricId } = root.cert.subject;
    if ("nodeId" in subject) {
      throw new CertificateError(`Ica certificate must not contain a nodeId.`);
    }
    if (fabricId !== void 0) {
      if (Array.isArray(fabricId)) {
        throw new CertificateError(`Invalid fabricId in NoC certificate: ${Diagnostic.json(fabricId)}`);
      }
      if (fabricId === FabricId(0)) {
        throw new CertificateError(`Invalid fabricId in NoC certificate: ${Diagnostic.json(fabricId)}`);
      }
    }
    if (icacId === void 0 || Array.isArray(icacId)) {
      throw new CertificateError(`Invalid icacId in Ica certificate: ${Diagnostic.json(icacId)}`);
    }
    if ("rcacId" in subject) {
      throw new CertificateError(`Ica certificate must not contain an rcacId.`);
    }
    if ("caseAuthenticatedTags" in subject) {
      throw new CertificateError(`Ica certificate must not contain a caseAuthenticatedTags.`);
    }
    if (rootFabricId !== void 0 && fabricId !== void 0 && rootFabricId !== fabricId) {
      throw new CertificateError(
        `FabricId in Ica certificate does not match the fabricId in the parent certificate. ${Diagnostic.json(
          rootFabricId
        )} !== ${Diagnostic.json(fabricId)}`
      );
    }
    if (root.cert.subject.rcacId !== rcacId) {
      throw new CertificateError(
        `RcacId in Ica certificate does not match the rcacId in the parent certificate. ${Diagnostic.json(
          root.cert.subject.rcacId
        )} !== ${Diagnostic.json(rcacId)}`
      );
    }
    if (!basicConstraints.isCa) {
      throw new CertificateError(`Ica certificate must have isCa set to true.`);
    }
    const keyUsage = ExtensionKeyUsageSchema.encode(extensions.keyUsage);
    if (keyUsage !== 96 && keyUsage !== 97) {
      throw new CertificateError(
        `Ica certificate keyUsage must have keyCertSign and CRLSign and optionally digitalSignature set.`
      );
    }
    if (extendedKeyUsage !== void 0) {
      throw new CertificateError(`Ica certificate must not have extendedKeyUsage set.`);
    }
    if (subjectKeyIdentifier === void 0) {
      throw new CertificateError(`Ica certificate must have subjectKeyIdentifier set.`);
    }
    if (subjectKeyIdentifier.length !== 20) {
      throw new CertificateError(`Ica certificate subjectKeyIdentifier must be 160 bit.`);
    }
    if (authorityKeyIdentifier === void 0) {
      throw new CertificateError(`Ica certificate must have authorityKeyIdentifier set.`);
    }
    if (authorityKeyIdentifier.length !== 20) {
      throw new CertificateError(`Ica certificate authorityKeyIdentifier must be 160 bit.`);
    }
    if (!Bytes.areEqual(authorityKeyIdentifier, root.cert.extensions.subjectKeyIdentifier)) {
      throw new CertificateError(
        `Ica certificate authorityKeyIdentifier must be equal to root cert subjectKeyIdentifier.`
      );
    }
    await crypto.verifyEcdsa(PublicKey(root.cert.ellipticCurvePublicKey), this.asUnsignedAsn1(), this.signature);
  }
}
export {
  Icac
};
