// @matter/node/dist/esm/node/client/ClientNodeInteraction.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { EndpointInitializer } from "#endpoint/properties/EndpointInitializer.js";
import { ClientInteraction } from "#protocol";
class ClientNodeInteraction {
  #node;
  constructor(node) {
    this.#node = node;
  }
  async *read(request, context) {
    request = this.structure.injectVersionFilters(request);
    const interaction = await this.#connect();
    const response = interaction.read(request, context);
    yield* this.structure.mutate(request, response);
  }
  async subscribe(request, context) {
    const intermediateRequest = {
      ...this.structure.injectVersionFilters(request),
      updated: async (data) => {
        const result = this.structure.mutate(request, data);
        if (request.updated) {
          await request.updated(result);
        } else {
          for await (const _chunk of result) ;
        }
      },
      closed(cause) {
        request.closed?.(cause);
      }
    };
    const interaction = await this.#connect();
    return interaction.subscribe(intermediateRequest, context);
  }
  async write(request, context) {
    return (await this.#connect()).write(request, context);
  }
  async *invoke(request, context) {
    yield* (await this.#connect()).invoke(request, context);
  }
  async #connect() {
    if (!this.#node.lifecycle.isOnline) {
      await this.#node.start();
    }
    return this.#node.env.get(ClientInteraction);
  }
  get structure() {
    return this.#node.env.get(EndpointInitializer).structure;
  }
}
export {
  ClientNodeInteraction
};
