// @matter/protocol/dist/cjs/certificate/kinds/AttestationCertificates.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AttestationCertificates_exports = {};
__export(AttestationCertificates_exports, {
  AttestationBaseCertificate: () => AttestationBaseCertificate,
  Dac: () => Dac,
  Paa: () => Paa,
  Pai: () => Pai
});
module.exports = __toCommonJS(AttestationCertificates_exports);
var import_general = require("#general");
var import_common = require("./common.js");
var import_X509Base = require("./X509Base.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class AttestationBaseCertificate extends import_X509Base.X509Base {
  /**
   * Sign the certificate using the provided crypto and key.
   * If the certificate is already signed, it throws a CertificateError.
   */
  async sign(crypto, key) {
    this.signature = await crypto.signEcdsa(key, this.asUnsignedAsn1(), "der");
  }
  /**
   * Returns the signed certificate in ASN.1 DER format.
   * If the certificate is not signed, it throws a CertificateError.
   */
  asSignedAsn1() {
    const certificate = this.genericBuildAsn1Structure(this.cert);
    const certBytes = import_general.DerCodec.encode({
      certificate,
      signAlgorithm: import_general.X962.EcdsaWithSHA256,
      signature: (0, import_general.DerBitString)(this.signature)
    });
    (0, import_common.assertCertificateDerSize)(certBytes);
    return certBytes;
  }
}
class Paa extends AttestationBaseCertificate {
}
class Pai extends AttestationBaseCertificate {
}
class Dac extends AttestationBaseCertificate {
}
