// @matter/protocol/dist/cjs/certificate/kinds/CertificationDeclaration.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CertificationDeclaration_exports = {};
__export(CertificationDeclaration_exports, {
  CertificationDeclaration: () => CertificationDeclaration
});
module.exports = __toCommonJS(CertificationDeclaration_exports);
var import_general = require("#general");
var import_common = require("./common.js");
var import_certification_declaration = require("./definitions/certification-declaration.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TestCMS_SignerPrivateKey = import_general.Bytes.fromHex("AEF3484116E9481EC57BE0472DF41BF499064E5024AD869ECA5E889802D48075");
const TestCMS_SignerSubjectKeyIdentifier = import_general.Bytes.fromHex("62FA823359ACFAA9963E1CFA140ADDF504F37160");
class CertificationDeclaration {
  #eContent;
  #subjectKeyIdentifier;
  /**
   * Generator which is the main usage for the class from outside.
   * It constructs the class with the relevant details and returns a signed ASN.1 DER version of the CD.
   */
  static generate(crypto, vendorId, productId, provisional = false) {
    const cd = new CertificationDeclaration(
      {
        formatVersion: 1,
        vendorId,
        produceIdArray: [productId],
        deviceTypeId: 22,
        certificateId: "CSA00000SWC00000-00",
        securityLevel: 0,
        securityInformation: 0,
        versionNumber: 1,
        certificationType: provisional ? 1 : 0
        // 0 = Test, 1 = Provisional/In certification, 2 = official
      },
      TestCMS_SignerSubjectKeyIdentifier
    );
    return cd.asSignedAsn1(crypto, (0, import_general.PrivateKey)(TestCMS_SignerPrivateKey));
  }
  constructor(content, subjectKeyIdentifier) {
    this.#eContent = import_certification_declaration.CertificationDeclaration.TlvDc.encode(content);
    this.#subjectKeyIdentifier = subjectKeyIdentifier;
  }
  /**
   * Returns the signed certificate in ASN.1 DER format.
   */
  async asSignedAsn1(crypto, privateKey) {
    const cert = {
      version: 3,
      digestAlgorithm: [import_general.SHA256_CMS],
      encapContentInfo: import_general.Pkcs7.Data(this.#eContent),
      signerInfo: [
        {
          version: 3,
          subjectKeyIdentifier: (0, import_general.ContextTaggedBytes)(0, this.#subjectKeyIdentifier),
          digestAlgorithm: import_general.SHA256_CMS,
          signatureAlgorithm: import_general.X962.EcdsaWithSHA256,
          signature: await crypto.signEcdsa(privateKey, this.#eContent, "der")
        }
      ]
    };
    const certBytes = import_general.DerCodec.encode(import_general.Pkcs7.SignedData(cert));
    (0, import_common.assertCertificateDerSize)(certBytes);
    return certBytes;
  }
}
