// @matter/protocol/dist/cjs/certificate/kinds/Icac.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Icac_exports = {};
__export(Icac_exports, {
  Icac: () => Icac
});
module.exports = __toCommonJS(Icac_exports);
var import_general = require("#general");
var import_types = require("#types");
var import_common = require("./common.js");
var import_base = require("./definitions/base.js");
var import_operational = require("./definitions/operational.js");
var import_OperationalBase = require("./OperationalBase.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class Icac extends import_OperationalBase.OperationalBase {
  /** Construct the class from a Tlv version of the certificate */
  static fromTlv(tlv) {
    return new Icac(import_operational.OperationalCertificate.TlvIcac.decode(tlv));
  }
  /** Validates all basic certificate fields on construction. */
  validateFields() {
    const {
      extensions: {
        basicConstraints: { isCa }
      }
    } = this.cert;
    if (!isCa) {
      throw new import_common.CertificateError("Intermediate certificate must be a CA.");
    }
  }
  /**
   * Encodes the certificate with the signature as Matter Tlv.
   * If the certificate is not signed, it throws a CertificateError.
   */
  asSignedTlv() {
    return import_operational.OperationalCertificate.TlvIcac.encode({ ...this.cert, signature: this.signature });
  }
  /**
   * Verify requirements a Matter Intermediate CA certificate must fulfill.
   * Rules for this are listed in @see {@link MatterSpecification.v12.Core} ยง6.5.x
   */
  async verify(crypto, root) {
    this.generalVerify();
    const {
      subject,
      issuer: { rcacId },
      extensions
    } = this.cert;
    const { fabricId, icacId } = subject;
    const { basicConstraints, extendedKeyUsage, subjectKeyIdentifier, authorityKeyIdentifier } = extensions;
    const { fabricId: rootFabricId } = root.cert.subject;
    if ("nodeId" in subject) {
      throw new import_common.CertificateError(`Ica certificate must not contain a nodeId.`);
    }
    if (fabricId !== void 0) {
      if (Array.isArray(fabricId)) {
        throw new import_common.CertificateError(`Invalid fabricId in NoC certificate: ${import_general.Diagnostic.json(fabricId)}`);
      }
      if (fabricId === (0, import_types.FabricId)(0)) {
        throw new import_common.CertificateError(`Invalid fabricId in NoC certificate: ${import_general.Diagnostic.json(fabricId)}`);
      }
    }
    if (icacId === void 0 || Array.isArray(icacId)) {
      throw new import_common.CertificateError(`Invalid icacId in Ica certificate: ${import_general.Diagnostic.json(icacId)}`);
    }
    if ("rcacId" in subject) {
      throw new import_common.CertificateError(`Ica certificate must not contain an rcacId.`);
    }
    if ("caseAuthenticatedTags" in subject) {
      throw new import_common.CertificateError(`Ica certificate must not contain a caseAuthenticatedTags.`);
    }
    if (rootFabricId !== void 0 && fabricId !== void 0 && rootFabricId !== fabricId) {
      throw new import_common.CertificateError(
        `FabricId in Ica certificate does not match the fabricId in the parent certificate. ${import_general.Diagnostic.json(
          rootFabricId
        )} !== ${import_general.Diagnostic.json(fabricId)}`
      );
    }
    if (root.cert.subject.rcacId !== rcacId) {
      throw new import_common.CertificateError(
        `RcacId in Ica certificate does not match the rcacId in the parent certificate. ${import_general.Diagnostic.json(
          root.cert.subject.rcacId
        )} !== ${import_general.Diagnostic.json(rcacId)}`
      );
    }
    if (!basicConstraints.isCa) {
      throw new import_common.CertificateError(`Ica certificate must have isCa set to true.`);
    }
    const keyUsage = import_base.ExtensionKeyUsageSchema.encode(extensions.keyUsage);
    if (keyUsage !== 96 && keyUsage !== 97) {
      throw new import_common.CertificateError(
        `Ica certificate keyUsage must have keyCertSign and CRLSign and optionally digitalSignature set.`
      );
    }
    if (extendedKeyUsage !== void 0) {
      throw new import_common.CertificateError(`Ica certificate must not have extendedKeyUsage set.`);
    }
    if (subjectKeyIdentifier === void 0) {
      throw new import_common.CertificateError(`Ica certificate must have subjectKeyIdentifier set.`);
    }
    if (subjectKeyIdentifier.length !== 20) {
      throw new import_common.CertificateError(`Ica certificate subjectKeyIdentifier must be 160 bit.`);
    }
    if (authorityKeyIdentifier === void 0) {
      throw new import_common.CertificateError(`Ica certificate must have authorityKeyIdentifier set.`);
    }
    if (authorityKeyIdentifier.length !== 20) {
      throw new import_common.CertificateError(`Ica certificate authorityKeyIdentifier must be 160 bit.`);
    }
    if (!import_general.Bytes.areEqual(authorityKeyIdentifier, root.cert.extensions.subjectKeyIdentifier)) {
      throw new import_common.CertificateError(
        `Ica certificate authorityKeyIdentifier must be equal to root cert subjectKeyIdentifier.`
      );
    }
    await crypto.verifyEcdsa((0, import_general.PublicKey)(root.cert.ellipticCurvePublicKey), this.asUnsignedAsn1(), this.signature);
  }
}
