// @matter/protocol/dist/cjs/events/OccurrenceManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OccurrenceManager_exports = {};
__export(OccurrenceManager_exports, {
  OccurrenceManager: () => OccurrenceManager
});
module.exports = __toCommonJS(OccurrenceManager_exports);
var import_general = require("#general");
var import_types = require("#types");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("OccurrenceManager");
class OccurrenceManager {
  #store;
  #bufferConfig;
  #cull;
  #iteratingValuesInProgress = false;
  #added = new import_general.Observable();
  // As we don't (yet) have storage with secondary indices we currently maintain indices in memory regardless of
  // whether underlying store is volatile
  #occurrences = new Array();
  #construction;
  get construction() {
    return this.#construction;
  }
  static async create(context) {
    return (0, import_general.asyncNew)(OccurrenceManager, context);
  }
  constructor({ store, bufferConfig }) {
    if (bufferConfig === void 0) {
      bufferConfig = OccurrenceManager.DefaultBufferConfig;
    }
    const { minEventAllowance, maxEventAllowance } = bufferConfig;
    if (maxEventAllowance < minEventAllowance) {
      throw new import_general.ImplementationError(
        `Maximum allowance ${maxEventAllowance} is less than minimum allowance ${minEventAllowance}`
      );
    }
    this.#store = store;
    this.#bufferConfig = bufferConfig;
    this.#construction = (0, import_general.Construction)(this, () => {
      return import_general.MaybePromise.then(this.#store.load(), (index) => {
        index.sort(
          // sort that way because Bigint & Number mix
          ({ number: numberA }, { number: numberB }) => numberA < numberB ? -1 : numberA > numberB ? 1 : 0
        );
        this.#occurrences.push(...index);
        if (this.#occurrences.length > this.#bufferConfig.minEventAllowance) {
          this.#startCull();
        }
      });
    });
  }
  get added() {
    return this.#added;
  }
  async clear() {
    await this.construction;
    await this.#store.clear();
    this.#occurrences.length = 0;
  }
  /**
   * Find the index of the first event number in the list that is greater than or equal to eventMin to optimize
   * searching.
   */
  #findMinEventNumberIndex(eventMin) {
    if (this.#occurrences.length === 0 || eventMin > this.#occurrences[this.#occurrences.length - 1].number) {
      return -1;
    }
    if (eventMin <= this.#occurrences[0].number) {
      return 0;
    }
    let low = this.#occurrences.length - 1;
    let high = low;
    let step = 1;
    while (low > 0 && this.#occurrences[low].number >= eventMin) {
      high = low;
      low = Math.max(0, low - step);
      step *= 2;
    }
    while (low <= high) {
      const mid = Math.floor((low + high) / 2);
      if (this.#occurrences[mid].number < eventMin) {
        low = mid + 1;
      } else {
        high = mid - 1;
      }
    }
    return low;
  }
  /**
   * Query the event store for events matching the given path and filters.
   * @deprecated
   */
  query(eventPath, filters, filterForFabricIndex) {
    if (filters !== void 0 && filters.length > 1) {
      throw new import_general.InternalError("Multiple filters not supported");
    }
    const startIndex = filters?.length ? this.#findMinEventNumberIndex((0, import_types.EventNumber)(filters[0].eventMin)) : 0;
    if (startIndex === -1) {
      return [];
    }
    let isAsyncLoad = false;
    const { endpointId, clusterId, eventId } = eventPath;
    const occurrences = new Array();
    for (let i = startIndex; i < this.#occurrences.length; i++) {
      const entry = this.#occurrences[i];
      if (endpointId === entry.endpointId && clusterId === entry.clusterId && eventId === entry.eventId) {
        let occurrence = this.#store.get(entry.number);
        occurrence = import_general.MaybePromise.then(occurrence, (occurrence2) => {
          occurrence2.number = entry.number;
          return occurrence2;
        });
        occurrences.push(occurrence);
        if (import_general.MaybePromise.is(occurrence)) {
          isAsyncLoad = true;
        }
      }
    }
    logger.debug(
      `Got ${occurrences.length} events for ${(0, import_types.resolveEventName)(eventPath)} with filters: ${import_general.Diagnostic.json(filters)}`
    );
    let result;
    if (isAsyncLoad) {
      result = Promise.all(occurrences);
    } else {
      result = occurrences;
    }
    if (filterForFabricIndex !== void 0) {
      result = import_general.MaybePromise.then(
        result,
        (occurrences2) => occurrences2.filter(({ payload }) => {
          if (!(0, import_general.isObject)(payload)) {
            return true;
          }
          const { fabricIndex } = payload;
          return fabricIndex === void 0 || fabricIndex === filterForFabricIndex;
        })
      );
    }
    result = import_general.MaybePromise.then(result, (occurrences2) => {
      let filterDesc;
      if (filters?.length) {
        filterDesc = filters.map(({ nodeId, eventMin }) => `${nodeId}:${eventMin}`).join(",");
      }
      if (filterForFabricIndex) {
        if (filterDesc !== void 0) {
          filterDesc = `${filterDesc} and fabric #${filterForFabricIndex}`;
        } else {
          filterDesc = `fabric #${filterForFabricIndex}`;
        }
      }
      if (filterDesc) {
        filterDesc = ` (filters: ${filterDesc})`;
      } else {
        filterDesc = "";
      }
      logger.debug(
        `Loaded ${occurrences2.length} occurrence${occurrences2.length === 1 ? "" : "s"} for ${(0, import_types.resolveEventName)(eventPath)}${filterDesc}`
      );
      return occurrences2;
    });
    return result;
  }
  /**
   * Return an iterator over all occurrences in the store that are bigger or equal to the minimum eventNumber,
   * if provided.
   */
  async *get(eventMin) {
    if (this.#cull) {
      await this.#cull;
    }
    const startIndex = eventMin === void 0 ? 0 : this.#findMinEventNumberIndex(eventMin);
    if (startIndex === -1) {
      return;
    }
    this.#iteratingValuesInProgress = true;
    try {
      for (let i = startIndex; i < this.#occurrences.length; i++) {
        const eventNumber = this.#occurrences[i].number;
        const occurrence = this.#store.get(eventNumber);
        if (import_general.MaybePromise.is(occurrence)) {
          yield {
            ...await occurrence,
            number: eventNumber
          };
        } else {
          yield {
            ...occurrence,
            number: eventNumber
          };
        }
      }
    } finally {
      this.#iteratingValuesInProgress = false;
    }
  }
  // For ServerNode usage the ServerEnvironment takes care to close the OccurrenceManager
  close() {
    return import_general.MaybePromise.then(this.#cull, () => this.#store.close());
  }
  add(occurrence) {
    return import_general.MaybePromise.then(this.#store.add(occurrence), (entry) => {
      logger.debug(`Recorded event #${entry.number}: ${import_general.Diagnostic.json(occurrence)}`);
      this.#occurrences.push(entry);
      if (this.#occurrences.length > this.#bufferConfig.maxEventAllowance) {
        this.#startCull();
      }
      const numberedOccurrence = {
        number: entry.number,
        ...occurrence
      };
      this.#added.emit(numberedOccurrence);
      return numberedOccurrence;
    });
  }
  remove(number) {
    const index = this.#occurrences.findIndex((entry) => entry.number === number);
    if (index === -1) {
      return;
    }
    this.#occurrences.splice(index, 1);
    if (this.#cull) {
      return this.#cull.then(() => this.#store.delete(number));
    }
    return this.#store.delete(number);
  }
  #startCull() {
    if (this.#cull || this.#iteratingValuesInProgress) {
      return;
    }
    const cull = this.#dropOldOccurrences();
    if (import_general.MaybePromise.is(cull)) {
      this.#cull = Promise.resolve(cull).finally(() => this.#cull = void 0);
    }
  }
  #dropOldOccurrences() {
    let toDelete = this.#occurrences.length - this.#bufferConfig.minEventAllowance;
    if (toDelete <= 0) {
      return;
    }
    logger.debug(`Event store is full; dropping ${toDelete} old occurrence${toDelete === 1 ? "s" : ""}`);
    const prioData = {
      [import_types.EventPriority.Info]: {
        count: this.#bufferConfig.minPriorityEventAllowance["info"],
        minPosition: -1
      },
      [import_types.EventPriority.Debug]: {
        count: this.#bufferConfig.minPriorityEventAllowance["debug"],
        minPosition: -1
      }
    };
    const asyncDrops = Array();
    for (let i = this.#occurrences.length - 1; i >= 0; i--) {
      const { priority } = this.#occurrences[i];
      if (priority !== import_types.EventPriority.Critical) {
        const data = prioData[priority];
        if (data.count > 0) {
          data.count--;
          if (data.count === 0) {
            data.minPosition = i;
            if (prioData[import_types.EventPriority.Info].minPosition > -1 && prioData[import_types.EventPriority.Debug].minPosition > -1) {
              break;
            }
          }
        }
      }
    }
    const occurrences = this.#occurrences;
    for (const priority of [import_types.EventPriority.Debug, import_types.EventPriority.Info, import_types.EventPriority.Critical]) {
      const checkUpTo = priority === import_types.EventPriority.Critical ? this.#occurrences.length : prioData[priority].minPosition;
      if (checkUpTo === -1) {
        continue;
      }
      for (let i = 0; i < checkUpTo && toDelete > 0; i++) {
        if (occurrences[i] === void 0) continue;
        const { priority: entryPriority, number } = occurrences[i];
        if (entryPriority === priority) {
          const drop = import_general.MaybePromise.catch(
            this.#store.delete(number),
            (error) => logger.warn(`Error dropping occurrence #${number}: ${error}`)
          );
          if (import_general.MaybePromise.is(drop)) {
            asyncDrops.push(drop);
          }
          occurrences[i] = void 0;
          toDelete--;
        }
      }
      if (toDelete <= 0) {
        break;
      }
    }
    this.#occurrences = occurrences.filter((entry) => entry);
    if (asyncDrops.length) {
      return import_general.MatterAggregateError.allSettled(asyncDrops, "Error dropping occurrences").then(() => {
      }).catch((error) => logger.error(error));
    }
  }
}
((OccurrenceManager2) => {
  OccurrenceManager2.DefaultBufferConfig = {
    minEventAllowance: 1e4,
    maxEventAllowance: 11e3,
    minPriorityEventAllowance: {
      info: 2e3,
      debug: 2e3
    }
  };
})(OccurrenceManager || (OccurrenceManager = {}));
