// @matter/model/dist/esm/standard/elements/administrator-commissioning.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  CommandElement as Command,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const AdministratorCommissioning = Cluster(
  { name: "AdministratorCommissioning", id: 60 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "BC", constraint: "0", title: "Basic" })
  ),
  Attribute({ name: "WindowStatus", id: 0, type: "CommissioningWindowStatusEnum", access: "R V", conformance: "M" }),
  Attribute({ name: "AdminFabricIndex", id: 1, type: "fabric-idx", access: "R V", conformance: "M", quality: "X" }),
  Attribute({ name: "AdminVendorId", id: 2, type: "vendor-id", access: "R V", conformance: "M", quality: "X" }),
  Command(
    {
      name: "OpenCommissioningWindow",
      id: 0,
      access: "A T",
      conformance: "M",
      direction: "request",
      response: "status"
    },
    Field({ name: "CommissioningTimeout", id: 0, type: "uint16", conformance: "M", constraint: "desc" }),
    Field({ name: "PakePasscodeVerifier", id: 1, type: "octstr", conformance: "M", constraint: "97" }),
    Field({ name: "Discriminator", id: 2, type: "uint16", conformance: "M", constraint: "0 to 4095" }),
    Field({ name: "Iterations", id: 3, type: "uint32", conformance: "M", constraint: "1000 to 100000" }),
    Field({ name: "Salt", id: 4, type: "octstr", conformance: "M", constraint: "16 to 32" })
  ),
  Command(
    {
      name: "OpenBasicCommissioningWindow",
      id: 1,
      access: "A T",
      conformance: "BC",
      direction: "request",
      response: "status"
    },
    Field({ name: "CommissioningTimeout", id: 0, type: "uint16", conformance: "M", constraint: "desc" })
  ),
  Command({ name: "RevokeCommissioning", id: 2, access: "A T", conformance: "M", direction: "request", response: "status" }),
  Datatype(
    { name: "CommissioningWindowStatusEnum", type: "enum8" },
    Field({ name: "WindowNotOpen", id: 0, conformance: "M" }),
    Field({ name: "EnhancedWindowOpen", id: 1, conformance: "M" }),
    Field({ name: "BasicWindowOpen", id: 2, conformance: "BC" })
  ),
  Datatype(
    { name: "StatusCodeEnum", type: "enum8" },
    Field({ name: "Busy", id: 2, conformance: "M" }),
    Field({ name: "PakeParameterError", id: 3, conformance: "M" }),
    Field({ name: "WindowNotOpen", id: 4, conformance: "M" })
  )
);
MatterDefinition.children.push(AdministratorCommissioning);
export {
  AdministratorCommissioning
};
