// @matter/model/dist/esm/standard/elements/air-quality.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const AirQuality = Cluster(
  { name: "AirQuality", id: 91 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "FAIR", conformance: "O", constraint: "0", title: "Fair" }),
    Field({ name: "MOD", conformance: "O", constraint: "1", title: "Moderate" }),
    Field({ name: "VPOOR", conformance: "O", constraint: "2", title: "VeryPoor" }),
    Field({ name: "XPOOR", conformance: "O", constraint: "3", title: "ExtremelyPoor" })
  ),
  Attribute({
    name: "AirQuality",
    id: 0,
    type: "AirQualityEnum",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 0
  }),
  Datatype(
    { name: "AirQualityEnum", type: "enum8" },
    Field({ name: "Unknown", id: 0, conformance: "M" }),
    Field({ name: "Good", id: 1, conformance: "M" }),
    Field({ name: "Fair", id: 2, conformance: "FAIR" }),
    Field({ name: "Moderate", id: 3, conformance: "MOD" }),
    Field({ name: "Poor", id: 4, conformance: "M" }),
    Field({ name: "VeryPoor", id: 5, conformance: "VPOOR" }),
    Field({ name: "ExtremelyPoor", id: 6, conformance: "XPOOR" })
  )
);
MatterDefinition.children.push(AirQuality);
export {
  AirQuality
};
