// @matter/model/dist/esm/standard/elements/alarm-base.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  EventElement as Event,
  CommandElement as Command,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const AlarmBase = Cluster(
  { name: "AlarmBase" },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "RESET", constraint: "0", title: "Reset" })
  ),
  Attribute({ name: "Mask", id: 0, type: "AlarmBitmap", access: "R V", conformance: "M", default: 0 }),
  Attribute(
    { name: "Latch", id: 1, type: "AlarmBitmap", access: "R V", conformance: "RESET", default: 0, quality: "F" }
  ),
  Attribute({ name: "State", id: 2, type: "AlarmBitmap", access: "R V", conformance: "M", default: 0 }),
  Attribute(
    { name: "Supported", id: 3, type: "AlarmBitmap", access: "R V", conformance: "M", default: 0, quality: "F" }
  ),
  Event(
    { name: "Notify", id: 0, access: "V", conformance: "M", priority: "info" },
    Field({ name: "Active", id: 0, type: "AlarmBitmap", conformance: "M", default: 0 }),
    Field({ name: "Inactive", id: 1, type: "AlarmBitmap", conformance: "M", default: 0 }),
    Field({ name: "State", id: 2, type: "AlarmBitmap", conformance: "M", default: 0 }),
    Field({ name: "Mask", id: 3, type: "AlarmBitmap", conformance: "M", default: 0 })
  ),
  Command(
    { name: "Reset", id: 0, access: "O", conformance: "RESET", direction: "request", response: "status" },
    Field({ name: "Alarms", id: 0, type: "AlarmBitmap", conformance: "M", default: 0 })
  ),
  Command(
    {
      name: "ModifyEnabledAlarms",
      id: 1,
      access: "O",
      conformance: "O",
      direction: "request",
      response: "status"
    },
    Field({ name: "Mask", id: 0, type: "AlarmBitmap", conformance: "M", default: 0 })
  ),
  Datatype({ name: "AlarmBitmap", type: "map32" })
);
MatterDefinition.children.push(AlarmBase);
export {
  AlarmBase
};
