// @matter/model/dist/esm/standard/elements/application-basic.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const ApplicationBasic = Cluster(
  { name: "ApplicationBasic", id: 1293 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute({
    name: "VendorName",
    id: 0,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 32",
    quality: "F"
  }),
  Attribute({ name: "VendorId", id: 1, type: "vendor-id", access: "R V", conformance: "O", quality: "F" }),
  Attribute({
    name: "ApplicationName",
    id: 2,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    quality: "F"
  }),
  Attribute({ name: "ProductId", id: 3, type: "uint16", access: "R V", conformance: "O", quality: "F" }),
  Attribute({
    name: "Application",
    id: 4,
    type: "ApplicationStruct",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    quality: "F"
  }),
  Attribute(
    { name: "Status", id: 5, type: "ApplicationStatusEnum", access: "R V", conformance: "M", constraint: "desc" }
  ),
  Attribute({
    name: "ApplicationVersion",
    id: 6,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "max 32",
    quality: "F"
  }),
  Attribute(
    { name: "AllowedVendorList", id: 7, type: "list", access: "R A", conformance: "M", quality: "F" },
    Field({ name: "entry", type: "vendor-id" })
  ),
  Datatype(
    { name: "ApplicationStatusEnum", type: "enum8" },
    Field({ name: "Stopped", id: 0, conformance: "M" }),
    Field({ name: "ActiveVisibleFocus", id: 1, conformance: "M" }),
    Field({ name: "ActiveHidden", id: 2, conformance: "M" }),
    Field({ name: "ActiveVisibleNotFocus", id: 3, conformance: "M" })
  ),
  Datatype(
    { name: "ApplicationStruct", type: "struct" },
    Field({ name: "CatalogVendorId", id: 0, type: "uint16", conformance: "M" }),
    Field({ name: "ApplicationId", id: 1, type: "string", conformance: "M" })
  )
);
MatterDefinition.children.push(ApplicationBasic);
export {
  ApplicationBasic
};
