// @matter/model/dist/esm/standard/elements/ballast-configuration.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  DatatypeElement as Datatype,
  FieldElement as Field
} from "../../elements/index.js";
const BallastConfiguration = Cluster(
  { name: "BallastConfiguration", id: 769 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 4 }),
  Attribute({
    name: "PhysicalMinLevel",
    id: 0,
    type: "uint8",
    access: "R V",
    conformance: "M",
    constraint: "1 to 254",
    default: 1
  }),
  Attribute({
    name: "PhysicalMaxLevel",
    id: 1,
    type: "uint8",
    access: "R V",
    conformance: "M",
    constraint: "1 to 254",
    default: 254
  }),
  Attribute({ name: "BallastStatus", id: 2, type: "BallastStatusBitmap", access: "R V", conformance: "O", default: 0 }),
  Attribute({
    name: "MinLevel",
    id: 16,
    type: "uint8",
    access: "RW VM",
    conformance: "M",
    constraint: "physicalMinLevel to maxLevel",
    default: { type: "reference", name: "PhysicalMinLevel" }
  }),
  Attribute({
    name: "MaxLevel",
    id: 17,
    type: "uint8",
    access: "RW VM",
    conformance: "M",
    constraint: "minLevel to physicalMaxLevel",
    default: { type: "reference", name: "PhysicalMaxLevel" }
  }),
  Attribute({ name: "PowerOnLevel", id: 18, conformance: "D" }),
  Attribute({ name: "PowerOnFadeTime", id: 19, conformance: "D" }),
  Attribute({ name: "IntrinsicBallastFactor", id: 20, type: "uint8", access: "RW VM", conformance: "O", quality: "X" }),
  Attribute({
    name: "BallastFactorAdjustment",
    id: 21,
    type: "uint8",
    access: "RW VM",
    conformance: "O",
    constraint: "100 to ms",
    default: null,
    quality: "X"
  }),
  Attribute({ name: "LampQuantity", id: 32, type: "uint8", access: "R V", conformance: "M" }),
  Attribute({ name: "LampType", id: 48, type: "string", access: "RW VM", conformance: "O", constraint: "max 16" }),
  Attribute(
    { name: "LampManufacturer", id: 49, type: "string", access: "RW VM", conformance: "O", constraint: "max 16" }
  ),
  Attribute({
    name: "LampRatedHours",
    id: 50,
    type: "uint24",
    access: "RW VM",
    conformance: "O",
    default: null,
    quality: "X"
  }),
  Attribute({ name: "LampBurnHours", id: 51, type: "uint24", access: "RW VM", conformance: "O", default: 0, quality: "X" }),
  Attribute(
    { name: "LampAlarmMode", id: 52, type: "LampAlarmModeBitmap", access: "RW VM", conformance: "O", default: 0 }
  ),
  Attribute({
    name: "LampBurnHoursTripPoint",
    id: 53,
    type: "uint24",
    access: "RW VM",
    conformance: "O",
    default: null,
    quality: "X"
  }),
  Datatype(
    { name: "BallastStatusBitmap", type: "map8" },
    Field({ name: "BallastNonOperational", constraint: "0" }),
    Field({ name: "LampFailure", constraint: "1" })
  ),
  Datatype({ name: "LampAlarmModeBitmap", type: "map8" }, Field({ name: "LampBurnHours", constraint: "0" }))
);
MatterDefinition.children.push(BallastConfiguration);
export {
  BallastConfiguration
};
