// @matter/model/dist/esm/standard/elements/basic-information.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  EventElement as Event,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const BasicInformation = Cluster(
  { name: "BasicInformation", id: 40 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 4 }),
  Attribute({
    name: "DataModelRevision",
    id: 0,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    quality: "F"
  }),
  Attribute({
    name: "VendorName",
    id: 1,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "max 32",
    quality: "F"
  }),
  Attribute({ name: "VendorId", id: 2, type: "vendor-id", access: "R V", conformance: "M", quality: "F" }),
  Attribute({
    name: "ProductName",
    id: 3,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "max 32",
    quality: "F"
  }),
  Attribute({ name: "ProductId", id: 4, type: "uint16", access: "R V", conformance: "M", quality: "F" }),
  Attribute({
    name: "NodeLabel",
    id: 5,
    type: "string",
    access: "RW VM",
    conformance: "M",
    constraint: "max 32",
    quality: "N"
  }),
  Attribute({
    name: "Location",
    id: 6,
    type: "string",
    access: "RW VA",
    conformance: "M",
    constraint: "2",
    default: "XX",
    quality: "N"
  }),
  Attribute(
    { name: "HardwareVersion", id: 7, type: "uint16", access: "R V", conformance: "M", default: 0, quality: "F" }
  ),
  Attribute({
    name: "HardwareVersionString",
    id: 8,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "1 to 64",
    quality: "F"
  }),
  Attribute({
    name: "SoftwareVersion",
    id: 9,
    type: "uint32",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 0,
    quality: "F"
  }),
  Attribute({
    name: "SoftwareVersionString",
    id: 10,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "1 to 64",
    quality: "F"
  }),
  Attribute({
    name: "ManufacturingDate",
    id: 11,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "8 to 16",
    quality: "F"
  }),
  Attribute({
    name: "PartNumber",
    id: 12,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 32",
    quality: "F"
  }),
  Attribute({
    name: "ProductUrl",
    id: 13,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 256",
    quality: "F"
  }),
  Attribute({
    name: "ProductLabel",
    id: 14,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 64",
    quality: "F"
  }),
  Attribute({
    name: "SerialNumber",
    id: 15,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 32",
    quality: "F"
  }),
  Attribute({
    name: "LocalConfigDisabled",
    id: 16,
    type: "bool",
    access: "RW VM",
    conformance: "O",
    default: false,
    quality: "N"
  }),
  Attribute({ name: "Reachable", id: 17, type: "bool", access: "R V", conformance: "O", default: true }),
  Attribute({ name: "UniqueId", id: 18, type: "string", access: "R V", conformance: "M", constraint: "max 32", quality: "F" }),
  Attribute({
    name: "CapabilityMinima",
    id: 19,
    type: "CapabilityMinimaStruct",
    access: "R V",
    conformance: "M",
    quality: "F"
  }),
  Attribute({
    name: "ProductAppearance",
    id: 20,
    type: "ProductAppearanceStruct",
    access: "R V",
    conformance: "O",
    quality: "F"
  }),
  Attribute({
    name: "SpecificationVersion",
    id: 21,
    type: "uint32",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 0,
    quality: "F"
  }),
  Attribute({
    name: "MaxPathsPerInvoke",
    id: 22,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "min 1",
    default: 1,
    quality: "F"
  }),
  Event(
    { name: "StartUp", id: 0, access: "V", conformance: "M", priority: "critical" },
    Field({ name: "SoftwareVersion", id: 0, type: "uint32", conformance: "M" })
  ),
  Event({ name: "ShutDown", id: 1, access: "V", conformance: "O", priority: "critical" }),
  Event(
    { name: "Leave", id: 2, access: "V", conformance: "O", priority: "info" },
    Field({ name: "FabricIndex", id: 0, type: "fabric-idx", conformance: "M", constraint: "1 to 254" })
  ),
  Event(
    { name: "ReachableChanged", id: 3, access: "V", conformance: "desc", priority: "info" },
    Field({ name: "ReachableNewValue", id: 0, type: "bool", conformance: "M" })
  ),
  Datatype(
    { name: "ProductFinishEnum", type: "enum8" },
    Field({ name: "Other", id: 0, conformance: "M" }),
    Field({ name: "Matte", id: 1, conformance: "M" }),
    Field({ name: "Satin", id: 2, conformance: "M" }),
    Field({ name: "Polished", id: 3, conformance: "M" }),
    Field({ name: "Rugged", id: 4, conformance: "M" }),
    Field({ name: "Fabric", id: 5, conformance: "M" })
  ),
  Datatype(
    { name: "ColorEnum", type: "enum8" },
    Field({ name: "Black", id: 0, conformance: "M" }),
    Field({ name: "Navy", id: 1, conformance: "M" }),
    Field({ name: "Green", id: 2, conformance: "M" }),
    Field({ name: "Teal", id: 3, conformance: "M" }),
    Field({ name: "Maroon", id: 4, conformance: "M" }),
    Field({ name: "Purple", id: 5, conformance: "M" }),
    Field({ name: "Olive", id: 6, conformance: "M" }),
    Field({ name: "Gray", id: 7, conformance: "M" }),
    Field({ name: "Blue", id: 8, conformance: "M" }),
    Field({ name: "Lime", id: 9, conformance: "M" }),
    Field({ name: "Aqua", id: 10, conformance: "M" }),
    Field({ name: "Red", id: 11, conformance: "M" }),
    Field({ name: "Fuchsia", id: 12, conformance: "M" }),
    Field({ name: "Yellow", id: 13, conformance: "M" }),
    Field({ name: "White", id: 14, conformance: "M" }),
    Field({ name: "Nickel", id: 15, conformance: "M" }),
    Field({ name: "Chrome", id: 16, conformance: "M" }),
    Field({ name: "Brass", id: 17, conformance: "M" }),
    Field({ name: "Copper", id: 18, conformance: "M" }),
    Field({ name: "Silver", id: 19, conformance: "M" }),
    Field({ name: "Gold", id: 20, conformance: "M" })
  ),
  Datatype(
    { name: "ProductAppearanceStruct", type: "struct" },
    Field({ name: "Finish", id: 0, type: "ProductFinishEnum", conformance: "M" }),
    Field({ name: "PrimaryColor", id: 1, type: "ColorEnum", conformance: "M", quality: "X" })
  ),
  Datatype(
    { name: "CapabilityMinimaStruct", type: "struct" },
    Field({ name: "CaseSessionsPerFabric", id: 0, type: "uint16", conformance: "M", constraint: "min 3", default: 3 }),
    Field({ name: "SubscriptionsPerFabric", id: 1, type: "uint16", conformance: "M", constraint: "min 3", default: 3 })
  )
);
MatterDefinition.children.push(BasicInformation);
export {
  BasicInformation
};
