// @matter/model/dist/esm/standard/elements/binding.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const Binding = Cluster(
  { name: "Binding", id: 30 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    {
      name: "Binding",
      id: 0,
      type: "list",
      access: "RW F VM",
      conformance: "M",
      constraint: "desc",
      default: [],
      quality: "N"
    },
    Field({ name: "entry", type: "TargetStruct" })
  ),
  Datatype(
    { name: "TargetStruct", type: "struct" },
    Field({ name: "Node", id: 1, type: "node-id", access: "F", conformance: "Endpoint" }),
    Field({ name: "Group", id: 2, type: "group-id", access: "F", conformance: "!Endpoint", constraint: "min 1" }),
    Field({ name: "Endpoint", id: 3, type: "endpoint-no", access: "F", conformance: "!Group" }),
    Field({ name: "Cluster", id: 4, type: "cluster-id", access: "F", conformance: "O" }),
    Field({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  )
);
MatterDefinition.children.push(Binding);
export {
  Binding
};
