// @matter/model/dist/esm/standard/elements/descriptor.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const Descriptor = Cluster(
  { name: "Descriptor", id: 29 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "TAGLIST", conformance: "desc", constraint: "0", title: "TagList" })
  ),
  Attribute(
    {
      name: "DeviceTypeList",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "min 1",
      quality: "F"
    },
    Field({ name: "entry", type: "DeviceTypeStruct" })
  ),
  Attribute(
    { name: "ServerList", id: 1, type: "list", access: "R V", conformance: "M", default: [], quality: "F" },
    Field({ name: "entry", type: "cluster-id" })
  ),
  Attribute(
    { name: "ClientList", id: 2, type: "list", access: "R V", conformance: "M", default: [], quality: "F" },
    Field({ name: "entry", type: "cluster-id" })
  ),
  Attribute(
    { name: "PartsList", id: 3, type: "list", access: "R V", conformance: "M", default: [] },
    Field({ name: "entry", type: "endpoint-no" })
  ),
  Attribute(
    {
      name: "TagList",
      id: 4,
      type: "list",
      access: "R V",
      conformance: "TAGLIST",
      constraint: "1 to 6",
      quality: "F"
    },
    Field({ name: "entry", type: "semtag" })
  ),
  Datatype(
    { name: "DeviceTypeStruct", type: "struct" },
    Field({ name: "DeviceType", id: 0, type: "devtype-id", conformance: "M" }),
    Field({ name: "Revision", id: 1, type: "uint16", conformance: "M", constraint: "min 1" })
  )
);
MatterDefinition.children.push(Descriptor);
export {
  Descriptor
};
