// @matter/model/dist/esm/standard/elements/diagnostic-logs.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  CommandElement as Command,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const DiagnosticLogs = Cluster(
  { name: "DiagnosticLogs", id: 50 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Command(
    {
      name: "RetrieveLogsRequest",
      id: 0,
      access: "O",
      conformance: "M",
      direction: "request",
      response: "RetrieveLogsResponse"
    },
    Field({ name: "Intent", id: 0, type: "IntentEnum", conformance: "M" }),
    Field({ name: "RequestedProtocol", id: 1, type: "TransferProtocolEnum", conformance: "M" }),
    Field({ name: "TransferFileDesignator", id: 2, type: "string", conformance: "O", constraint: "max 32" })
  ),
  Command(
    { name: "RetrieveLogsResponse", id: 1, conformance: "M", direction: "response" },
    Field({ name: "Status", id: 0, type: "StatusEnum", conformance: "M" }),
    Field({ name: "LogContent", id: 1, type: "octstr", conformance: "M", constraint: "max 1024" }),
    Field({ name: "UtcTimeStamp", id: 2, type: "epoch-us", conformance: "O" }),
    Field({ name: "TimeSinceBoot", id: 3, type: "systime-us", conformance: "O" })
  ),
  Datatype(
    { name: "IntentEnum", type: "enum8" },
    Field({ name: "EndUserSupport", id: 0, conformance: "M" }),
    Field({ name: "NetworkDiag", id: 1, conformance: "M" }),
    Field({ name: "CrashLogs", id: 2, conformance: "M" })
  ),
  Datatype(
    { name: "StatusEnum", type: "enum8" },
    Field({ name: "Success", id: 0, conformance: "M" }),
    Field({ name: "Exhausted", id: 1, conformance: "M" }),
    Field({ name: "NoLogs", id: 2, conformance: "M" }),
    Field({ name: "Busy", id: 3, conformance: "M" }),
    Field({ name: "Denied", id: 4, conformance: "M" })
  ),
  Datatype(
    { name: "TransferProtocolEnum", type: "enum8" },
    Field({ name: "ResponsePayload", id: 0, conformance: "M" }),
    Field({ name: "Bdx", id: 1, conformance: "M" })
  )
);
MatterDefinition.children.push(DiagnosticLogs);
export {
  DiagnosticLogs
};
