// @matter/model/dist/esm/standard/elements/ecosystem-information.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const EcosystemInformation = Cluster(
  { name: "EcosystemInformation", id: 1872 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "DeviceDirectory", id: 0, type: "list", access: "R F M", conformance: "M", quality: "N" },
    Field({ name: "entry", type: "EcosystemDeviceStruct" })
  ),
  Attribute(
    { name: "LocationDirectory", id: 1, type: "list", access: "R F M", conformance: "M", quality: "N" },
    Field({ name: "entry", type: "EcosystemLocationStruct" })
  ),
  Datatype(
    { name: "EcosystemDeviceStruct", type: "struct" },
    Field({ name: "DeviceName", id: 0, type: "string", access: "S", conformance: "O", constraint: "max 64" }),
    Field({ name: "DeviceNameLastEdit", id: 1, type: "epoch-us", access: "S", conformance: "desc", default: 0 }),
    Field({ name: "BridgedEndpoint", id: 2, type: "endpoint-no", access: "S", conformance: "desc", constraint: "desc" }),
    Field({
      name: "OriginalEndpoint",
      id: 3,
      type: "endpoint-no",
      access: "S",
      conformance: "desc",
      constraint: "desc"
    }),
    Field(
      { name: "DeviceTypes", id: 4, type: "list", access: "S", conformance: "M", constraint: "desc" },
      Field({ name: "entry", type: "Descriptor.DeviceTypeStruct" })
    ),
    Field(
      {
        name: "UniqueLocationIDs",
        id: 5,
        type: "list",
        access: "S",
        conformance: "M",
        constraint: "max 64[max 64]"
      },
      Field({ name: "entry", type: "string" })
    ),
    Field({ name: "UniqueLocationIDsLastEdit", id: 6, type: "epoch-us", access: "S", conformance: "M", default: 0 }),
    Field({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  ),
  Datatype(
    { name: "EcosystemLocationStruct", type: "struct" },
    Field({ name: "UniqueLocationId", id: 0, type: "string", access: "S", conformance: "M", constraint: "max 64" }),
    Field({ name: "LocationDescriptor", id: 1, type: "locationdesc", access: "S", conformance: "M" }),
    Field({ name: "LocationDescriptorLastEdit", id: 2, type: "epoch-us", access: "S", conformance: "M", default: 0 }),
    Field({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  )
);
MatterDefinition.children.push(EcosystemInformation);
export {
  EcosystemInformation
};
