// @matter/model/dist/esm/standard/elements/ethernet-network-diagnostics.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  CommandElement as Command,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const EthernetNetworkDiagnostics = Cluster(
  { name: "EthernetNetworkDiagnostics", id: 55, quality: "K" },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "PKTCNT", constraint: "0", title: "PacketCounts" }),
    Field({ name: "ERRCNT", constraint: "1", title: "ErrorCounts" })
  ),
  Attribute(
    { name: "PhyRate", id: 0, type: "PHYRateEnum", access: "R V", conformance: "O", default: null, quality: "X" }
  ),
  Attribute({ name: "FullDuplex", id: 1, type: "bool", access: "R V", conformance: "O", default: null, quality: "X" }),
  Attribute({ name: "PacketRxCount", id: 2, type: "uint64", access: "R V", conformance: "PKTCNT", default: 0, quality: "C" }),
  Attribute({ name: "PacketTxCount", id: 3, type: "uint64", access: "R V", conformance: "PKTCNT", default: 0, quality: "C" }),
  Attribute(
    { name: "TxErrCount", id: 4, type: "uint64", access: "R V", conformance: "ERRCNT", default: 0, quality: "C" }
  ),
  Attribute({ name: "CollisionCount", id: 5, type: "uint64", access: "R V", conformance: "ERRCNT", default: 0, quality: "C" }),
  Attribute({ name: "OverrunCount", id: 6, type: "uint64", access: "R V", conformance: "ERRCNT", default: 0, quality: "C" }),
  Attribute({ name: "CarrierDetect", id: 7, type: "bool", access: "R V", conformance: "O", default: null, quality: "X C" }),
  Attribute(
    { name: "TimeSinceReset", id: 8, type: "uint64", access: "R V", conformance: "O", default: 0, quality: "C" }
  ),
  Command({
    name: "ResetCounts",
    id: 0,
    access: "M",
    conformance: "PKTCNT | ERRCNT",
    direction: "request",
    response: "status"
  }),
  Datatype(
    { name: "PHYRateEnum", type: "enum8" },
    Field({ name: "Rate10M", id: 0, conformance: "M" }),
    Field({ name: "Rate100M", id: 1, conformance: "M" }),
    Field({ name: "Rate1G", id: 2, conformance: "M" }),
    Field({ name: "Rate25G", id: 3, conformance: "M" }),
    Field({ name: "Rate5G", id: 4, conformance: "M" }),
    Field({ name: "Rate10G", id: 5, conformance: "M" }),
    Field({ name: "Rate40G", id: 6, conformance: "M" }),
    Field({ name: "Rate100G", id: 7, conformance: "M" }),
    Field({ name: "Rate200G", id: 8, conformance: "M" }),
    Field({ name: "Rate400G", id: 9, conformance: "M" })
  )
);
MatterDefinition.children.push(EthernetNetworkDiagnostics);
export {
  EthernetNetworkDiagnostics
};
