// @matter/model/dist/esm/standard/elements/flow-measurement.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { ClusterElement as Cluster, AttributeElement as Attribute } from "../../elements/index.js";
const FlowMeasurement = Cluster(
  { name: "FlowMeasurement", id: 1028 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  Attribute({
    name: "MeasuredValue",
    id: 0,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "minMeasuredValue to maxMeasuredValue",
    default: null,
    quality: "X P"
  }),
  Attribute({
    name: "MinMeasuredValue",
    id: 1,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "max 65533",
    quality: "X"
  }),
  Attribute({
    name: "MaxMeasuredValue",
    id: 2,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "min minMeasuredValue + 1",
    quality: "X"
  }),
  Attribute({ name: "Tolerance", id: 3, type: "uint16", access: "R V", conformance: "O", constraint: "max 2048", default: 0 })
);
MatterDefinition.children.push(FlowMeasurement);
export {
  FlowMeasurement
};
