// @matter/model/dist/esm/standard/elements/illuminance-measurement.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  DatatypeElement as Datatype,
  FieldElement as Field
} from "../../elements/index.js";
const IlluminanceMeasurement = Cluster(
  { name: "IlluminanceMeasurement", id: 1024 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  Attribute({
    name: "MeasuredValue",
    id: 0,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "0, minMeasuredValue to maxMeasuredValue",
    default: 0,
    quality: "X P"
  }),
  Attribute({
    name: "MinMeasuredValue",
    id: 1,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "1 to 65533",
    quality: "X"
  }),
  Attribute({
    name: "MaxMeasuredValue",
    id: 2,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "min minMeasuredValue + 1",
    quality: "X"
  }),
  Attribute({ name: "Tolerance", id: 3, type: "uint16", access: "R V", conformance: "O", constraint: "max 2048" }),
  Attribute({ name: "LightSensorType", id: 4, type: "uint8", access: "R V", conformance: "O", default: null, quality: "X" }),
  Datatype(
    { name: "LightSensorTypeEnum", type: "enum8" },
    Field({ name: "Photodiode", id: 0, conformance: "M" }),
    Field({ name: "Cmos", id: 1, conformance: "M" })
  )
);
MatterDefinition.children.push(IlluminanceMeasurement);
export {
  IlluminanceMeasurement
};
