// @matter/model/dist/esm/standard/elements/joint-fabric-pki.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  CommandElement as Command,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const JointFabricPki = Cluster(
  { name: "JointFabricPki", id: 1875 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Command(
    {
      name: "IcacsrRequest",
      id: 0,
      access: "A",
      conformance: "M",
      direction: "request",
      response: "IcacsrResponse"
    },
    Field({ name: "Icacsr", id: 0, type: "octstr", conformance: "M", constraint: "max 400" })
  ),
  Command(
    { name: "IcacsrResponse", id: 1, access: "A", conformance: "M", direction: "response" },
    Field({ name: "StatusCode", id: 0, type: "IcacsrRequestStatusEnum", conformance: "M" }),
    Field({ name: "Icac", id: 1, type: "octstr", conformance: "O", constraint: "max 400" })
  ),
  Command({
    name: "TransferAnchorRequest",
    id: 2,
    access: "A",
    conformance: "M",
    direction: "request",
    response: "TransferAnchorResponse"
  }),
  Command({ name: "TransferAnchorResponse", id: 3, access: "A", conformance: "M", direction: "response" }),
  Command({
    name: "TransferAnchorComplete",
    id: 4,
    access: "A",
    conformance: "M",
    direction: "request",
    response: "status"
  }),
  Datatype(
    { name: "IcacsrRequestStatusEnum", type: "enum8" },
    Field({ name: "Ok", id: 0, conformance: "M" }),
    Field({ name: "InvalidIcaCsrFormat", id: 1, conformance: "M" }),
    Field({ name: "InvalidIcaCsrSignature", id: 2, conformance: "M" }),
    Field({ name: "FailedDclVendorIdValidation", id: 3, conformance: "M" }),
    Field({ name: "NotAnIcac", id: 4, conformance: "M" }),
    Field({ name: "BusyAnchorTransfer", id: 5, conformance: "M" }),
    Field({ name: "IcaCsrSigningFailed", id: 6, conformance: "M" }),
    Field({ name: "IcaCsrRequestNoUserConsent", id: 7, conformance: "M" })
  ),
  Datatype(
    { name: "TransferAnchorResponseStatusEnum", type: "enum8" },
    Field({ name: "Ok", id: 0, conformance: "M" }),
    Field({ name: "TransferAnchorStatusDatastoreBusy", id: 1, conformance: "M" }),
    Field({ name: "TransferAnchorStatusNoUserConsent", id: 2, conformance: "M" })
  )
);
MatterDefinition.children.push(JointFabricPki);
export {
  JointFabricPki
};
