// @matter/model/dist/esm/standard/elements/laundry-washer-controls.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const LaundryWasherControls = Cluster(
  { name: "LaundryWasherControls", id: 83 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "SPIN", conformance: "O.a+", constraint: "0", title: "Spin" }),
    Field({ name: "RINSE", conformance: "O.a+", constraint: "1", title: "Rinse" })
  ),
  Attribute(
    { name: "SpinSpeeds", id: 0, type: "list", access: "R V", conformance: "SPIN", constraint: "max 16[max 64]" },
    Field({ name: "entry", type: "string" })
  ),
  Attribute({
    name: "SpinSpeedCurrent",
    id: 1,
    type: "uint8",
    access: "RW VO",
    conformance: "SPIN",
    constraint: "max 15",
    quality: "X"
  }),
  Attribute({
    name: "NumberOfRinses",
    id: 2,
    type: "NumberOfRinsesEnum",
    access: "RW VO",
    conformance: "RINSE",
    constraint: "desc",
    default: 1
  }),
  Attribute(
    { name: "SupportedRinses", id: 3, type: "list", access: "R V", conformance: "RINSE", constraint: "max 4" },
    Field({ name: "entry", type: "NumberOfRinsesEnum" })
  ),
  Datatype(
    { name: "NumberOfRinsesEnum", type: "enum8" },
    Field({ name: "None", id: 0, conformance: "RINSE" }),
    Field({ name: "Normal", id: 1, conformance: "RINSE" }),
    Field({ name: "Extra", id: 2, conformance: "RINSE" }),
    Field({ name: "Max", id: 3, conformance: "RINSE" })
  )
);
MatterDefinition.children.push(LaundryWasherControls);
export {
  LaundryWasherControls
};
