// @matter/model/dist/esm/standard/elements/laundry-dryer-controls.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const LaundryDryerControls = Cluster(
  { name: "LaundryDryerControls", id: 74 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    {
      name: "SupportedDrynessLevels",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "1 to 4"
    },
    Field({ name: "entry", type: "DrynessLevelEnum" })
  ),
  Attribute({
    name: "SelectedDrynessLevel",
    id: 1,
    type: "DrynessLevelEnum",
    access: "RW VO",
    conformance: "M",
    constraint: "desc",
    quality: "X"
  }),
  Datatype(
    { name: "DrynessLevelEnum", type: "enum8" },
    Field({ name: "Low", id: 0, conformance: "M" }),
    Field({ name: "Normal", id: 1, conformance: "M" }),
    Field({ name: "Extra", id: 2, conformance: "M" }),
    Field({ name: "Max", id: 3, conformance: "M" })
  )
);
MatterDefinition.children.push(LaundryDryerControls);
export {
  LaundryDryerControls
};
