// @matter/model/dist/esm/standard/elements/measurement-accuracy-range-struct.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DatatypeElement as Datatype, FieldElement as Field } from "../../elements/index.js";
const MeasurementAccuracyRangeStruct = Datatype(
  { name: "MeasurementAccuracyRangeStruct", type: "struct" },
  Field({ name: "RangeMin", id: 0, type: "int64", conformance: "M", quality: "F" }),
  Field({ name: "RangeMax", id: 1, type: "int64", conformance: "M", quality: "F" }),
  Field({ name: "PercentMax", id: 2, type: "percent100ths", conformance: "O.a+", quality: "F" }),
  Field({
    name: "PercentMin",
    id: 3,
    type: "percent100ths",
    conformance: "[PercentMax]",
    constraint: "max percentTypical",
    quality: "F"
  }),
  Field({
    name: "PercentTypical",
    id: 4,
    type: "percent100ths",
    conformance: "[PercentMin]",
    constraint: "percentMin to percentMax",
    quality: "F"
  }),
  Field({ name: "FixedMax", id: 5, type: "uint64", conformance: "O.a+", quality: "F" }),
  Field({ name: "FixedMin", id: 6, type: "uint64", conformance: "[FixedMax]", constraint: "max fixedMax", quality: "F" }),
  Field({
    name: "FixedTypical",
    id: 7,
    type: "uint64",
    conformance: "[FixedMin]",
    constraint: "fixedMin to fixedMax",
    quality: "F"
  })
);
MatterDefinition.children.push(MeasurementAccuracyRangeStruct);
export {
  MeasurementAccuracyRangeStruct
};
