// @matter/model/dist/esm/standard/elements/power-topology.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field
} from "../../elements/index.js";
const PowerTopology = Cluster(
  { name: "PowerTopology", id: 156 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "NODE", conformance: "O.a", constraint: "0", title: "NodeTopology" }),
    Field({ name: "TREE", conformance: "O.a", constraint: "1", title: "TreeTopology" }),
    Field({ name: "SET", conformance: "O.a", constraint: "2", title: "SetTopology" }),
    Field({ name: "DYPF", conformance: "[SET]", constraint: "3", title: "DynamicPowerFlow" })
  ),
  Attribute(
    {
      name: "AvailableEndpoints",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "SET",
      constraint: "max 20",
      quality: "F"
    },
    Field({ name: "entry", type: "endpoint-no" })
  ),
  Attribute(
    {
      name: "ActiveEndpoints",
      id: 1,
      type: "list",
      access: "R V",
      conformance: "DYPF",
      constraint: "max 20",
      quality: "N"
    },
    Field({ name: "entry", type: "endpoint-no" })
  )
);
MatterDefinition.children.push(PowerTopology);
export {
  PowerTopology
};
