// @matter/model/dist/esm/standard/elements/pressure-measurement.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field
} from "../../elements/index.js";
const PressureMeasurement = Cluster(
  { name: "PressureMeasurement", id: 1027 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "EXT", conformance: "O", constraint: "0", title: "Extended" })
  ),
  Attribute({
    name: "MeasuredValue",
    id: 0,
    type: "int16",
    access: "R V",
    conformance: "M",
    constraint: "minMeasuredValue to maxMeasuredValue",
    quality: "X P"
  }),
  Attribute({
    name: "MinMeasuredValue",
    id: 1,
    type: "int16",
    access: "R V",
    conformance: "M",
    constraint: "max 32766",
    quality: "X"
  }),
  Attribute({
    name: "MaxMeasuredValue",
    id: 2,
    type: "int16",
    access: "R V",
    conformance: "M",
    constraint: "minMeasuredValue + 1 to 32767",
    quality: "X"
  }),
  Attribute({ name: "Tolerance", id: 3, type: "uint16", access: "R V", conformance: "O", constraint: "max 2048", default: 0 }),
  Attribute({
    name: "ScaledValue",
    id: 16,
    type: "int16",
    access: "R V",
    conformance: "EXT",
    constraint: "minScaledValue to maxScaledValue",
    default: 0,
    quality: "X"
  }),
  Attribute({
    name: "MinScaledValue",
    id: 17,
    type: "int16",
    access: "R V",
    conformance: "EXT",
    constraint: "max 32766",
    default: 0,
    quality: "X"
  }),
  Attribute({
    name: "MaxScaledValue",
    id: 18,
    type: "int16",
    access: "R V",
    conformance: "EXT",
    constraint: "minScaledValue + 1 to 32767",
    default: 0,
    quality: "X"
  }),
  Attribute({
    name: "ScaledTolerance",
    id: 19,
    type: "uint16",
    access: "R V",
    conformance: "[EXT]",
    constraint: "max 2048",
    default: 0
  }),
  Attribute({ name: "Scale", id: 20, type: "int8", access: "R V", conformance: "EXT", constraint: "min -127", default: 0 })
);
MatterDefinition.children.push(PressureMeasurement);
export {
  PressureMeasurement
};
