// @matter/model/dist/esm/standard/elements/proxy-configuration.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const ProxyConfiguration = Cluster(
  { name: "ProxyConfiguration", id: 66 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "ConfigurationList", id: 0, type: "list", access: "RW", conformance: "M", default: [], quality: "N" },
    Field({ name: "entry", type: "ConfigurationStruct" })
  ),
  Datatype(
    { name: "ConfigurationStruct", type: "struct" },
    Field({
      name: "ProxyAllNodes",
      id: 1,
      type: "bool",
      access: "RW",
      conformance: "M",
      constraint: "desc",
      default: false
    }),
    Field(
      {
        name: "SourceList",
        id: 2,
        type: "list",
        access: "RW",
        conformance: "M",
        constraint: "desc",
        default: []
      },
      Field({ name: "entry", type: "node-id" })
    )
  )
);
MatterDefinition.children.push(ProxyConfiguration);
export {
  ProxyConfiguration
};
