// @matter/model/dist/esm/standard/elements/relative-humidity-measurement.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { ClusterElement as Cluster, AttributeElement as Attribute } from "../../elements/index.js";
const RelativeHumidityMeasurement = Cluster(
  { name: "RelativeHumidityMeasurement", id: 1029 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  Attribute({
    name: "MeasuredValue",
    id: 0,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "minMeasuredValue to maxMeasuredValue",
    quality: "X P"
  }),
  Attribute({
    name: "MinMeasuredValue",
    id: 1,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "max 9999",
    quality: "X"
  }),
  Attribute({
    name: "MaxMeasuredValue",
    id: 2,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "minMeasuredValue + 1 to 10000",
    quality: "X"
  }),
  Attribute({ name: "Tolerance", id: 3, type: "uint16", access: "R V", conformance: "O", constraint: "max 2048" })
);
MatterDefinition.children.push(RelativeHumidityMeasurement);
export {
  RelativeHumidityMeasurement
};
