// @matter/model/dist/esm/standard/elements/root-node.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  DeviceTypeElement as DeviceType,
  RequirementElement as Requirement,
  FieldElement as Field
} from "../../elements/index.js";
const RootNodeDt = DeviceType(
  { name: "RootNode", id: 22 },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 22, revision: 3 }], element: "attribute" })
  ),
  Requirement({ name: "BasicInformation", id: 40, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement(
    { name: "AccessControl", id: 31, conformance: "M", element: "serverCluster", quality: "I" },
    Requirement({ name: "MNGD", conformance: "[ManagedAcl, Allowed]", constraint: "desc", element: "feature" })
  ),
  Requirement(
    { name: "PowerSourceConfiguration", id: 46, conformance: "O, D", element: "serverCluster", quality: "I" }
  ),
  Requirement({ name: "TimeSynchronization", id: 56, conformance: "O", element: "serverCluster", quality: "I" }),
  Requirement({ name: "GroupKeyManagement", id: 63, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({ name: "GeneralCommissioning", id: 48, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({ name: "NetworkCommissioning", id: 49, conformance: "!CustomNetworkConfig", element: "serverCluster" }),
  Requirement({ name: "AdministratorCommissioning", id: 60, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({ name: "OperationalCredentials", id: 62, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({
    name: "LocalizationConfiguration",
    id: 43,
    conformance: "LanguageLocale",
    element: "serverCluster",
    quality: "I"
  }),
  Requirement(
    { name: "TimeFormatLocalization", id: 44, conformance: "TimeLocale", element: "serverCluster", quality: "I" }
  ),
  Requirement({ name: "UnitLocalization", id: 45, conformance: "UnitLocale", element: "serverCluster", quality: "I" }),
  Requirement({ name: "GeneralDiagnostics", id: 51, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({ name: "DiagnosticLogs", id: 50, conformance: "O", element: "serverCluster", quality: "I" }),
  Requirement({ name: "SoftwareDiagnostics", id: 52, conformance: "O", element: "serverCluster", quality: "I" }),
  Requirement({ name: "EthernetNetworkDiagnostics", id: 55, conformance: "[Ethernet]", element: "serverCluster" }),
  Requirement({ name: "WiFiNetworkDiagnostics", id: 54, conformance: "[Wi, Fi]", element: "serverCluster" }),
  Requirement({ name: "ThreadNetworkDiagnostics", id: 53, conformance: "[Thread]", element: "serverCluster" }),
  Requirement(
    { name: "IcdManagement", id: 70, conformance: "SIT | LIT", element: "serverCluster", quality: "I" },
    Requirement({ name: "LONGIDLETIMESUPPORT", conformance: "LIT", element: "feature" })
  ),
  Field(
    { name: "conditions", type: "enum8" },
    Field({ name: "CustomNetworkConfig" }),
    Field({ name: "ManagedAclAllowed" })
  )
);
MatterDefinition.children.push(RootNodeDt);
export {
  RootNodeDt
};
