// @matter/model/dist/esm/standard/elements/temperature-measurement.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { ClusterElement as Cluster, AttributeElement as Attribute } from "../../elements/index.js";
const TemperatureMeasurement = Cluster(
  { name: "TemperatureMeasurement", id: 1026 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 4 }),
  Attribute({
    name: "MeasuredValue",
    id: 0,
    type: "temperature",
    access: "R V",
    conformance: "M",
    constraint: "minMeasuredValue to maxMeasuredValue",
    quality: "X P"
  }),
  Attribute({
    name: "MinMeasuredValue",
    id: 1,
    type: "temperature",
    access: "R V",
    conformance: "M",
    constraint: "-27315 to maxMeasuredValue - 1",
    default: -27315,
    quality: "X"
  }),
  Attribute({
    name: "MaxMeasuredValue",
    id: 2,
    type: "temperature",
    access: "R V",
    conformance: "M",
    constraint: "minMeasuredValue + 1 to 32767",
    default: 32767,
    quality: "X"
  }),
  Attribute({ name: "Tolerance", id: 3, type: "uint16", access: "R V", conformance: "O", constraint: "max 2048", default: 0 })
);
MatterDefinition.children.push(TemperatureMeasurement);
export {
  TemperatureMeasurement
};
