// @matter/model/dist/esm/standard/elements/thermostat-user-interface-configuration.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  DatatypeElement as Datatype,
  FieldElement as Field
} from "../../elements/index.js";
const ThermostatUserInterfaceConfiguration = Cluster(
  { name: "ThermostatUserInterfaceConfiguration", id: 516 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  Attribute({
    name: "TemperatureDisplayMode",
    id: 0,
    type: "TemperatureDisplayModeEnum",
    access: "RW VO",
    conformance: "M",
    default: 0
  }),
  Attribute({ name: "KeypadLockout", id: 1, type: "KeypadLockoutEnum", access: "RW VM", conformance: "M", default: 0 }),
  Attribute({
    name: "ScheduleProgrammingVisibility",
    id: 2,
    type: "ScheduleProgrammingVisibilityEnum",
    access: "RW VM",
    conformance: "O",
    default: 0
  }),
  Datatype(
    { name: "TemperatureDisplayModeEnum", type: "enum8" },
    Field({ name: "Celsius", id: 0, conformance: "M" }),
    Field({ name: "Fahrenheit", id: 1, conformance: "M" })
  ),
  Datatype(
    { name: "KeypadLockoutEnum", type: "enum8" },
    Field({ name: "NoLockout", id: 0, conformance: "M" }),
    Field({ name: "Lockout1", id: 1, conformance: "M" }),
    Field({ name: "Lockout2", id: 2, conformance: "M" }),
    Field({ name: "Lockout3", id: 3, conformance: "M" }),
    Field({ name: "Lockout4", id: 4, conformance: "M" }),
    Field({ name: "Lockout5", id: 5, conformance: "M" })
  ),
  Datatype(
    { name: "ScheduleProgrammingVisibilityEnum", type: "enum8" },
    Field({ name: "ScheduleProgrammingPermitted", id: 0, conformance: "M" }),
    Field({ name: "ScheduleProgrammingDenied", id: 1, conformance: "M" })
  )
);
MatterDefinition.children.push(ThermostatUserInterfaceConfiguration);
export {
  ThermostatUserInterfaceConfiguration
};
