// @matter/model/dist/esm/standard/elements/thread-border-router-management.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  CommandElement as Command
} from "../../elements/index.js";
const ThreadBorderRouterManagement = Cluster(
  { name: "ThreadBorderRouterManagement", id: 1106 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "PC", conformance: "O", constraint: "0", title: "PanChange" })
  ),
  Attribute(
    { name: "BorderRouterName", id: 0, type: "string", access: "R V", conformance: "M", constraint: "1 to 63" }
  ),
  Attribute({ name: "BorderAgentId", id: 1, type: "octstr", access: "R V", conformance: "M", constraint: "16" }),
  Attribute({ name: "ThreadVersion", id: 2, type: "uint16", access: "R V", conformance: "M", quality: "F" }),
  Attribute({ name: "InterfaceEnabled", id: 3, type: "bool", access: "R V", conformance: "M", default: false, quality: "N" }),
  Attribute({
    name: "ActiveDatasetTimestamp",
    id: 4,
    type: "uint64",
    access: "R V",
    conformance: "M",
    default: 0,
    quality: "X N"
  }),
  Attribute({
    name: "PendingDatasetTimestamp",
    id: 5,
    type: "uint64",
    access: "R V",
    conformance: "M",
    default: 0,
    quality: "X N"
  }),
  Command({
    name: "GetActiveDatasetRequest",
    id: 0,
    access: "M",
    conformance: "M",
    direction: "request",
    response: "DatasetResponse"
  }),
  Command({
    name: "GetPendingDatasetRequest",
    id: 1,
    access: "M",
    conformance: "M",
    direction: "request",
    response: "DatasetResponse"
  }),
  Command(
    { name: "DatasetResponse", id: 2, conformance: "M", direction: "response" },
    Field({ name: "Dataset", id: 0, type: "octstr", conformance: "M", constraint: "max 254" })
  ),
  Command(
    {
      name: "SetActiveDatasetRequest",
      id: 3,
      access: "M T",
      conformance: "M",
      direction: "request",
      response: "status"
    },
    Field({ name: "ActiveDataset", id: 0, type: "octstr", conformance: "M", constraint: "max 254" }),
    Field({ name: "Breadcrumb", id: 1, type: "uint64", conformance: "O" })
  ),
  Command(
    {
      name: "SetPendingDatasetRequest",
      id: 4,
      access: "M T",
      conformance: "PC",
      direction: "request",
      response: "status"
    },
    Field({ name: "PendingDataset", id: 0, type: "octstr", conformance: "M", constraint: "max 254" })
  )
);
MatterDefinition.children.push(ThreadBorderRouterManagement);
export {
  ThreadBorderRouterManagement
};
