// @matter/model/dist/esm/standard/elements/time-format-localization.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const TimeFormatLocalization = Cluster(
  { name: "TimeFormatLocalization", id: 44 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "CALFMT", constraint: "0", title: "CalendarFormat" })
  ),
  Attribute({ name: "HourFormat", id: 0, type: "HourFormatEnum", access: "RW VM", conformance: "M", quality: "N" }),
  Attribute({
    name: "ActiveCalendarType",
    id: 1,
    type: "CalendarTypeEnum",
    access: "RW VM",
    conformance: "CALFMT",
    constraint: "in SupportedCalendarTypes",
    quality: "N"
  }),
  Attribute(
    {
      name: "SupportedCalendarTypes",
      id: 2,
      type: "list",
      access: "R V",
      conformance: "CALFMT",
      constraint: "desc",
      quality: "F"
    },
    Field({ name: "entry", type: "CalendarTypeEnum" })
  ),
  Datatype(
    { name: "HourFormatEnum", type: "enum8" },
    Field({ name: "12Hr", id: 0, conformance: "M" }),
    Field({ name: "24Hr", id: 1, conformance: "M" }),
    Field({ name: "UseActiveLocale", id: 255, conformance: "M" })
  ),
  Datatype(
    { name: "CalendarTypeEnum", type: "enum8" },
    Field({ name: "Buddhist", id: 0, conformance: "O.a+" }),
    Field({ name: "Chinese", id: 1, conformance: "O.a+" }),
    Field({ name: "Coptic", id: 2, conformance: "O.a+" }),
    Field({ name: "Ethiopian", id: 3, conformance: "O.a+" }),
    Field({ name: "Gregorian", id: 4, conformance: "O.a+" }),
    Field({ name: "Hebrew", id: 5, conformance: "O.a+" }),
    Field({ name: "Indian", id: 6, conformance: "O.a+" }),
    Field({ name: "Islamic", id: 7, conformance: "O.a+" }),
    Field({ name: "Japanese", id: 8, conformance: "O.a+" }),
    Field({ name: "Korean", id: 9, conformance: "O.a+" }),
    Field({ name: "Persian", id: 10, conformance: "O.a+" }),
    Field({ name: "Taiwanese", id: 11, conformance: "O.a+" }),
    Field({ name: "UseActiveLocale", id: 255, conformance: "O.a+" })
  )
);
MatterDefinition.children.push(TimeFormatLocalization);
export {
  TimeFormatLocalization
};
