// @matter/model/dist/esm/standard/elements/wi-fi-network-management.element.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  CommandElement as Command,
  FieldElement as Field
} from "../../elements/index.js";
const WiFiNetworkManagement = Cluster(
  { name: "WiFiNetworkManagement", id: 1105 },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute({
    name: "Ssid",
    id: 0,
    type: "octstr",
    access: "R V",
    conformance: "M",
    constraint: "1 to 32",
    default: null,
    quality: "X N"
  }),
  Attribute({
    name: "PassphraseSurrogate",
    id: 1,
    type: "uint64",
    access: "R M",
    conformance: "M",
    default: null,
    quality: "X N"
  }),
  Command({
    name: "NetworkPassphraseRequest",
    id: 0,
    access: "M",
    conformance: "M",
    direction: "request",
    response: "NetworkPassphraseResponse"
  }),
  Command(
    { name: "NetworkPassphraseResponse", id: 1, conformance: "M", direction: "response" },
    Field({ name: "Passphrase", id: 0, type: "octstr", conformance: "M", constraint: "max 64" })
  )
);
MatterDefinition.children.push(WiFiNetworkManagement);
export {
  WiFiNetworkManagement
};
