// @matter/model/dist/cjs/standard/elements/accepted-command-list.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var accepted_command_list_element_exports = {};
__export(accepted_command_list_element_exports, {
  AcceptedCommandList: () => AcceptedCommandList
});
module.exports = __toCommonJS(accepted_command_list_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AcceptedCommandList = (0, import_elements.AttributeElement)(
  {
    name: "AcceptedCommandList",
    id: 65529,
    type: "list",
    access: "R V",
    conformance: "M",
    isSeed: true,
    quality: "F"
  },
  (0, import_elements.FieldElement)({ name: "entry", type: "command-id" })
);
import_MatterDefinition.MatterDefinition.children.push(AcceptedCommandList);
