// @matter/model/dist/cjs/standard/elements/account-login.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var account_login_element_exports = {};
__export(account_login_element_exports, {
  AccountLogin: () => AccountLogin
});
module.exports = __toCommonJS(account_login_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AccountLogin = (0, import_elements.ClusterElement)(
  { name: "AccountLogin", id: 1294 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.EventElement)(
    { name: "LoggedOut", id: 0, access: "S A", conformance: "O", priority: "critical" },
    (0, import_elements.FieldElement)({ name: "Node", id: 0, type: "node-id", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "GetSetupPin",
      id: 0,
      access: "F A T",
      conformance: "M",
      direction: "request",
      response: "GetSetupPinResponse"
    },
    (0, import_elements.FieldElement)({ name: "TempAccountIdentifier", id: 0, type: "string", conformance: "M", constraint: "16 to 100" })
  ),
  (0, import_elements.CommandElement)(
    { name: "GetSetupPinResponse", id: 1, access: "F", conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "SetupPin", id: 0, type: "string", conformance: "M", constraint: "desc" })
  ),
  (0, import_elements.CommandElement)(
    { name: "Login", id: 2, access: "F A T", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "TempAccountIdentifier", id: 0, type: "string", conformance: "M", constraint: "16 to 100" }),
    (0, import_elements.FieldElement)({ name: "SetupPin", id: 1, type: "string", conformance: "M", constraint: "min 8" }),
    (0, import_elements.FieldElement)({ name: "Node", id: 2, type: "node-id", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    { name: "Logout", id: 3, access: "F O T", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "Node", id: 0, type: "node-id", conformance: "O" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(AccountLogin);
