// @matter/model/dist/cjs/standard/elements/actions.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var actions_element_exports = {};
__export(actions_element_exports, {
  Actions: () => Actions
});
module.exports = __toCommonJS(actions_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const Actions = (0, import_elements.ClusterElement)(
  { name: "Actions", id: 37 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    {
      name: "ActionList",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "max 256",
      default: []
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "ActionStruct" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "EndpointLists",
      id: 1,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "max 256",
      default: []
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "EndpointListStruct" })
  ),
  (0, import_elements.AttributeElement)({ name: "SetupUrl", id: 2, type: "string", access: "R V", conformance: "O", constraint: "max 512" }),
  (0, import_elements.EventElement)(
    { name: "StateChanged", id: 0, access: "V", conformance: "M", priority: "info" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NewState", id: 2, type: "ActionStateEnum", conformance: "M" })
  ),
  (0, import_elements.EventElement)(
    { name: "ActionFailed", id: 1, access: "V", conformance: "M", priority: "info" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NewState", id: 2, type: "ActionStateEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Error", id: 3, type: "ActionErrorEnum", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "InstantAction", id: 0, access: "O", conformance: "desc", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "InstantActionWithTransition",
      id: 1,
      access: "O",
      conformance: "desc",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint16", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "StartAction", id: 2, access: "O", conformance: "desc", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "StartActionWithDuration",
      id: 3,
      access: "O",
      conformance: "desc",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Duration", id: 2, type: "uint32", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "StopAction", id: 4, access: "O", conformance: "desc", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    { name: "PauseAction", id: 5, access: "O", conformance: "desc", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "PauseActionWithDuration",
      id: 6,
      access: "O",
      conformance: "desc",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Duration", id: 2, type: "uint32", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "ResumeAction", id: 7, access: "O", conformance: "desc", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    { name: "EnableAction", id: 8, access: "O", conformance: "desc", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "EnableActionWithDuration",
      id: 9,
      access: "O",
      conformance: "desc",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Duration", id: 2, type: "uint32", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "DisableAction", id: 10, access: "O", conformance: "desc", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "DisableActionWithDuration",
      id: 11,
      access: "O",
      conformance: "desc",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvokeId", id: 1, type: "uint32", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Duration", id: 2, type: "uint32", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "CommandBits", type: "map16" },
    (0, import_elements.FieldElement)({ name: "InstantAction", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "InstantActionWithTransition", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "StartAction", constraint: "2" }),
    (0, import_elements.FieldElement)({ name: "StartActionWithDuration", constraint: "3" }),
    (0, import_elements.FieldElement)({ name: "StopAction", constraint: "4" }),
    (0, import_elements.FieldElement)({ name: "PauseAction", constraint: "5" }),
    (0, import_elements.FieldElement)({ name: "PauseActionWithDuration", constraint: "6" }),
    (0, import_elements.FieldElement)({ name: "ResumeAction", constraint: "7" }),
    (0, import_elements.FieldElement)({ name: "EnableAction", constraint: "8" }),
    (0, import_elements.FieldElement)({ name: "EnableActionWithDuration", constraint: "9" }),
    (0, import_elements.FieldElement)({ name: "DisableAction", constraint: "10" }),
    (0, import_elements.FieldElement)({ name: "DisableActionWithDuration", constraint: "11" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ActionTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Other", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Scene", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Sequence", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Automation", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Exception", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Notification", id: 5, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Alarm", id: 6, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ActionStateEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Inactive", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Active", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Paused", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Disabled", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ActionErrorEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unknown", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Interrupted", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "EndpointListTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Other", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Room", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Zone", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ActionStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "ActionId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Name", id: 1, type: "string", conformance: "M", constraint: "max 128{32}" }),
    (0, import_elements.FieldElement)({ name: "Type", id: 2, type: "ActionTypeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndpointListId", id: 3, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SupportedCommands", id: 4, type: "CommandBits", conformance: "M", constraint: "0 to 4095" }),
    (0, import_elements.FieldElement)({ name: "State", id: 5, type: "ActionStateEnum", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "EndpointListStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "EndpointListId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Name", id: 1, type: "string", conformance: "M", constraint: "max 128{32}" }),
    (0, import_elements.FieldElement)({ name: "Type", id: 2, type: "EndpointListTypeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)(
      { name: "Endpoints", id: 3, type: "list", conformance: "M", constraint: "max 256" },
      (0, import_elements.FieldElement)({ name: "entry", type: "endpoint-no" })
    )
  )
);
import_MatterDefinition.MatterDefinition.children.push(Actions);
