// @matter/model/dist/cjs/standard/elements/administrator-commissioning.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var administrator_commissioning_element_exports = {};
__export(administrator_commissioning_element_exports, {
  AdministratorCommissioning: () => AdministratorCommissioning
});
module.exports = __toCommonJS(administrator_commissioning_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AdministratorCommissioning = (0, import_elements.ClusterElement)(
  { name: "AdministratorCommissioning", id: 60 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "BC", constraint: "0", title: "Basic" })
  ),
  (0, import_elements.AttributeElement)({ name: "WindowStatus", id: 0, type: "CommissioningWindowStatusEnum", access: "R V", conformance: "M" }),
  (0, import_elements.AttributeElement)({ name: "AdminFabricIndex", id: 1, type: "fabric-idx", access: "R V", conformance: "M", quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "AdminVendorId", id: 2, type: "vendor-id", access: "R V", conformance: "M", quality: "X" }),
  (0, import_elements.CommandElement)(
    {
      name: "OpenCommissioningWindow",
      id: 0,
      access: "A T",
      conformance: "M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "CommissioningTimeout", id: 0, type: "uint16", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "PakePasscodeVerifier", id: 1, type: "octstr", conformance: "M", constraint: "97" }),
    (0, import_elements.FieldElement)({ name: "Discriminator", id: 2, type: "uint16", conformance: "M", constraint: "0 to 4095" }),
    (0, import_elements.FieldElement)({ name: "Iterations", id: 3, type: "uint32", conformance: "M", constraint: "1000 to 100000" }),
    (0, import_elements.FieldElement)({ name: "Salt", id: 4, type: "octstr", conformance: "M", constraint: "16 to 32" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "OpenBasicCommissioningWindow",
      id: 1,
      access: "A T",
      conformance: "BC",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "CommissioningTimeout", id: 0, type: "uint16", conformance: "M", constraint: "desc" })
  ),
  (0, import_elements.CommandElement)({ name: "RevokeCommissioning", id: 2, access: "A T", conformance: "M", direction: "request", response: "status" }),
  (0, import_elements.DatatypeElement)(
    { name: "CommissioningWindowStatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "WindowNotOpen", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EnhancedWindowOpen", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "BasicWindowOpen", id: 2, conformance: "BC" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StatusCodeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Busy", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "PakeParameterError", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "WindowNotOpen", id: 4, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(AdministratorCommissioning);
