// @matter/model/dist/cjs/standard/elements/air-quality.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var air_quality_element_exports = {};
__export(air_quality_element_exports, {
  AirQuality: () => AirQuality
});
module.exports = __toCommonJS(air_quality_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AirQuality = (0, import_elements.ClusterElement)(
  { name: "AirQuality", id: 91 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "FAIR", conformance: "O", constraint: "0", title: "Fair" }),
    (0, import_elements.FieldElement)({ name: "MOD", conformance: "O", constraint: "1", title: "Moderate" }),
    (0, import_elements.FieldElement)({ name: "VPOOR", conformance: "O", constraint: "2", title: "VeryPoor" }),
    (0, import_elements.FieldElement)({ name: "XPOOR", conformance: "O", constraint: "3", title: "ExtremelyPoor" })
  ),
  (0, import_elements.AttributeElement)({
    name: "AirQuality",
    id: 0,
    type: "AirQualityEnum",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 0
  }),
  (0, import_elements.DatatypeElement)(
    { name: "AirQualityEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unknown", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Good", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Fair", id: 2, conformance: "FAIR" }),
    (0, import_elements.FieldElement)({ name: "Moderate", id: 3, conformance: "MOD" }),
    (0, import_elements.FieldElement)({ name: "Poor", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "VeryPoor", id: 5, conformance: "VPOOR" }),
    (0, import_elements.FieldElement)({ name: "ExtremelyPoor", id: 6, conformance: "XPOOR" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(AirQuality);
