// @matter/model/dist/cjs/standard/elements/alarm-base.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var alarm_base_element_exports = {};
__export(alarm_base_element_exports, {
  AlarmBase: () => AlarmBase
});
module.exports = __toCommonJS(alarm_base_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AlarmBase = (0, import_elements.ClusterElement)(
  { name: "AlarmBase" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "RESET", constraint: "0", title: "Reset" })
  ),
  (0, import_elements.AttributeElement)({ name: "Mask", id: 0, type: "AlarmBitmap", access: "R V", conformance: "M", default: 0 }),
  (0, import_elements.AttributeElement)(
    { name: "Latch", id: 1, type: "AlarmBitmap", access: "R V", conformance: "RESET", default: 0, quality: "F" }
  ),
  (0, import_elements.AttributeElement)({ name: "State", id: 2, type: "AlarmBitmap", access: "R V", conformance: "M", default: 0 }),
  (0, import_elements.AttributeElement)(
    { name: "Supported", id: 3, type: "AlarmBitmap", access: "R V", conformance: "M", default: 0, quality: "F" }
  ),
  (0, import_elements.EventElement)(
    { name: "Notify", id: 0, access: "V", conformance: "M", priority: "info" },
    (0, import_elements.FieldElement)({ name: "Active", id: 0, type: "AlarmBitmap", conformance: "M", default: 0 }),
    (0, import_elements.FieldElement)({ name: "Inactive", id: 1, type: "AlarmBitmap", conformance: "M", default: 0 }),
    (0, import_elements.FieldElement)({ name: "State", id: 2, type: "AlarmBitmap", conformance: "M", default: 0 }),
    (0, import_elements.FieldElement)({ name: "Mask", id: 3, type: "AlarmBitmap", conformance: "M", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "Reset", id: 0, access: "O", conformance: "RESET", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "Alarms", id: 0, type: "AlarmBitmap", conformance: "M", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "ModifyEnabledAlarms",
      id: 1,
      access: "O",
      conformance: "O",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "Mask", id: 0, type: "AlarmBitmap", conformance: "M", default: 0 })
  ),
  (0, import_elements.DatatypeElement)({ name: "AlarmBitmap", type: "map32" })
);
import_MatterDefinition.MatterDefinition.children.push(AlarmBase);
