// @matter/model/dist/cjs/standard/elements/application-basic.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var application_basic_element_exports = {};
__export(application_basic_element_exports, {
  ApplicationBasic: () => ApplicationBasic
});
module.exports = __toCommonJS(application_basic_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ApplicationBasic = (0, import_elements.ClusterElement)(
  { name: "ApplicationBasic", id: 1293 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)({
    name: "VendorName",
    id: 0,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 32",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({ name: "VendorId", id: 1, type: "vendor-id", access: "R V", conformance: "O", quality: "F" }),
  (0, import_elements.AttributeElement)({
    name: "ApplicationName",
    id: 2,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({ name: "ProductId", id: 3, type: "uint16", access: "R V", conformance: "O", quality: "F" }),
  (0, import_elements.AttributeElement)({
    name: "Application",
    id: 4,
    type: "ApplicationStruct",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)(
    { name: "Status", id: 5, type: "ApplicationStatusEnum", access: "R V", conformance: "M", constraint: "desc" }
  ),
  (0, import_elements.AttributeElement)({
    name: "ApplicationVersion",
    id: 6,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "max 32",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)(
    { name: "AllowedVendorList", id: 7, type: "list", access: "R A", conformance: "M", quality: "F" },
    (0, import_elements.FieldElement)({ name: "entry", type: "vendor-id" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ApplicationStatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Stopped", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ActiveVisibleFocus", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ActiveHidden", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ActiveVisibleNotFocus", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ApplicationStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "CatalogVendorId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ApplicationId", id: 1, type: "string", conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ApplicationBasic);
