// @matter/model/dist/cjs/standard/elements/application-launcher.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var application_launcher_element_exports = {};
__export(application_launcher_element_exports, {
  ApplicationLauncher: () => ApplicationLauncher
});
module.exports = __toCommonJS(application_launcher_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ApplicationLauncher = (0, import_elements.ClusterElement)(
  { name: "ApplicationLauncher", id: 1292 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "AP", constraint: "0", title: "ApplicationPlatform" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "CatalogList", id: 0, type: "list", access: "R V", conformance: "AP", quality: "N" },
    (0, import_elements.FieldElement)({ name: "entry", type: "uint16" })
  ),
  (0, import_elements.AttributeElement)({
    name: "CurrentApp",
    id: 1,
    type: "ApplicationEPStruct",
    access: "R V",
    conformance: "O",
    constraint: "desc",
    default: null,
    quality: "X"
  }),
  (0, import_elements.CommandElement)(
    {
      name: "LaunchApp",
      id: 0,
      access: "O",
      conformance: "M",
      direction: "request",
      response: "LauncherResponse"
    },
    (0, import_elements.FieldElement)({ name: "Application", id: 0, type: "ApplicationStruct", conformance: "AP", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "Data", id: 1, type: "octstr", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    { name: "StopApp", id: 1, access: "O", conformance: "M", direction: "request", response: "LauncherResponse" },
    (0, import_elements.FieldElement)({ name: "Application", id: 0, type: "ApplicationStruct", conformance: "AP", constraint: "desc" })
  ),
  (0, import_elements.CommandElement)(
    { name: "HideApp", id: 2, access: "O", conformance: "M", direction: "request", response: "LauncherResponse" },
    (0, import_elements.FieldElement)({ name: "Application", id: 0, type: "ApplicationStruct", conformance: "AP", constraint: "desc" })
  ),
  (0, import_elements.CommandElement)(
    { name: "LauncherResponse", id: 3, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "StatusEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Data", id: 1, type: "octstr", conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Success", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AppNotAvailable", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SystemBusy", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "PendingUserApproval", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Downloading", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Installing", id: 5, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ApplicationStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "CatalogVendorId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ApplicationId", id: 1, type: "string", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ApplicationEPStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Application", id: 0, type: "ApplicationStruct", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Endpoint", id: 1, type: "endpoint-no", conformance: "O" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ApplicationLauncher);
