// @matter/model/dist/cjs/standard/elements/ballast-configuration.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ballast_configuration_element_exports = {};
__export(ballast_configuration_element_exports, {
  BallastConfiguration: () => BallastConfiguration
});
module.exports = __toCommonJS(ballast_configuration_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const BallastConfiguration = (0, import_elements.ClusterElement)(
  { name: "BallastConfiguration", id: 769 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 4 }),
  (0, import_elements.AttributeElement)({
    name: "PhysicalMinLevel",
    id: 0,
    type: "uint8",
    access: "R V",
    conformance: "M",
    constraint: "1 to 254",
    default: 1
  }),
  (0, import_elements.AttributeElement)({
    name: "PhysicalMaxLevel",
    id: 1,
    type: "uint8",
    access: "R V",
    conformance: "M",
    constraint: "1 to 254",
    default: 254
  }),
  (0, import_elements.AttributeElement)({ name: "BallastStatus", id: 2, type: "BallastStatusBitmap", access: "R V", conformance: "O", default: 0 }),
  (0, import_elements.AttributeElement)({
    name: "MinLevel",
    id: 16,
    type: "uint8",
    access: "RW VM",
    conformance: "M",
    constraint: "physicalMinLevel to maxLevel",
    default: { type: "reference", name: "PhysicalMinLevel" }
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxLevel",
    id: 17,
    type: "uint8",
    access: "RW VM",
    conformance: "M",
    constraint: "minLevel to physicalMaxLevel",
    default: { type: "reference", name: "PhysicalMaxLevel" }
  }),
  (0, import_elements.AttributeElement)({ name: "PowerOnLevel", id: 18, conformance: "D" }),
  (0, import_elements.AttributeElement)({ name: "PowerOnFadeTime", id: 19, conformance: "D" }),
  (0, import_elements.AttributeElement)({ name: "IntrinsicBallastFactor", id: 20, type: "uint8", access: "RW VM", conformance: "O", quality: "X" }),
  (0, import_elements.AttributeElement)({
    name: "BallastFactorAdjustment",
    id: 21,
    type: "uint8",
    access: "RW VM",
    conformance: "O",
    constraint: "100 to ms",
    default: null,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({ name: "LampQuantity", id: 32, type: "uint8", access: "R V", conformance: "M" }),
  (0, import_elements.AttributeElement)({ name: "LampType", id: 48, type: "string", access: "RW VM", conformance: "O", constraint: "max 16" }),
  (0, import_elements.AttributeElement)(
    { name: "LampManufacturer", id: 49, type: "string", access: "RW VM", conformance: "O", constraint: "max 16" }
  ),
  (0, import_elements.AttributeElement)({
    name: "LampRatedHours",
    id: 50,
    type: "uint24",
    access: "RW VM",
    conformance: "O",
    default: null,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({ name: "LampBurnHours", id: 51, type: "uint24", access: "RW VM", conformance: "O", default: 0, quality: "X" }),
  (0, import_elements.AttributeElement)(
    { name: "LampAlarmMode", id: 52, type: "LampAlarmModeBitmap", access: "RW VM", conformance: "O", default: 0 }
  ),
  (0, import_elements.AttributeElement)({
    name: "LampBurnHoursTripPoint",
    id: 53,
    type: "uint24",
    access: "RW VM",
    conformance: "O",
    default: null,
    quality: "X"
  }),
  (0, import_elements.DatatypeElement)(
    { name: "BallastStatusBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "BallastNonOperational", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "LampFailure", constraint: "1" })
  ),
  (0, import_elements.DatatypeElement)({ name: "LampAlarmModeBitmap", type: "map8" }, (0, import_elements.FieldElement)({ name: "LampBurnHours", constraint: "0" }))
);
import_MatterDefinition.MatterDefinition.children.push(BallastConfiguration);
