// @matter/model/dist/cjs/standard/elements/basic-information.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var basic_information_element_exports = {};
__export(basic_information_element_exports, {
  BasicInformation: () => BasicInformation
});
module.exports = __toCommonJS(basic_information_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const BasicInformation = (0, import_elements.ClusterElement)(
  { name: "BasicInformation", id: 40 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 4 }),
  (0, import_elements.AttributeElement)({
    name: "DataModelRevision",
    id: 0,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "VendorName",
    id: 1,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "max 32",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({ name: "VendorId", id: 2, type: "vendor-id", access: "R V", conformance: "M", quality: "F" }),
  (0, import_elements.AttributeElement)({
    name: "ProductName",
    id: 3,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "max 32",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({ name: "ProductId", id: 4, type: "uint16", access: "R V", conformance: "M", quality: "F" }),
  (0, import_elements.AttributeElement)({
    name: "NodeLabel",
    id: 5,
    type: "string",
    access: "RW VM",
    conformance: "M",
    constraint: "max 32",
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "Location",
    id: 6,
    type: "string",
    access: "RW VA",
    conformance: "M",
    constraint: "2",
    default: "XX",
    quality: "N"
  }),
  (0, import_elements.AttributeElement)(
    { name: "HardwareVersion", id: 7, type: "uint16", access: "R V", conformance: "M", default: 0, quality: "F" }
  ),
  (0, import_elements.AttributeElement)({
    name: "HardwareVersionString",
    id: 8,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "1 to 64",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "SoftwareVersion",
    id: 9,
    type: "uint32",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "SoftwareVersionString",
    id: 10,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "1 to 64",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "ManufacturingDate",
    id: 11,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "8 to 16",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "PartNumber",
    id: 12,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 32",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "ProductUrl",
    id: 13,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 256",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "ProductLabel",
    id: 14,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 64",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "SerialNumber",
    id: 15,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 32",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "LocalConfigDisabled",
    id: 16,
    type: "bool",
    access: "RW VM",
    conformance: "O",
    default: false,
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({ name: "Reachable", id: 17, type: "bool", access: "R V", conformance: "O", default: true }),
  (0, import_elements.AttributeElement)({ name: "UniqueId", id: 18, type: "string", access: "R V", conformance: "M", constraint: "max 32", quality: "F" }),
  (0, import_elements.AttributeElement)({
    name: "CapabilityMinima",
    id: 19,
    type: "CapabilityMinimaStruct",
    access: "R V",
    conformance: "M",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "ProductAppearance",
    id: 20,
    type: "ProductAppearanceStruct",
    access: "R V",
    conformance: "O",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "SpecificationVersion",
    id: 21,
    type: "uint32",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxPathsPerInvoke",
    id: 22,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "min 1",
    default: 1,
    quality: "F"
  }),
  (0, import_elements.EventElement)(
    { name: "StartUp", id: 0, access: "V", conformance: "M", priority: "critical" },
    (0, import_elements.FieldElement)({ name: "SoftwareVersion", id: 0, type: "uint32", conformance: "M" })
  ),
  (0, import_elements.EventElement)({ name: "ShutDown", id: 1, access: "V", conformance: "O", priority: "critical" }),
  (0, import_elements.EventElement)(
    { name: "Leave", id: 2, access: "V", conformance: "O", priority: "info" },
    (0, import_elements.FieldElement)({ name: "FabricIndex", id: 0, type: "fabric-idx", conformance: "M", constraint: "1 to 254" })
  ),
  (0, import_elements.EventElement)(
    { name: "ReachableChanged", id: 3, access: "V", conformance: "desc", priority: "info" },
    (0, import_elements.FieldElement)({ name: "ReachableNewValue", id: 0, type: "bool", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ProductFinishEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Other", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Matte", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Satin", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Polished", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rugged", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Fabric", id: 5, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ColorEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Black", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Navy", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Green", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Teal", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Maroon", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Purple", id: 5, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Olive", id: 6, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Gray", id: 7, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Blue", id: 8, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Lime", id: 9, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Aqua", id: 10, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Red", id: 11, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Fuchsia", id: 12, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Yellow", id: 13, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "White", id: 14, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Nickel", id: 15, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Chrome", id: 16, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Brass", id: 17, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Copper", id: 18, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Silver", id: 19, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Gold", id: 20, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ProductAppearanceStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Finish", id: 0, type: "ProductFinishEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "PrimaryColor", id: 1, type: "ColorEnum", conformance: "M", quality: "X" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "CapabilityMinimaStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "CaseSessionsPerFabric", id: 0, type: "uint16", conformance: "M", constraint: "min 3", default: 3 }),
    (0, import_elements.FieldElement)({ name: "SubscriptionsPerFabric", id: 1, type: "uint16", conformance: "M", constraint: "min 3", default: 3 })
  )
);
import_MatterDefinition.MatterDefinition.children.push(BasicInformation);
