// @matter/model/dist/cjs/standard/elements/binding.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var binding_element_exports = {};
__export(binding_element_exports, {
  Binding: () => Binding
});
module.exports = __toCommonJS(binding_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const Binding = (0, import_elements.ClusterElement)(
  { name: "Binding", id: 30 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    {
      name: "Binding",
      id: 0,
      type: "list",
      access: "RW F VM",
      conformance: "M",
      constraint: "desc",
      default: [],
      quality: "N"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "TargetStruct" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "TargetStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Node", id: 1, type: "node-id", access: "F", conformance: "Endpoint" }),
    (0, import_elements.FieldElement)({ name: "Group", id: 2, type: "group-id", access: "F", conformance: "!Endpoint", constraint: "min 1" }),
    (0, import_elements.FieldElement)({ name: "Endpoint", id: 3, type: "endpoint-no", access: "F", conformance: "!Group" }),
    (0, import_elements.FieldElement)({ name: "Cluster", id: 4, type: "cluster-id", access: "F", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(Binding);
