// @matter/model/dist/cjs/standard/elements/boolean-state-configuration.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var boolean_state_configuration_element_exports = {};
__export(boolean_state_configuration_element_exports, {
  BooleanStateConfiguration: () => BooleanStateConfiguration
});
module.exports = __toCommonJS(boolean_state_configuration_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const BooleanStateConfiguration = (0, import_elements.ClusterElement)(
  { name: "BooleanStateConfiguration", id: 128 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "VIS", conformance: "O", constraint: "0", title: "Visual" }),
    (0, import_elements.FieldElement)({ name: "AUD", conformance: "O", constraint: "1", title: "Audible" }),
    (0, import_elements.FieldElement)({ name: "SPRS", conformance: "[VIS | AUD]", constraint: "2", title: "AlarmSuppress" }),
    (0, import_elements.FieldElement)({ name: "SENSLVL", conformance: "O", constraint: "3", title: "SensitivityLevel" })
  ),
  (0, import_elements.AttributeElement)({
    name: "CurrentSensitivityLevel",
    id: 0,
    type: "uint8",
    access: "RW VO",
    conformance: "SENSLVL",
    constraint: "max supportedSensitivityLevels - 1",
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "SupportedSensitivityLevels",
    id: 1,
    type: "uint8",
    access: "R V",
    conformance: "SENSLVL",
    constraint: "2 to 10",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "DefaultSensitivityLevel",
    id: 2,
    type: "uint8",
    access: "R V",
    conformance: "[SENSLVL]",
    constraint: "max supportedSensitivityLevels - 1",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)(
    { name: "AlarmsActive", id: 3, type: "AlarmModeBitmap", access: "R V", conformance: "VIS | AUD", default: 0 }
  ),
  (0, import_elements.AttributeElement)(
    { name: "AlarmsSuppressed", id: 4, type: "AlarmModeBitmap", access: "R V", conformance: "SPRS", default: 0 }
  ),
  (0, import_elements.AttributeElement)({
    name: "AlarmsEnabled",
    id: 5,
    type: "AlarmModeBitmap",
    access: "R V",
    conformance: "[VIS | AUD]",
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "AlarmsSupported",
    id: 6,
    type: "AlarmModeBitmap",
    access: "R V",
    conformance: "VIS | AUD",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({ name: "SensorFault", id: 7, type: "SensorFaultBitmap", access: "R V", conformance: "O", default: 0 }),
  (0, import_elements.EventElement)(
    { name: "AlarmsStateChanged", id: 0, access: "V", conformance: "VIS | AUD", priority: "info" },
    (0, import_elements.FieldElement)({ name: "AlarmsActive", id: 0, type: "AlarmModeBitmap", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AlarmsSuppressed", id: 1, type: "AlarmModeBitmap", conformance: "SPRS" })
  ),
  (0, import_elements.EventElement)(
    { name: "SensorFault", id: 1, access: "V", conformance: "O", priority: "info" },
    (0, import_elements.FieldElement)({ name: "SensorFault", id: 0, type: "SensorFaultBitmap", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "SuppressAlarm", id: 0, access: "O", conformance: "SPRS", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "AlarmsToSuppress", id: 0, type: "AlarmModeBitmap", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "EnableDisableAlarm",
      id: 1,
      access: "O",
      conformance: "VIS | AUD",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "AlarmsToEnableDisable", id: 0, type: "AlarmModeBitmap", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "AlarmModeBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "Visual", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "Audible", constraint: "1" })
  ),
  (0, import_elements.DatatypeElement)({ name: "SensorFaultBitmap", type: "map16" }, (0, import_elements.FieldElement)({ name: "GeneralFault", constraint: "0" }))
);
import_MatterDefinition.MatterDefinition.children.push(BooleanStateConfiguration);
