// @matter/model/dist/cjs/standard/elements/channel.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var channel_element_exports = {};
__export(channel_element_exports, {
  Channel: () => Channel
});
module.exports = __toCommonJS(channel_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const Channel = (0, import_elements.ClusterElement)(
  { name: "Channel", id: 1284 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "CL", constraint: "0", title: "ChannelList" }),
    (0, import_elements.FieldElement)({ name: "LI", constraint: "1", title: "LineupInfo" }),
    (0, import_elements.FieldElement)({ name: "EG", constraint: "2", title: "ElectronicGuide" }),
    (0, import_elements.FieldElement)({ name: "RP", constraint: "3", title: "RecordProgram" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "ChannelList", id: 0, type: "list", access: "R V", conformance: "CL", default: [] },
    (0, import_elements.FieldElement)({ name: "entry", type: "ChannelInfoStruct" })
  ),
  (0, import_elements.AttributeElement)({
    name: "Lineup",
    id: 1,
    type: "LineupInfoStruct",
    access: "R V",
    conformance: "LI",
    constraint: "desc",
    default: null,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentChannel",
    id: 2,
    type: "ChannelInfoStruct",
    access: "R V",
    conformance: "O",
    constraint: "desc",
    default: null,
    quality: "X"
  }),
  (0, import_elements.CommandElement)(
    {
      name: "ChangeChannel",
      id: 0,
      access: "O",
      conformance: "CL | LI",
      direction: "request",
      response: "ChangeChannelResponse"
    },
    (0, import_elements.FieldElement)({ name: "Match", id: 0, type: "string", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "ChangeChannelResponse", id: 1, conformance: "CL | LI", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "StatusEnum", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "Data", id: 1, type: "string", conformance: "O", constraint: "any" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "ChangeChannelByNumber",
      id: 2,
      access: "O",
      conformance: "M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "MajorNumber", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "MinorNumber", id: 1, type: "uint16", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "SkipChannel", id: 3, access: "O", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "Count", id: 0, type: "int16", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "GetProgramGuide",
      id: 4,
      access: "O",
      conformance: "EG",
      direction: "request",
      response: "ProgramGuideResponse"
    },
    (0, import_elements.FieldElement)({ name: "StartTime", id: 0, type: "epoch-s", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndTime", id: 1, type: "epoch-s", conformance: "M" }),
    (0, import_elements.FieldElement)(
      { name: "ChannelList", id: 2, type: "list", conformance: "O", constraint: "max 255", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "ChannelInfoStruct" })
    ),
    (0, import_elements.FieldElement)({ name: "PageToken", id: 3, type: "PageTokenStruct", conformance: "O", default: null, quality: "X" }),
    (0, import_elements.FieldElement)({ name: "RecordingFlag", id: 5, type: "RecordingFlagBitmap", conformance: "O", default: null, quality: "X" }),
    (0, import_elements.FieldElement)(
      { name: "ExternalIdList", id: 6, type: "list", conformance: "O", constraint: "max 255", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "ContentLauncher.AdditionalInfoStruct" })
    ),
    (0, import_elements.FieldElement)({ name: "Data", id: 7, type: "octstr", conformance: "O", constraint: "max 8092" })
  ),
  (0, import_elements.CommandElement)(
    { name: "ProgramGuideResponse", id: 5, conformance: "EG", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Paging", id: 0, type: "ChannelPagingStruct", conformance: "M" }),
    (0, import_elements.FieldElement)(
      { name: "ProgramList", id: 1, type: "list", conformance: "M", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "ProgramStruct" })
    )
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RecordProgram",
      id: 6,
      access: "O",
      conformance: "RP & EG",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ProgramIdentifier", id: 0, type: "string", conformance: "M", constraint: "max 255" }),
    (0, import_elements.FieldElement)({ name: "ShouldRecordSeries", id: 1, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)(
      { name: "ExternalIdList", id: 2, type: "list", conformance: "O", constraint: "max 255", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "ContentLauncher.AdditionalInfoStruct" })
    ),
    (0, import_elements.FieldElement)({ name: "Data", id: 3, type: "octstr", conformance: "O", constraint: "max 8092" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "CancelRecordProgram",
      id: 7,
      access: "O",
      conformance: "RP & EG",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ProgramIdentifier", id: 0, type: "string", conformance: "M", constraint: "max 255" }),
    (0, import_elements.FieldElement)({ name: "ShouldRecordSeries", id: 1, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)(
      { name: "ExternalIdList", id: 2, type: "list", conformance: "O", constraint: "max 255", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "ContentLauncher.AdditionalInfoStruct" })
    ),
    (0, import_elements.FieldElement)({ name: "Data", id: 3, type: "octstr", conformance: "O", constraint: "max 8092" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "RecordingFlagBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "Scheduled", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "RecordSeries", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "Recorded", constraint: "2" })
  ),
  (0, import_elements.DatatypeElement)({ name: "LineupInfoTypeEnum", type: "enum8" }, (0, import_elements.FieldElement)({ name: "Mso", id: 0, conformance: "M" })),
  (0, import_elements.DatatypeElement)(
    { name: "StatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Success", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "MultipleMatches", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NoMatches", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ChannelTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Satellite", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Cable", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Terrestrial", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Ott", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ChannelInfoStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "MajorNumber", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "MinorNumber", id: 1, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Name", id: 2, type: "string", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "CallSign", id: 3, type: "string", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "AffiliateCallSign", id: 4, type: "string", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Identifier", id: 5, type: "string", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Type", id: 6, type: "ChannelTypeEnum", conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "LineupInfoStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "OperatorName", id: 0, type: "string", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "LineupName", id: 1, type: "string", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "PostalCode", id: 2, type: "string", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "LineupInfoType", id: 3, type: "LineupInfoTypeEnum", conformance: "M", constraint: "desc" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ProgramStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Identifier", id: 0, type: "string", conformance: "M", constraint: "max 255" }),
    (0, import_elements.FieldElement)({ name: "Channel", id: 1, type: "ChannelInfoStruct", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StartTime", id: 2, type: "epoch-s", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndTime", id: 3, type: "epoch-s", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Title", id: 4, type: "string", conformance: "M", constraint: "max 255" }),
    (0, import_elements.FieldElement)({ name: "Subtitle", id: 5, type: "string", conformance: "O", constraint: "max 255" }),
    (0, import_elements.FieldElement)({ name: "Description", id: 6, type: "string", conformance: "O", constraint: "max 8192" }),
    (0, import_elements.FieldElement)(
      {
        name: "AudioLanguages",
        id: 7,
        type: "list",
        conformance: "O",
        constraint: "max 10[max 50]",
        default: []
      },
      (0, import_elements.FieldElement)({ name: "entry", type: "string" })
    ),
    (0, import_elements.FieldElement)(
      { name: "Ratings", id: 8, type: "list", conformance: "O", constraint: "max 255", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "string" })
    ),
    (0, import_elements.FieldElement)({ name: "ThumbnailUrl", id: 9, type: "string", conformance: "O", constraint: "max 8192" }),
    (0, import_elements.FieldElement)({ name: "PosterArtUrl", id: 10, type: "string", conformance: "O", constraint: "max 8192" }),
    (0, import_elements.FieldElement)({ name: "DvbiUrl", id: 11, type: "string", conformance: "O", constraint: "max 8192" }),
    (0, import_elements.FieldElement)({ name: "ReleaseDate", id: 12, type: "string", conformance: "O", constraint: "max 30" }),
    (0, import_elements.FieldElement)({ name: "ParentalGuidanceText", id: 13, type: "string", conformance: "O", constraint: "max 255" }),
    (0, import_elements.FieldElement)({ name: "RecordingFlag", id: 14, type: "RecordingFlagBitmap", conformance: "RP" }),
    (0, import_elements.FieldElement)({ name: "SeriesInfo", id: 15, type: "SeriesInfoStruct", conformance: "O", default: null, quality: "X" }),
    (0, import_elements.FieldElement)(
      { name: "CategoryList", id: 16, type: "list", conformance: "O", constraint: "max 255", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "ProgramCategoryStruct" })
    ),
    (0, import_elements.FieldElement)(
      { name: "CastList", id: 17, type: "list", conformance: "O", constraint: "max 255", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "ProgramCastStruct" })
    ),
    (0, import_elements.FieldElement)(
      { name: "ExternalIdList", id: 18, type: "list", conformance: "O", constraint: "max 255", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "ContentLauncher.AdditionalInfoStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ProgramCategoryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Category", id: 0, type: "string", conformance: "M", constraint: "max 256" }),
    (0, import_elements.FieldElement)({ name: "SubCategory", id: 1, type: "string", conformance: "O", constraint: "max 256" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "SeriesInfoStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Season", id: 0, type: "string", conformance: "M", constraint: "max 256" }),
    (0, import_elements.FieldElement)({ name: "Episode", id: 1, type: "string", conformance: "M", constraint: "max 256" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ProgramCastStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Name", id: 0, type: "string", conformance: "M", constraint: "max 256" }),
    (0, import_elements.FieldElement)({ name: "Role", id: 1, type: "string", conformance: "M", constraint: "max 256" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "PageTokenStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Limit", id: 0, type: "uint16", conformance: "O", default: 0 }),
    (0, import_elements.FieldElement)({ name: "After", id: 1, type: "string", conformance: "O", constraint: "max 8192" }),
    (0, import_elements.FieldElement)({ name: "Before", id: 2, type: "string", conformance: "O", constraint: "max 8192" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ChannelPagingStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "PreviousToken", id: 0, type: "PageTokenStruct", conformance: "O", default: null, quality: "X" }),
    (0, import_elements.FieldElement)({ name: "NextToken", id: 1, type: "PageTokenStruct", conformance: "O", default: null, quality: "X" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(Channel);
