// @matter/model/dist/cjs/standard/elements/color-control.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var color_control_element_exports = {};
__export(color_control_element_exports, {
  ColorControl: () => ColorControl
});
module.exports = __toCommonJS(color_control_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ColorControl = (0, import_elements.ClusterElement)(
  { name: "ColorControl", id: 768 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 7 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "HS", conformance: "EHUE, O", constraint: "0", title: "HueSaturation" }),
    (0, import_elements.FieldElement)({ name: "EHUE", conformance: "CL, O", constraint: "1", title: "EnhancedHue" }),
    (0, import_elements.FieldElement)({ name: "CL", conformance: "O", constraint: "2", title: "ColorLoop" }),
    (0, import_elements.FieldElement)({ name: "XY", conformance: "O", constraint: "3", title: "Xy" }),
    (0, import_elements.FieldElement)({ name: "CT", conformance: "O", constraint: "4", title: "ColorTemperature" })
  ),
  (0, import_elements.AttributeElement)({
    name: "CurrentHue",
    id: 0,
    type: "uint8",
    access: "R V",
    conformance: "HS",
    constraint: "max 254",
    default: 0,
    quality: "N P Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentSaturation",
    id: 1,
    type: "uint8",
    access: "R V",
    conformance: "HS",
    constraint: "max 254",
    default: 0,
    quality: "N S P Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "RemainingTime",
    id: 2,
    type: "uint16",
    access: "R V",
    conformance: "O",
    constraint: "0 to 65535",
    default: 0,
    quality: "Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentX",
    id: 3,
    type: "uint16",
    access: "R V",
    conformance: "XY",
    constraint: "max 65279",
    default: 24939,
    quality: "N S P Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentY",
    id: 4,
    type: "uint16",
    access: "R V",
    conformance: "XY",
    constraint: "max 65279",
    default: 24701,
    quality: "N S P Q"
  }),
  (0, import_elements.AttributeElement)({ name: "DriftCompensation", id: 5, type: "DriftCompensationEnum", access: "R V", conformance: "O" }),
  (0, import_elements.AttributeElement)(
    { name: "CompensationText", id: 6, type: "string", access: "R V", conformance: "O", constraint: "max 254" }
  ),
  (0, import_elements.AttributeElement)({
    name: "ColorTemperatureMireds",
    id: 7,
    type: "uint16",
    access: "R V",
    conformance: "CT",
    constraint: "colorTempPhysicalMinMireds to colorTempPhysicalMaxMireds",
    default: 250,
    quality: "N S P Q"
  }),
  (0, import_elements.AttributeElement)({ name: "ColorMode", id: 8, type: "ColorModeEnum", access: "R V", conformance: "M", quality: "N" }),
  (0, import_elements.AttributeElement)({
    name: "Options",
    id: 15,
    type: "OptionsBitmap",
    access: "RW VO",
    conformance: "M",
    constraint: "desc",
    default: 0
  }),
  (0, import_elements.AttributeElement)({
    name: "NumberOfPrimaries",
    id: 16,
    type: "uint8",
    access: "R V",
    conformance: "M",
    constraint: "max 6",
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary1X",
    id: 17,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 0",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary1Y",
    id: 18,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 0",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary1Intensity",
    id: 19,
    type: "uint8",
    access: "R V",
    conformance: "NumberOfPrimaries > 0",
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary2X",
    id: 21,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 1",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary2Y",
    id: 22,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 1",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary2Intensity",
    id: 23,
    type: "uint8",
    access: "R V",
    conformance: "NumberOfPrimaries > 1",
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary3X",
    id: 25,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 2",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary3Y",
    id: 26,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 2",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary3Intensity",
    id: 27,
    type: "uint8",
    access: "R V",
    conformance: "NumberOfPrimaries > 2",
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary4X",
    id: 32,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 3",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary4Y",
    id: 33,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 3",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary4Intensity",
    id: 34,
    type: "uint8",
    access: "R V",
    conformance: "NumberOfPrimaries > 3",
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary5X",
    id: 36,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 4",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary5Y",
    id: 37,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 4",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary5Intensity",
    id: 38,
    type: "uint8",
    access: "R V",
    conformance: "NumberOfPrimaries > 4",
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary6X",
    id: 40,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 5",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary6Y",
    id: 41,
    type: "uint16",
    access: "R V",
    conformance: "NumberOfPrimaries > 5",
    constraint: "max 65279",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Primary6Intensity",
    id: 42,
    type: "uint8",
    access: "R V",
    conformance: "NumberOfPrimaries > 5",
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)(
    { name: "WhitePointX", id: 48, type: "uint16", access: "RW VM", conformance: "O", constraint: "max 65279" }
  ),
  (0, import_elements.AttributeElement)(
    { name: "WhitePointY", id: 49, type: "uint16", access: "RW VM", conformance: "O", constraint: "max 65279" }
  ),
  (0, import_elements.AttributeElement)(
    { name: "ColorPointRx", id: 50, type: "uint16", access: "RW VM", conformance: "O", constraint: "max 65279" }
  ),
  (0, import_elements.AttributeElement)(
    { name: "ColorPointRy", id: 51, type: "uint16", access: "RW VM", conformance: "O", constraint: "max 65279" }
  ),
  (0, import_elements.AttributeElement)({ name: "ColorPointRIntensity", id: 52, type: "uint8", access: "RW VM", conformance: "O", quality: "X" }),
  (0, import_elements.AttributeElement)(
    { name: "ColorPointGx", id: 54, type: "uint16", access: "RW VM", conformance: "O", constraint: "max 65279" }
  ),
  (0, import_elements.AttributeElement)(
    { name: "ColorPointGy", id: 55, type: "uint16", access: "RW VM", conformance: "O", constraint: "max 65279" }
  ),
  (0, import_elements.AttributeElement)({ name: "ColorPointGIntensity", id: 56, type: "uint8", access: "RW VM", conformance: "O", quality: "X" }),
  (0, import_elements.AttributeElement)(
    { name: "ColorPointBx", id: 58, type: "uint16", access: "RW VM", conformance: "O", constraint: "max 65279" }
  ),
  (0, import_elements.AttributeElement)(
    { name: "ColorPointBy", id: 59, type: "uint16", access: "RW VM", conformance: "O", constraint: "max 65279" }
  ),
  (0, import_elements.AttributeElement)({ name: "ColorPointBIntensity", id: 60, type: "uint8", access: "RW VM", conformance: "O", quality: "X" }),
  (0, import_elements.AttributeElement)({
    name: "EnhancedCurrentHue",
    id: 16384,
    type: "uint16",
    access: "R V",
    conformance: "EHUE",
    default: 0,
    quality: "N S Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "EnhancedColorMode",
    id: 16385,
    type: "EnhancedColorModeEnum",
    access: "R V",
    conformance: "M",
    default: 1,
    quality: "N S"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "ColorLoopActive",
      id: 16386,
      type: "enum16",
      access: "R V",
      conformance: "CL",
      constraint: "max 1",
      default: 0,
      quality: "N S"
    },
    (0, import_elements.FieldElement)({ name: "Inactive", id: 0 }),
    (0, import_elements.FieldElement)({ name: "Active", id: 1 })
  ),
  (0, import_elements.AttributeElement)({
    name: "ColorLoopDirection",
    id: 16387,
    type: "ColorLoopDirectionEnum",
    access: "R V",
    conformance: "CL",
    constraint: "max 1",
    default: 0,
    quality: "N S"
  }),
  (0, import_elements.AttributeElement)({
    name: "ColorLoopTime",
    id: 16388,
    type: "uint16",
    access: "R V",
    conformance: "CL",
    default: 25,
    quality: "N S"
  }),
  (0, import_elements.AttributeElement)({ name: "ColorLoopStartEnhancedHue", id: 16389, type: "uint16", access: "R V", conformance: "CL", default: 8960 }),
  (0, import_elements.AttributeElement)({ name: "ColorLoopStoredEnhancedHue", id: 16390, type: "uint16", access: "R V", conformance: "CL", default: 0 }),
  (0, import_elements.AttributeElement)(
    {
      name: "ColorCapabilities",
      id: 16394,
      type: "map16",
      access: "R V",
      conformance: "M",
      constraint: "max 31",
      default: 0
    },
    (0, import_elements.FieldElement)({ name: "HueSaturation", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "EnhancedHue", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "ColorLoop", constraint: "2" }),
    (0, import_elements.FieldElement)({ name: "XY", constraint: "3" }),
    (0, import_elements.FieldElement)({ name: "ColorTemperature", constraint: "4" })
  ),
  (0, import_elements.AttributeElement)({
    name: "ColorTempPhysicalMinMireds",
    id: 16395,
    type: "uint16",
    access: "R V",
    conformance: "CT",
    constraint: "1 to 65279"
  }),
  (0, import_elements.AttributeElement)({
    name: "ColorTempPhysicalMaxMireds",
    id: 16396,
    type: "uint16",
    access: "R V",
    conformance: "CT",
    constraint: "max 65279"
  }),
  (0, import_elements.AttributeElement)({
    name: "CoupleColorTempToLevelMinMireds",
    id: 16397,
    type: "uint16",
    access: "R V",
    conformance: "CT & ColorTemperatureMireds",
    constraint: "colorTempPhysicalMinMireds to colorTemperatureMireds"
  }),
  (0, import_elements.AttributeElement)({
    name: "StartUpColorTemperatureMireds",
    id: 16400,
    type: "uint16",
    access: "RW VM",
    conformance: "CT & ColorTemperatureMireds",
    constraint: "1 to 65279",
    quality: "X N"
  }),
  (0, import_elements.CommandElement)(
    { name: "MoveToHue", id: 0, access: "O", conformance: "HS", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "Hue", id: 0, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "Direction", id: 1, type: "DirectionEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "MoveHue", id: 1, access: "O", conformance: "HS", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "MoveMode", id: 0, type: "MoveModeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate", id: 1, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 2, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "StepHue", id: 2, access: "O", conformance: "HS", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "StepMode", id: 0, type: "StepModeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StepSize", id: 1, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "MoveToSaturation", id: 3, access: "O", conformance: "HS", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "Saturation", id: 0, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 1, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 2, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "MoveSaturation", id: 4, access: "O", conformance: "HS", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "MoveMode", id: 0, type: "MoveModeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate", id: 1, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 2, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "StepSaturation", id: 5, access: "O", conformance: "HS", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "StepMode", id: 0, type: "StepModeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StepSize", id: 1, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "MoveToHueAndSaturation",
      id: 6,
      access: "O",
      conformance: "HS",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "Hue", id: 0, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "Saturation", id: 1, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "MoveToColor", id: 7, access: "O", conformance: "XY", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "ColorX", id: 0, type: "uint16", conformance: "M", constraint: "max 65279" }),
    (0, import_elements.FieldElement)({ name: "ColorY", id: 1, type: "uint16", conformance: "M", constraint: "max 65279" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "MoveColor", id: 8, access: "O", conformance: "XY", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "RateX", id: 0, type: "int16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "RateY", id: 1, type: "int16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 2, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "StepColor", id: 9, access: "O", conformance: "XY", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "StepX", id: 0, type: "int16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StepY", id: 1, type: "int16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "MoveToColorTemperature",
      id: 10,
      access: "O",
      conformance: "CT",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ColorTemperatureMireds", id: 0, type: "uint16", conformance: "M", constraint: "max 65279" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 1, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 2, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "EnhancedMoveToHue",
      id: 64,
      access: "O",
      conformance: "EHUE",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "EnhancedHue", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Direction", id: 1, type: "DirectionEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "EnhancedMoveHue",
      id: 65,
      access: "O",
      conformance: "EHUE",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "MoveMode", id: 0, type: "MoveModeEnum", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "Rate", id: 1, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 2, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "EnhancedStepHue",
      id: 66,
      access: "O",
      conformance: "EHUE",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "StepMode", id: 0, type: "StepModeEnum", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "StepSize", id: 1, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "EnhancedMoveToHueAndSaturation",
      id: 67,
      access: "O",
      conformance: "EHUE",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "EnhancedHue", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Saturation", id: 1, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 3, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    { name: "ColorLoopSet", id: 68, access: "O", conformance: "CL", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "UpdateFlags", id: 0, type: "UpdateFlagsBitmap", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Action", id: 1, type: "ColorLoopActionEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Direction", id: 2, type: "ColorLoopDirectionEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Time", id: 3, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StartHue", id: 4, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 5, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 6, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "StopMoveStep",
      id: 71,
      access: "O",
      conformance: "HS | XY | CT",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 0, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 1, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "MoveColorTemperature",
      id: 75,
      access: "O",
      conformance: "CT",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "MoveMode", id: 0, type: "MoveModeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate", id: 1, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ColorTemperatureMinimumMireds", id: 2, type: "uint16", conformance: "M", constraint: "max 65279" }),
    (0, import_elements.FieldElement)({ name: "ColorTemperatureMaximumMireds", id: 3, type: "uint16", conformance: "M", constraint: "max 65279" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 4, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 5, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "StepColorTemperature",
      id: 76,
      access: "O",
      conformance: "CT",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "StepMode", id: 0, type: "StepModeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StepSize", id: 1, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint16", conformance: "M", constraint: "max 65534" }),
    (0, import_elements.FieldElement)({ name: "ColorTemperatureMinimumMireds", id: 3, type: "uint16", conformance: "M", constraint: "max 65279" }),
    (0, import_elements.FieldElement)({ name: "ColorTemperatureMaximumMireds", id: 4, type: "uint16", conformance: "M", constraint: "max 65279" }),
    (0, import_elements.FieldElement)({ name: "OptionsMask", id: 5, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "OptionsOverride", id: 6, type: "OptionsBitmap", conformance: "M", constraint: "desc", default: 0 })
  ),
  (0, import_elements.DatatypeElement)({ name: "OptionsBitmap", type: "map8" }, (0, import_elements.FieldElement)({ name: "ExecuteIfOff", constraint: "0" })),
  (0, import_elements.DatatypeElement)(
    { name: "UpdateFlagsBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "UpdateAction", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "UpdateDirection", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "UpdateTime", constraint: "2" }),
    (0, import_elements.FieldElement)({ name: "UpdateStartHue", constraint: "3" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DriftCompensationEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "None", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OtherOrUnknown", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TemperatureMonitoring", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OpticalLuminanceMonitoringAndFeedback", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "OpticalColorMonitoringAndFeedback", id: 4, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ColorModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "CurrentHueAndCurrentSaturation", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "CurrentXAndCurrentY", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ColorTemperatureMireds", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "EnhancedColorModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "CurrentHueAndCurrentSaturation", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "CurrentXAndCurrentY", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ColorTemperatureMireds", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EnhancedCurrentHueAndCurrentSaturation", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DirectionEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Shortest", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Longest", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Up", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Down", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "MoveModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Stop", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Up", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Down", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StepModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Up", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Down", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ColorLoopActionEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Deactivate", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ActivateFromColorLoopStartEnhancedHue", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ActivateFromEnhancedCurrentHue", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ColorLoopDirectionEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Decrement", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Increment", id: 1, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ColorControl);
