// @matter/model/dist/cjs/standard/elements/concentration-measurement.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var concentration_measurement_element_exports = {};
__export(concentration_measurement_element_exports, {
  ConcentrationMeasurement: () => ConcentrationMeasurement
});
module.exports = __toCommonJS(concentration_measurement_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ConcentrationMeasurement = (0, import_elements.ClusterElement)(
  { name: "ConcentrationMeasurement" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "MEA", conformance: "O.a+", constraint: "0", title: "NumericMeasurement" }),
    (0, import_elements.FieldElement)({ name: "LEV", conformance: "O.a+", constraint: "1", title: "LevelIndication" }),
    (0, import_elements.FieldElement)({ name: "MED", conformance: "[LEV]", constraint: "2", title: "MediumLevel" }),
    (0, import_elements.FieldElement)({ name: "CRI", conformance: "[LEV]", constraint: "3", title: "CriticalLevel" }),
    (0, import_elements.FieldElement)({ name: "PEA", conformance: "[MEA]", constraint: "4", title: "PeakMeasurement" }),
    (0, import_elements.FieldElement)({ name: "AVG", conformance: "[MEA]", constraint: "5", title: "AverageMeasurement" })
  ),
  (0, import_elements.AttributeElement)({
    name: "MeasuredValue",
    id: 0,
    type: "single",
    access: "R V",
    conformance: "MEA",
    constraint: "minMeasuredValue to maxMeasuredValue",
    default: null,
    quality: "X P"
  }),
  (0, import_elements.AttributeElement)({
    name: "MinMeasuredValue",
    id: 1,
    type: "single",
    access: "R V",
    conformance: "MEA",
    default: null,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxMeasuredValue",
    id: 2,
    type: "single",
    access: "R V",
    conformance: "MEA",
    constraint: "min minMeasuredValue",
    default: null,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "PeakMeasuredValue",
    id: 3,
    type: "single",
    access: "R V",
    conformance: "PEA",
    constraint: "minMeasuredValue to maxMeasuredValue",
    default: null,
    quality: "X P"
  }),
  (0, import_elements.AttributeElement)({
    name: "PeakMeasuredValueWindow",
    id: 4,
    type: "elapsed-s",
    access: "R V",
    conformance: "PEA",
    constraint: "max 604800",
    default: 1,
    quality: "P"
  }),
  (0, import_elements.AttributeElement)({
    name: "AverageMeasuredValue",
    id: 5,
    type: "single",
    access: "R V",
    conformance: "AVG",
    constraint: "minMeasuredValue to maxMeasuredValue",
    default: null,
    quality: "X P"
  }),
  (0, import_elements.AttributeElement)({
    name: "AverageMeasuredValueWindow",
    id: 6,
    type: "elapsed-s",
    access: "R V",
    conformance: "AVG",
    constraint: "max 604800",
    default: 1,
    quality: "P"
  }),
  (0, import_elements.AttributeElement)({ name: "Uncertainty", id: 7, type: "single", access: "R V", conformance: "[MEA]", constraint: "ms" }),
  (0, import_elements.AttributeElement)({ name: "MeasurementUnit", id: 8, type: "MeasurementUnitEnum", access: "R V", conformance: "MEA", quality: "F" }),
  (0, import_elements.AttributeElement)({
    name: "MeasurementMedium",
    id: 9,
    type: "MeasurementMediumEnum",
    access: "R V",
    conformance: "M",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({ name: "LevelValue", id: 10, type: "LevelValueEnum", access: "R V", conformance: "LEV", default: 0 }),
  (0, import_elements.DatatypeElement)(
    { name: "MeasurementUnitEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Ppm", id: 0, conformance: "MEA" }),
    (0, import_elements.FieldElement)({ name: "Ppb", id: 1, conformance: "MEA" }),
    (0, import_elements.FieldElement)({ name: "Ppt", id: 2, conformance: "MEA" }),
    (0, import_elements.FieldElement)({ name: "Mgm3", id: 3, conformance: "MEA" }),
    (0, import_elements.FieldElement)({ name: "Ugm3", id: 4, conformance: "MEA" }),
    (0, import_elements.FieldElement)({ name: "Ngm3", id: 5, conformance: "MEA" }),
    (0, import_elements.FieldElement)({ name: "Pm3", id: 6, conformance: "MEA" }),
    (0, import_elements.FieldElement)({ name: "Bqm3", id: 7, conformance: "MEA" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "MeasurementMediumEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Air", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Water", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Soil", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "LevelValueEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unknown", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Low", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Medium", id: 2, conformance: "MED" }),
    (0, import_elements.FieldElement)({ name: "High", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Critical", id: 4, conformance: "CRI" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ConcentrationMeasurement);
