// @matter/model/dist/cjs/standard/elements/content-launcher.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var content_launcher_element_exports = {};
__export(content_launcher_element_exports, {
  ContentLauncher: () => ContentLauncher
});
module.exports = __toCommonJS(content_launcher_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ContentLauncher = (0, import_elements.ClusterElement)(
  { name: "ContentLauncher", id: 1290 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "CS", constraint: "0", title: "ContentSearch" }),
    (0, import_elements.FieldElement)({ name: "UP", constraint: "1", title: "UrlPlayback" }),
    (0, import_elements.FieldElement)({ name: "AS", constraint: "2", title: "AdvancedSeek" }),
    (0, import_elements.FieldElement)({ name: "TT", constraint: "3", title: "TextTracks" }),
    (0, import_elements.FieldElement)({ name: "AT", constraint: "4", title: "AudioTracks" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "AcceptHeader",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "UP",
      constraint: "max 100[max 1024]",
      default: [],
      quality: "N"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "string" })
  ),
  (0, import_elements.AttributeElement)({
    name: "SupportedStreamingProtocols",
    id: 1,
    type: "SupportedProtocolsBitmap",
    access: "R V",
    conformance: "UP",
    default: 0,
    quality: "N"
  }),
  (0, import_elements.CommandElement)(
    {
      name: "LaunchContent",
      id: 0,
      access: "O",
      conformance: "CS",
      direction: "request",
      response: "LauncherResponse"
    },
    (0, import_elements.FieldElement)({ name: "Search", id: 0, type: "ContentSearchStruct", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "AutoPlay", id: 1, type: "bool", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "Data", id: 2, type: "string", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "PlaybackPreferences", id: 3, type: "PlaybackPreferencesStruct", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "UseCurrentContext", id: 4, type: "bool", conformance: "O", constraint: "desc", default: true })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "LaunchUrl",
      id: 1,
      access: "O",
      conformance: "UP",
      direction: "request",
      response: "LauncherResponse"
    },
    (0, import_elements.FieldElement)({ name: "ContentUrl", id: 0, type: "string", conformance: "M", constraint: "any" }),
    (0, import_elements.FieldElement)({ name: "DisplayString", id: 1, type: "string", conformance: "O", constraint: "any" }),
    (0, import_elements.FieldElement)({
      name: "BrandingInformation",
      id: 2,
      type: "BrandingInformationStruct",
      conformance: "O",
      constraint: "any"
    }),
    (0, import_elements.FieldElement)({
      name: "PlaybackPreferences",
      id: 3,
      type: "PlaybackPreferencesStruct",
      conformance: "O",
      constraint: "any"
    })
  ),
  (0, import_elements.CommandElement)(
    { name: "LauncherResponse", id: 2, conformance: "CS | UP", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "StatusEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Data", id: 1, type: "string", conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "SupportedProtocolsBitmap", type: "map32" },
    (0, import_elements.FieldElement)({ name: "Dash", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "Hls", constraint: "1" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Success", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "UrlNotAvailable", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AuthFailed", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TextTrackNotAvailable", id: 3, conformance: "TT" }),
    (0, import_elements.FieldElement)({ name: "AudioTrackNotAvailable", id: 4, conformance: "AT" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ParameterEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Actor", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Channel", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Character", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Director", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Event", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Franchise", id: 5, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Genre", id: 6, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "League", id: 7, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Popularity", id: 8, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Provider", id: 9, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Sport", id: 10, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SportsTeam", id: 11, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Type", id: 12, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Video", id: 13, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Season", id: 14, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Episode", id: 15, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Any", id: 16, conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "MetricTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Pixels", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Percentage", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "AdditionalInfoStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Name", id: 0, type: "string", conformance: "M", constraint: "max 256" }),
    (0, import_elements.FieldElement)({ name: "Value", id: 1, type: "string", conformance: "M", constraint: "max 8192" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ParameterStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Type", id: 0, type: "ParameterEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Value", id: 1, type: "string", conformance: "M", constraint: "max 1024" }),
    (0, import_elements.FieldElement)(
      { name: "ExternalIdList", id: 2, type: "list", conformance: "O", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "AdditionalInfoStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ContentSearchStruct", type: "struct" },
    (0, import_elements.FieldElement)(
      { name: "ParameterList", id: 0, type: "list", conformance: "M" },
      (0, import_elements.FieldElement)({ name: "entry", type: "ParameterStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DimensionStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Width", id: 0, type: "double", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Height", id: 1, type: "double", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Metric", id: 2, type: "MetricTypeEnum", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StyleInformationStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "ImageUrl", id: 0, type: "string", conformance: "O", constraint: "max 8192" }),
    (0, import_elements.FieldElement)({ name: "Color", id: 1, type: "string", conformance: "O", constraint: "7, 9" }),
    (0, import_elements.FieldElement)({ name: "Size", id: 2, type: "DimensionStruct", conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "BrandingInformationStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "ProviderName", id: 0, type: "string", conformance: "M", constraint: "max 256" }),
    (0, import_elements.FieldElement)({ name: "Background", id: 1, type: "StyleInformationStruct", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Logo", id: 2, type: "StyleInformationStruct", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "ProgressBar", id: 3, type: "StyleInformationStruct", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Splash", id: 4, type: "StyleInformationStruct", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Watermark", id: 5, type: "StyleInformationStruct", conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "PlaybackPreferencesStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "PlaybackPosition", id: 0, type: "uint64", conformance: "AS", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "TextTrack", id: 1, type: "TrackPreferenceStruct", conformance: "TT", quality: "X" }),
    (0, import_elements.FieldElement)(
      { name: "AudioTracks", id: 2, type: "list", conformance: "AT", quality: "X" },
      (0, import_elements.FieldElement)({ name: "entry", type: "TrackPreferenceStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "TrackPreferenceStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "LanguageCode", id: 0, type: "string", conformance: "M", constraint: "max 32" }),
    (0, import_elements.FieldElement)(
      { name: "Characteristics", id: 1, type: "list", conformance: "O", default: null, quality: "X" },
      (0, import_elements.FieldElement)({ name: "entry", type: "MediaPlayback.CharacteristicEnum" })
    ),
    (0, import_elements.FieldElement)({ name: "AudioOutputIndex", id: 2, type: "uint8", conformance: "AT", quality: "X" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ContentLauncher);
