// @matter/model/dist/cjs/standard/elements/content-app-observer.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var content_app_observer_element_exports = {};
__export(content_app_observer_element_exports, {
  ContentAppObserver: () => ContentAppObserver
});
module.exports = __toCommonJS(content_app_observer_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ContentAppObserver = (0, import_elements.ClusterElement)(
  { name: "ContentAppObserver", id: 1296 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.CommandElement)(
    {
      name: "ContentAppMessage",
      id: 0,
      access: "O",
      conformance: "M",
      direction: "request",
      response: "ContentAppMessageResponse"
    },
    (0, import_elements.FieldElement)({ name: "Data", id: 0, type: "string", conformance: "M", constraint: "max 500" }),
    (0, import_elements.FieldElement)({ name: "EncodingHint", id: 1, type: "string", conformance: "O", constraint: "max 100" })
  ),
  (0, import_elements.CommandElement)(
    { name: "ContentAppMessageResponse", id: 1, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "StatusEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Data", id: 1, type: "string", conformance: "O", constraint: "max 500" }),
    (0, import_elements.FieldElement)({ name: "EncodingHint", id: 2, type: "string", conformance: "O", constraint: "max 100" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Success", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "UnexpectedData", id: 1, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ContentAppObserver);
