// @matter/model/dist/cjs/standard/elements/content-control.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var content_control_element_exports = {};
__export(content_control_element_exports, {
  ContentControl: () => ContentControl
});
module.exports = __toCommonJS(content_control_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ContentControl = (0, import_elements.ClusterElement)(
  { name: "ContentControl", id: 1295 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "ST", constraint: "0", title: "ScreenTime" }),
    (0, import_elements.FieldElement)({ name: "PM", constraint: "1", title: "PinManagement" }),
    (0, import_elements.FieldElement)({ name: "BU", constraint: "2", title: "BlockUnrated" }),
    (0, import_elements.FieldElement)({ name: "OCR", constraint: "3", title: "OnDemandContentRating" }),
    (0, import_elements.FieldElement)({ name: "SCR", constraint: "4", title: "ScheduledContentRating" }),
    (0, import_elements.FieldElement)({ name: "BC", constraint: "5", title: "BlockChannels" }),
    (0, import_elements.FieldElement)({ name: "BA", constraint: "6", title: "BlockApplications" }),
    (0, import_elements.FieldElement)({ name: "BTW", constraint: "7", title: "BlockContentTimeWindow" })
  ),
  (0, import_elements.AttributeElement)({ name: "Enabled", id: 0, type: "bool", access: "R V", conformance: "M" }),
  (0, import_elements.AttributeElement)(
    { name: "OnDemandRatings", id: 1, type: "list", access: "R V", conformance: "OCR" },
    (0, import_elements.FieldElement)({ name: "entry", type: "RatingNameStruct" })
  ),
  (0, import_elements.AttributeElement)({
    name: "OnDemandRatingThreshold",
    id: 2,
    type: "string",
    access: "R V",
    conformance: "OCR",
    constraint: "max 8"
  }),
  (0, import_elements.AttributeElement)(
    { name: "ScheduledContentRatings", id: 3, type: "list", access: "R V", conformance: "SCR" },
    (0, import_elements.FieldElement)({ name: "entry", type: "RatingNameStruct" })
  ),
  (0, import_elements.AttributeElement)({
    name: "ScheduledContentRatingThreshold",
    id: 4,
    type: "string",
    access: "R V",
    conformance: "SCR",
    constraint: "max 8"
  }),
  (0, import_elements.AttributeElement)({ name: "ScreenDailyTime", id: 5, type: "elapsed-s", access: "R V", conformance: "ST", constraint: "max 86400" }),
  (0, import_elements.AttributeElement)({
    name: "RemainingScreenTime",
    id: 6,
    type: "elapsed-s",
    access: "R V",
    conformance: "ST",
    constraint: "max 86400"
  }),
  (0, import_elements.AttributeElement)({ name: "BlockUnrated", id: 7, type: "bool", access: "R V", conformance: "BU" }),
  (0, import_elements.AttributeElement)(
    { name: "BlockChannelList", id: 8, type: "list", access: "R V", conformance: "BC" },
    (0, import_elements.FieldElement)({ name: "entry", type: "BlockChannelStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "BlockApplicationList", id: 9, type: "list", access: "R V", conformance: "BA" },
    (0, import_elements.FieldElement)({ name: "entry", type: "AppInfoStruct" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "BlockContentTimeWindow",
      id: 10,
      type: "list",
      access: "R V",
      conformance: "BTW",
      constraint: "max 7"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "TimeWindowStruct" })
  ),
  (0, import_elements.EventElement)({ name: "RemainingScreenTimeExpired", id: 0, access: "V", conformance: "ST", priority: "info" }),
  (0, import_elements.EventElement)({ name: "EnteringBlockContentTimeWindow", id: 1, access: "V", conformance: "BTW", priority: "info" }),
  (0, import_elements.CommandElement)(
    { name: "UpdatePin", id: 0, access: "M T", conformance: "PM", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "OldPin", id: 0, type: "string", conformance: "M", constraint: "max 6" }),
    (0, import_elements.FieldElement)({ name: "NewPin", id: 1, type: "string", conformance: "M", constraint: "max 6" })
  ),
  (0, import_elements.CommandElement)({ name: "ResetPin", id: 1, access: "A T", conformance: "PM", direction: "request", response: "ResetPinResponse" }),
  (0, import_elements.CommandElement)(
    { name: "ResetPinResponse", id: 2, conformance: "PM", direction: "response" },
    (0, import_elements.FieldElement)({ name: "PinCode", id: 0, type: "string", conformance: "M", constraint: "max 6" })
  ),
  (0, import_elements.CommandElement)({ name: "Enable", id: 3, access: "M T", conformance: "M", direction: "request", response: "status" }),
  (0, import_elements.CommandElement)({ name: "Disable", id: 4, access: "M T", conformance: "M", direction: "request", response: "status" }),
  (0, import_elements.CommandElement)(
    { name: "AddBonusTime", id: 5, access: "O", conformance: "ST", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "PinCode", id: 0, type: "string", conformance: "O", constraint: "max 6" }),
    (0, import_elements.FieldElement)({ name: "BonusTime", id: 1, type: "elapsed-s", conformance: "M", constraint: "desc", default: 300 })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "SetScreenDailyTime",
      id: 6,
      access: "M",
      conformance: "ST",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ScreenTime", id: 0, type: "elapsed-s", conformance: "M", constraint: "max 86400" })
  ),
  (0, import_elements.CommandElement)({ name: "BlockUnratedContent", id: 7, access: "M", conformance: "BU", direction: "request", response: "status" }),
  (0, import_elements.CommandElement)({
    name: "UnblockUnratedContent",
    id: 8,
    access: "M",
    conformance: "BU",
    direction: "request",
    response: "status"
  }),
  (0, import_elements.CommandElement)(
    {
      name: "SetOnDemandRatingThreshold",
      id: 9,
      access: "M",
      conformance: "OCR",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "Rating", id: 0, type: "string", conformance: "M", constraint: "max 8" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "SetScheduledContentRatingThreshold",
      id: 10,
      access: "M",
      conformance: "SCR",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "Rating", id: 0, type: "string", conformance: "M", constraint: "max 8" })
  ),
  (0, import_elements.CommandElement)(
    { name: "AddBlockChannels", id: 11, access: "M", conformance: "BC", direction: "request", response: "status" },
    (0, import_elements.FieldElement)(
      { name: "Channels", id: 0, type: "list", conformance: "M" },
      (0, import_elements.FieldElement)({ name: "entry", type: "BlockChannelStruct" })
    )
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RemoveBlockChannels",
      id: 12,
      access: "M",
      conformance: "BC",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)(
      { name: "ChannelIndexes", id: 0, type: "list", conformance: "M" },
      (0, import_elements.FieldElement)({ name: "entry", type: "uint16" })
    )
  ),
  (0, import_elements.CommandElement)(
    {
      name: "AddBlockApplications",
      id: 13,
      access: "M",
      conformance: "BA",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)(
      { name: "Applications", id: 0, type: "list", conformance: "M" },
      (0, import_elements.FieldElement)({ name: "entry", type: "AppInfoStruct" })
    )
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RemoveBlockApplications",
      id: 14,
      access: "M",
      conformance: "BA",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)(
      { name: "Applications", id: 0, type: "list", conformance: "M" },
      (0, import_elements.FieldElement)({ name: "entry", type: "AppInfoStruct" })
    )
  ),
  (0, import_elements.CommandElement)(
    {
      name: "SetBlockContentTimeWindow",
      id: 15,
      access: "M",
      conformance: "BTW",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "TimeWindow", id: 0, type: "TimeWindowStruct", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RemoveBlockContentTimeWindow",
      id: 16,
      access: "M",
      conformance: "BTW",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)(
      { name: "TimeWindowIndexes", id: 0, type: "list", conformance: "M" },
      (0, import_elements.FieldElement)({ name: "entry", type: "uint16" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DayOfWeekBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "Sunday", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "Monday", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "Tuesday", constraint: "2" }),
    (0, import_elements.FieldElement)({ name: "Wednesday", constraint: "3" }),
    (0, import_elements.FieldElement)({ name: "Thursday", constraint: "4" }),
    (0, import_elements.FieldElement)({ name: "Friday", constraint: "5" }),
    (0, import_elements.FieldElement)({ name: "Saturday", constraint: "6" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "RatingNameStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "RatingName", id: 0, type: "string", conformance: "M", constraint: "max 8" }),
    (0, import_elements.FieldElement)({ name: "RatingNameDesc", id: 1, type: "string", conformance: "O", constraint: "max 64" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "BlockChannelStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "BlockChannelIndex", id: 0, type: "uint16", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "MajorNumber", id: 1, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "MinorNumber", id: 2, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Identifier", id: 3, type: "string", conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "AppInfoStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "CatalogVendorId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ApplicationId", id: 1, type: "string", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "TimeWindowStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "TimeWindowIndex", id: 0, type: "uint16", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "DayOfWeek", id: 1, type: "DayOfWeekBitmap", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)(
      { name: "TimePeriod", id: 2, type: "list", conformance: "M", constraint: "desc" },
      (0, import_elements.FieldElement)({ name: "entry", type: "TimePeriodStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "TimePeriodStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "StartHour", id: 0, type: "uint8", conformance: "M", constraint: "0 to 23" }),
    (0, import_elements.FieldElement)({ name: "StartMinute", id: 1, type: "uint8", conformance: "M", constraint: "0 to 59" }),
    (0, import_elements.FieldElement)({ name: "EndHour", id: 2, type: "uint8", conformance: "M", constraint: "0 to 23" }),
    (0, import_elements.FieldElement)({ name: "EndMinute", id: 3, type: "uint8", conformance: "M", constraint: "0 to 59" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StatusCodeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "InvalidPinCode", id: 2 }),
    (0, import_elements.FieldElement)({ name: "InvalidRating", id: 3 }),
    (0, import_elements.FieldElement)({ name: "InvalidChannel", id: 4 }),
    (0, import_elements.FieldElement)({ name: "ChannelAlreadyExist", id: 5 }),
    (0, import_elements.FieldElement)({ name: "ChannelNotExist", id: 6 }),
    (0, import_elements.FieldElement)({ name: "UnidentifiableApplication", id: 7 }),
    (0, import_elements.FieldElement)({ name: "ApplicationAlreadyExist", id: 8 }),
    (0, import_elements.FieldElement)({ name: "ApplicationNotExist", id: 9 }),
    (0, import_elements.FieldElement)({ name: "TimeWindowAlreadyExist", id: 10 }),
    (0, import_elements.FieldElement)({ name: "TimeWindowNotExist", id: 11 })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ContentControl);
