// @matter/model/dist/cjs/standard/elements/descriptor.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var descriptor_element_exports = {};
__export(descriptor_element_exports, {
  Descriptor: () => Descriptor
});
module.exports = __toCommonJS(descriptor_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const Descriptor = (0, import_elements.ClusterElement)(
  { name: "Descriptor", id: 29 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "TAGLIST", conformance: "desc", constraint: "0", title: "TagList" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "DeviceTypeList",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "min 1",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "DeviceTypeStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "ServerList", id: 1, type: "list", access: "R V", conformance: "M", default: [], quality: "F" },
    (0, import_elements.FieldElement)({ name: "entry", type: "cluster-id" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "ClientList", id: 2, type: "list", access: "R V", conformance: "M", default: [], quality: "F" },
    (0, import_elements.FieldElement)({ name: "entry", type: "cluster-id" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "PartsList", id: 3, type: "list", access: "R V", conformance: "M", default: [] },
    (0, import_elements.FieldElement)({ name: "entry", type: "endpoint-no" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "TagList",
      id: 4,
      type: "list",
      access: "R V",
      conformance: "TAGLIST",
      constraint: "1 to 6",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "semtag" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DeviceTypeStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "DeviceType", id: 0, type: "devtype-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Revision", id: 1, type: "uint16", conformance: "M", constraint: "min 1" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(Descriptor);
