// @matter/model/dist/cjs/standard/elements/diagnostic-logs.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var diagnostic_logs_element_exports = {};
__export(diagnostic_logs_element_exports, {
  DiagnosticLogs: () => DiagnosticLogs
});
module.exports = __toCommonJS(diagnostic_logs_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const DiagnosticLogs = (0, import_elements.ClusterElement)(
  { name: "DiagnosticLogs", id: 50 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.CommandElement)(
    {
      name: "RetrieveLogsRequest",
      id: 0,
      access: "O",
      conformance: "M",
      direction: "request",
      response: "RetrieveLogsResponse"
    },
    (0, import_elements.FieldElement)({ name: "Intent", id: 0, type: "IntentEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "RequestedProtocol", id: 1, type: "TransferProtocolEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransferFileDesignator", id: 2, type: "string", conformance: "O", constraint: "max 32" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RetrieveLogsResponse", id: 1, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "StatusEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "LogContent", id: 1, type: "octstr", conformance: "M", constraint: "max 1024" }),
    (0, import_elements.FieldElement)({ name: "UtcTimeStamp", id: 2, type: "epoch-us", conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "TimeSinceBoot", id: 3, type: "systime-us", conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "IntentEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "EndUserSupport", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NetworkDiag", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "CrashLogs", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Success", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Exhausted", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NoLogs", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Busy", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Denied", id: 4, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "TransferProtocolEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "ResponsePayload", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Bdx", id: 1, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(DiagnosticLogs);
