// @matter/model/dist/cjs/standard/elements/ecosystem-information.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ecosystem_information_element_exports = {};
__export(ecosystem_information_element_exports, {
  EcosystemInformation: () => EcosystemInformation
});
module.exports = __toCommonJS(ecosystem_information_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const EcosystemInformation = (0, import_elements.ClusterElement)(
  { name: "EcosystemInformation", id: 1872 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "DeviceDirectory", id: 0, type: "list", access: "R F M", conformance: "M", quality: "N" },
    (0, import_elements.FieldElement)({ name: "entry", type: "EcosystemDeviceStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "LocationDirectory", id: 1, type: "list", access: "R F M", conformance: "M", quality: "N" },
    (0, import_elements.FieldElement)({ name: "entry", type: "EcosystemLocationStruct" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "EcosystemDeviceStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "DeviceName", id: 0, type: "string", access: "S", conformance: "O", constraint: "max 64" }),
    (0, import_elements.FieldElement)({ name: "DeviceNameLastEdit", id: 1, type: "epoch-us", access: "S", conformance: "desc", default: 0 }),
    (0, import_elements.FieldElement)({ name: "BridgedEndpoint", id: 2, type: "endpoint-no", access: "S", conformance: "desc", constraint: "desc" }),
    (0, import_elements.FieldElement)({
      name: "OriginalEndpoint",
      id: 3,
      type: "endpoint-no",
      access: "S",
      conformance: "desc",
      constraint: "desc"
    }),
    (0, import_elements.FieldElement)(
      { name: "DeviceTypes", id: 4, type: "list", access: "S", conformance: "M", constraint: "desc" },
      (0, import_elements.FieldElement)({ name: "entry", type: "Descriptor.DeviceTypeStruct" })
    ),
    (0, import_elements.FieldElement)(
      {
        name: "UniqueLocationIDs",
        id: 5,
        type: "list",
        access: "S",
        conformance: "M",
        constraint: "max 64[max 64]"
      },
      (0, import_elements.FieldElement)({ name: "entry", type: "string" })
    ),
    (0, import_elements.FieldElement)({ name: "UniqueLocationIDsLastEdit", id: 6, type: "epoch-us", access: "S", conformance: "M", default: 0 }),
    (0, import_elements.FieldElement)({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "EcosystemLocationStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "UniqueLocationId", id: 0, type: "string", access: "S", conformance: "M", constraint: "max 64" }),
    (0, import_elements.FieldElement)({ name: "LocationDescriptor", id: 1, type: "locationdesc", access: "S", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "LocationDescriptorLastEdit", id: 2, type: "epoch-us", access: "S", conformance: "M", default: 0 }),
    (0, import_elements.FieldElement)({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(EcosystemInformation);
