// @matter/model/dist/cjs/standard/elements/electrical-power-measurement.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var electrical_power_measurement_element_exports = {};
__export(electrical_power_measurement_element_exports, {
  ElectricalPowerMeasurement: () => ElectricalPowerMeasurement
});
module.exports = __toCommonJS(electrical_power_measurement_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ElectricalPowerMeasurement = (0, import_elements.ClusterElement)(
  { name: "ElectricalPowerMeasurement", id: 144 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "DIRC", conformance: "O.a+", constraint: "0", title: "DirectCurrent" }),
    (0, import_elements.FieldElement)({ name: "ALTC", conformance: "O.a+", constraint: "1", title: "AlternatingCurrent" }),
    (0, import_elements.FieldElement)({ name: "POLY", conformance: "[ALTC]", constraint: "2", title: "PolyphasePower" }),
    (0, import_elements.FieldElement)({ name: "HARM", conformance: "[ALTC]", constraint: "3", title: "Harmonics" }),
    (0, import_elements.FieldElement)({ name: "PWRQ", conformance: "[ALTC]", constraint: "4", title: "PowerQuality" })
  ),
  (0, import_elements.AttributeElement)({ name: "PowerMode", id: 0, type: "PowerModeEnum", access: "R V", conformance: "M" }),
  (0, import_elements.AttributeElement)({
    name: "NumberOfMeasurementTypes",
    id: 1,
    type: "uint8",
    access: "R V",
    conformance: "M",
    constraint: "min 1",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "Accuracy",
      id: 2,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "1 to numberOfMeasurementTypes",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "MeasurementAccuracyStruct" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "Ranges",
      id: 3,
      type: "list",
      access: "R V",
      conformance: "O",
      constraint: "0 to numberOfMeasurementTypes",
      default: [],
      quality: "Q"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "MeasurementRangeStruct" })
  ),
  (0, import_elements.AttributeElement)({ name: "Voltage", id: 4, type: "voltage-mV", access: "R V", conformance: "O", default: null, quality: "X Q" }),
  (0, import_elements.AttributeElement)({
    name: "ActiveCurrent",
    id: 5,
    type: "amperage-mA",
    access: "R V",
    conformance: "O",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "ReactiveCurrent",
    id: 6,
    type: "amperage-mA",
    access: "R V",
    conformance: "[ALTC]",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "ApparentCurrent",
    id: 7,
    type: "amperage-mA",
    access: "R V",
    conformance: "[ALTC]",
    constraint: "min 0",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)({ name: "ActivePower", id: 8, type: "power-mW", access: "R V", conformance: "M", default: null, quality: "X Q" }),
  (0, import_elements.AttributeElement)({
    name: "ReactivePower",
    id: 9,
    type: "power-mW",
    access: "R V",
    conformance: "[ALTC]",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "ApparentPower",
    id: 10,
    type: "power-mW",
    access: "R V",
    conformance: "[ALTC]",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "RmsVoltage",
    id: 11,
    type: "voltage-mV",
    access: "R V",
    conformance: "[ALTC]",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "RmsCurrent",
    id: 12,
    type: "amperage-mA",
    access: "R V",
    conformance: "[ALTC]",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "RmsPower",
    id: 13,
    type: "power-mW",
    access: "R V",
    conformance: "[ALTC]",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "Frequency",
    id: 14,
    type: "int64",
    access: "R V",
    conformance: "[ALTC]",
    constraint: "0 to 1000000",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "HarmonicCurrents",
      id: 15,
      type: "list",
      access: "R V",
      conformance: "HARM",
      constraint: "desc",
      default: null,
      quality: "X Q"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "HarmonicMeasurementStruct" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "HarmonicPhases",
      id: 16,
      type: "list",
      access: "R V",
      conformance: "PWRQ",
      constraint: "desc",
      default: null,
      quality: "X Q"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "HarmonicMeasurementStruct" })
  ),
  (0, import_elements.AttributeElement)({
    name: "PowerFactor",
    id: 17,
    type: "int64",
    access: "R V",
    conformance: "[ALTC]",
    constraint: "-10000 to 10000",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.AttributeElement)({
    name: "NeutralCurrent",
    id: 18,
    type: "amperage-mA",
    access: "R V",
    conformance: "[POLY]",
    default: null,
    quality: "X Q"
  }),
  (0, import_elements.EventElement)(
    { name: "MeasurementPeriodRanges", id: 0, access: "V", conformance: "Ranges", priority: "info" },
    (0, import_elements.FieldElement)(
      { name: "Ranges", id: 0, type: "list", access: "R V", conformance: "M", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "MeasurementRangeStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "PowerModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unknown", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Dc", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Ac", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "MeasurementRangeStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "MeasurementType", id: 0, type: "MeasurementTypeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Min", id: 1, type: "int64", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Max", id: 2, type: "int64", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StartTimestamp", id: 3, type: "epoch-s", conformance: "EndTimestamp" }),
    (0, import_elements.FieldElement)({ name: "EndTimestamp", id: 4, type: "epoch-s", conformance: "desc", constraint: "min startTimestamp + 1" }),
    (0, import_elements.FieldElement)({ name: "MinTimestamp", id: 5, type: "epoch-s", conformance: "EndTimestamp" }),
    (0, import_elements.FieldElement)({
      name: "MaxTimestamp",
      id: 6,
      type: "epoch-s",
      conformance: "EndTimestamp",
      constraint: "min minTimestamp + 1"
    }),
    (0, import_elements.FieldElement)({ name: "StartSystime", id: 7, type: "systime-ms", conformance: "EndSystime" }),
    (0, import_elements.FieldElement)({ name: "EndSystime", id: 8, type: "systime-ms", conformance: "desc", constraint: "min startSystime + 1" }),
    (0, import_elements.FieldElement)({ name: "MinSystime", id: 9, type: "systime-ms", conformance: "EndSystime" }),
    (0, import_elements.FieldElement)({
      name: "MaxSystime",
      id: 10,
      type: "systime-ms",
      conformance: "EndSystime",
      constraint: "min minSystime + 1"
    })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "HarmonicMeasurementStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Order", id: 0, type: "uint8", conformance: "M", constraint: "min 1", default: 1 }),
    (0, import_elements.FieldElement)({ name: "Measurement", id: 1, type: "int64", conformance: "M", default: null, quality: "X" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ElectricalPowerMeasurement);
