// @matter/model/dist/cjs/standard/elements/energy-preference.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var energy_preference_element_exports = {};
__export(energy_preference_element_exports, {
  EnergyPreference: () => EnergyPreference
});
module.exports = __toCommonJS(energy_preference_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const EnergyPreference = (0, import_elements.ClusterElement)(
  { name: "EnergyPreference", id: 155 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "BALA", conformance: "O.a+", constraint: "0", title: "EnergyBalance" }),
    (0, import_elements.FieldElement)({ name: "LPMS", conformance: "O.a+", constraint: "1", title: "LowPowerModeSensitivity" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "EnergyBalances",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "BALA",
      constraint: "2 to 10",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "BalanceStruct" })
  ),
  (0, import_elements.AttributeElement)({ name: "CurrentEnergyBalance", id: 1, type: "uint8", access: "RW VO", conformance: "BALA", quality: "N" }),
  (0, import_elements.AttributeElement)(
    {
      name: "EnergyPriorities",
      id: 2,
      type: "list",
      access: "R V",
      conformance: "BALA",
      constraint: "2",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "EnergyPriorityEnum" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "LowPowerModeSensitivities",
      id: 3,
      type: "list",
      access: "R V",
      conformance: "LPMS",
      constraint: "2 to 10",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "BalanceStruct" })
  ),
  (0, import_elements.AttributeElement)({
    name: "CurrentLowPowerModeSensitivity",
    id: 4,
    type: "uint8",
    access: "RW VO",
    conformance: "LPMS",
    quality: "N"
  }),
  (0, import_elements.DatatypeElement)(
    { name: "EnergyPriorityEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Comfort", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Speed", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Efficiency", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "WaterConsumption", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "BalanceStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Step", id: 0, type: "percent", conformance: "M", quality: "F" }),
    (0, import_elements.FieldElement)({ name: "Label", id: 1, type: "string", conformance: "O", constraint: "max 64", quality: "F" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(EnergyPreference);
