// @matter/model/dist/cjs/standard/elements/ethernet-network-diagnostics.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ethernet_network_diagnostics_element_exports = {};
__export(ethernet_network_diagnostics_element_exports, {
  EthernetNetworkDiagnostics: () => EthernetNetworkDiagnostics
});
module.exports = __toCommonJS(ethernet_network_diagnostics_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const EthernetNetworkDiagnostics = (0, import_elements.ClusterElement)(
  { name: "EthernetNetworkDiagnostics", id: 55, quality: "K" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "PKTCNT", constraint: "0", title: "PacketCounts" }),
    (0, import_elements.FieldElement)({ name: "ERRCNT", constraint: "1", title: "ErrorCounts" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "PhyRate", id: 0, type: "PHYRateEnum", access: "R V", conformance: "O", default: null, quality: "X" }
  ),
  (0, import_elements.AttributeElement)({ name: "FullDuplex", id: 1, type: "bool", access: "R V", conformance: "O", default: null, quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "PacketRxCount", id: 2, type: "uint64", access: "R V", conformance: "PKTCNT", default: 0, quality: "C" }),
  (0, import_elements.AttributeElement)({ name: "PacketTxCount", id: 3, type: "uint64", access: "R V", conformance: "PKTCNT", default: 0, quality: "C" }),
  (0, import_elements.AttributeElement)(
    { name: "TxErrCount", id: 4, type: "uint64", access: "R V", conformance: "ERRCNT", default: 0, quality: "C" }
  ),
  (0, import_elements.AttributeElement)({ name: "CollisionCount", id: 5, type: "uint64", access: "R V", conformance: "ERRCNT", default: 0, quality: "C" }),
  (0, import_elements.AttributeElement)({ name: "OverrunCount", id: 6, type: "uint64", access: "R V", conformance: "ERRCNT", default: 0, quality: "C" }),
  (0, import_elements.AttributeElement)({ name: "CarrierDetect", id: 7, type: "bool", access: "R V", conformance: "O", default: null, quality: "X C" }),
  (0, import_elements.AttributeElement)(
    { name: "TimeSinceReset", id: 8, type: "uint64", access: "R V", conformance: "O", default: 0, quality: "C" }
  ),
  (0, import_elements.CommandElement)({
    name: "ResetCounts",
    id: 0,
    access: "M",
    conformance: "PKTCNT | ERRCNT",
    direction: "request",
    response: "status"
  }),
  (0, import_elements.DatatypeElement)(
    { name: "PHYRateEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Rate10M", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate100M", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate1G", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate25G", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate5G", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate10G", id: 5, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate40G", id: 6, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate100G", id: 7, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate200G", id: 8, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rate400G", id: 9, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(EthernetNetworkDiagnostics);
