// @matter/model/dist/cjs/standard/elements/fan-control.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var fan_control_element_exports = {};
__export(fan_control_element_exports, {
  FanControl: () => FanControl
});
module.exports = __toCommonJS(fan_control_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const FanControl = (0, import_elements.ClusterElement)(
  { name: "FanControl", id: 514 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 4 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "SPD", constraint: "0", title: "MultiSpeed" }),
    (0, import_elements.FieldElement)({ name: "AUT", constraint: "1", title: "Auto" }),
    (0, import_elements.FieldElement)({ name: "RCK", constraint: "2", title: "Rocking" }),
    (0, import_elements.FieldElement)({ name: "WND", constraint: "3", title: "Wind" }),
    (0, import_elements.FieldElement)({ name: "STEP", constraint: "4", title: "Step" }),
    (0, import_elements.FieldElement)({ name: "DIR", constraint: "5", title: "AirflowDirection" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "FanMode", id: 0, type: "FanModeEnum", access: "RW VO", conformance: "M", default: 0, quality: "N" }
  ),
  (0, import_elements.AttributeElement)({ name: "FanModeSequence", id: 1, type: "FanModeSequenceEnum", access: "R V", conformance: "M", quality: "F" }),
  (0, import_elements.AttributeElement)({
    name: "PercentSetting",
    id: 2,
    type: "percent",
    access: "RW VO",
    conformance: "M",
    constraint: "max 100",
    default: 0,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({ name: "PercentCurrent", id: 3, type: "percent", access: "R V", conformance: "M", constraint: "max 100" }),
  (0, import_elements.AttributeElement)({
    name: "SpeedMax",
    id: 4,
    type: "uint8",
    access: "R V",
    conformance: "SPD",
    constraint: "1 to 100",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "SpeedSetting",
    id: 5,
    type: "uint8",
    access: "RW VO",
    conformance: "SPD",
    constraint: "max speedMax",
    default: 0,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "SpeedCurrent",
    id: 6,
    type: "uint8",
    access: "R V",
    conformance: "SPD",
    constraint: "max speedMax",
    quality: "P"
  }),
  (0, import_elements.AttributeElement)({
    name: "RockSupport",
    id: 7,
    type: "RockBitmap",
    access: "R V",
    conformance: "RCK",
    constraint: "desc",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "RockSetting",
    id: 8,
    type: "RockBitmap",
    access: "RW VO",
    conformance: "RCK",
    constraint: "desc",
    default: 0,
    quality: "P"
  }),
  (0, import_elements.AttributeElement)({
    name: "WindSupport",
    id: 9,
    type: "WindBitmap",
    access: "R V",
    conformance: "WND",
    constraint: "desc",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "WindSetting",
    id: 10,
    type: "WindBitmap",
    access: "RW VO",
    conformance: "WND",
    constraint: "desc",
    default: 0,
    quality: "P"
  }),
  (0, import_elements.AttributeElement)({
    name: "AirflowDirection",
    id: 11,
    type: "AirflowDirectionEnum",
    access: "RW VO",
    conformance: "DIR",
    constraint: "desc",
    default: 0,
    quality: "P"
  }),
  (0, import_elements.CommandElement)(
    { name: "Step", id: 0, access: "O", conformance: "STEP", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "Direction", id: 0, type: "StepDirectionEnum", conformance: "M", default: 0 }),
    (0, import_elements.FieldElement)({ name: "Wrap", id: 1, type: "bool", conformance: "O", default: false }),
    (0, import_elements.FieldElement)({ name: "LowestOff", id: 2, type: "bool", conformance: "O", default: true })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "RockBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "RockLeftRight", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "RockUpDown", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "RockRound", constraint: "2" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "WindBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "SleepWind", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "NaturalWind", constraint: "1" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StepDirectionEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Increase", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Decrease", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "AirflowDirectionEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Forward", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Reverse", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "FanModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Off", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Low", id: 1, conformance: "desc" }),
    (0, import_elements.FieldElement)({ name: "Medium", id: 2, conformance: "desc" }),
    (0, import_elements.FieldElement)({ name: "High", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "On", id: 4, conformance: "D" }),
    (0, import_elements.FieldElement)({ name: "Auto", id: 5, conformance: "AUT" }),
    (0, import_elements.FieldElement)({ name: "Smart", id: 6, conformance: "D" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "FanModeSequenceEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "OffLowMedHigh", id: 0, conformance: "[!AUT].a" }),
    (0, import_elements.FieldElement)({ name: "OffLowHigh", id: 1, conformance: "[!AUT].a" }),
    (0, import_elements.FieldElement)({ name: "OffLowMedHighAuto", id: 2, conformance: "[AUT].a" }),
    (0, import_elements.FieldElement)({ name: "OffLowHighAuto", id: 3, conformance: "[AUT].a" }),
    (0, import_elements.FieldElement)({ name: "OffHighAuto", id: 4, conformance: "[AUT].a" }),
    (0, import_elements.FieldElement)({ name: "OffHigh", id: 5, conformance: "[!AUT].a" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(FanControl);
