// @matter/model/dist/cjs/standard/elements/general-diagnostics.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var general_diagnostics_element_exports = {};
__export(general_diagnostics_element_exports, {
  GeneralDiagnostics: () => GeneralDiagnostics
});
module.exports = __toCommonJS(general_diagnostics_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const GeneralDiagnostics = (0, import_elements.ClusterElement)(
  { name: "GeneralDiagnostics", id: 51 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "DMTEST", conformance: "desc", constraint: "0", title: "DataModelTest" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "NetworkInterfaces", id: 0, type: "list", access: "R V", conformance: "M", constraint: "max 8" },
    (0, import_elements.FieldElement)({ name: "entry", type: "NetworkInterface" })
  ),
  (0, import_elements.AttributeElement)({ name: "RebootCount", id: 1, type: "uint16", access: "R V", conformance: "M", quality: "N" }),
  (0, import_elements.AttributeElement)({ name: "UpTime", id: 2, type: "uint64", access: "R V", conformance: "M", quality: "C" }),
  (0, import_elements.AttributeElement)({ name: "TotalOperationalHours", id: 3, type: "uint32", access: "R V", conformance: "O", quality: "N C" }),
  (0, import_elements.AttributeElement)({ name: "BootReason", id: 4, type: "BootReasonEnum", access: "R V", conformance: "O" }),
  (0, import_elements.AttributeElement)(
    { name: "ActiveHardwareFaults", id: 5, type: "list", access: "R V", conformance: "O", constraint: "max 11" },
    (0, import_elements.FieldElement)({ name: "entry", type: "HardwareFaultEnum" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "ActiveRadioFaults", id: 6, type: "list", access: "R V", conformance: "O", constraint: "max 7" },
    (0, import_elements.FieldElement)({ name: "entry", type: "RadioFaultEnum" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "ActiveNetworkFaults", id: 7, type: "list", access: "R V", conformance: "O", constraint: "max 4" },
    (0, import_elements.FieldElement)({ name: "entry", type: "NetworkFaultEnum" })
  ),
  (0, import_elements.AttributeElement)({ name: "TestEventTriggersEnabled", id: 8, type: "bool", access: "R V", conformance: "M" }),
  (0, import_elements.AttributeElement)({ name: "DoNotUse", id: 9, conformance: "X" }),
  (0, import_elements.EventElement)(
    { name: "HardwareFaultChange", id: 0, access: "V", conformance: "O", priority: "critical" },
    (0, import_elements.FieldElement)(
      { name: "Current", id: 0, type: "list", conformance: "M", constraint: "max 11" },
      (0, import_elements.FieldElement)({ name: "entry", type: "HardwareFaultEnum" })
    ),
    (0, import_elements.FieldElement)(
      { name: "Previous", id: 1, type: "list", conformance: "M", constraint: "max 11" },
      (0, import_elements.FieldElement)({ name: "entry", type: "HardwareFaultEnum" })
    )
  ),
  (0, import_elements.EventElement)(
    { name: "RadioFaultChange", id: 1, access: "V", conformance: "O", priority: "critical" },
    (0, import_elements.FieldElement)(
      { name: "Current", id: 0, type: "list", conformance: "M", constraint: "max 7" },
      (0, import_elements.FieldElement)({ name: "entry", type: "RadioFaultEnum" })
    ),
    (0, import_elements.FieldElement)(
      { name: "Previous", id: 1, type: "list", conformance: "M", constraint: "max 7" },
      (0, import_elements.FieldElement)({ name: "entry", type: "RadioFaultEnum" })
    )
  ),
  (0, import_elements.EventElement)(
    { name: "NetworkFaultChange", id: 2, access: "V", conformance: "O", priority: "critical" },
    (0, import_elements.FieldElement)(
      { name: "Current", id: 0, type: "list", conformance: "M", constraint: "max 4" },
      (0, import_elements.FieldElement)({ name: "entry", type: "NetworkFaultEnum" })
    ),
    (0, import_elements.FieldElement)(
      { name: "Previous", id: 1, type: "list", conformance: "M", constraint: "max 4" },
      (0, import_elements.FieldElement)({ name: "entry", type: "NetworkFaultEnum" })
    )
  ),
  (0, import_elements.EventElement)(
    { name: "BootReason", id: 3, access: "V", conformance: "M", priority: "critical" },
    (0, import_elements.FieldElement)({ name: "BootReason", id: 0, type: "BootReasonEnum", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "TestEventTrigger", id: 0, access: "M", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "EnableKey", id: 0, type: "octstr", conformance: "M", constraint: "16" }),
    (0, import_elements.FieldElement)({ name: "EventTrigger", id: 1, type: "uint64", conformance: "M" })
  ),
  (0, import_elements.CommandElement)({
    name: "TimeSnapshot",
    id: 1,
    access: "O",
    conformance: "M",
    direction: "request",
    response: "TimeSnapshotResponse"
  }),
  (0, import_elements.CommandElement)(
    { name: "TimeSnapshotResponse", id: 2, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "SystemTimeMs", id: 0, type: "systime-ms", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "PosixTimeMs", id: 1, type: "posix-ms", conformance: "M", default: null, quality: "X" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "PayloadTestRequest",
      id: 3,
      access: "M",
      conformance: "DMTEST",
      direction: "request",
      response: "PayloadTestResponse"
    },
    (0, import_elements.FieldElement)({ name: "EnableKey", id: 0, type: "octstr", conformance: "M", constraint: "16" }),
    (0, import_elements.FieldElement)({ name: "Value", id: 1, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Count", id: 2, type: "uint16", conformance: "M", constraint: "max 2048" })
  ),
  (0, import_elements.CommandElement)(
    { name: "PayloadTestResponse", id: 4, conformance: "DMTEST", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Payload", id: 0, type: "octstr", conformance: "M", constraint: "max 2048" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "HardwareFaultEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unspecified", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Radio", id: 1, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Sensor", id: 2, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "ResettableOverTemp", id: 3, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "NonResettableOverTemp", id: 4, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "PowerSource", id: 5, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "VisualDisplayFault", id: 6, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "AudioOutputFault", id: 7, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "UserInterfaceFault", id: 8, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "NonVolatileMemoryError", id: 9, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "TamperDetected", id: 10, conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "RadioFaultEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unspecified", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "WiFiFault", id: 1, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "CellularFault", id: 2, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "ThreadFault", id: 3, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "NfcFault", id: 4, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "BleFault", id: 5, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "EthernetFault", id: 6, conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "NetworkFaultEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unspecified", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "HardwareFailure", id: 1, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "NetworkJammed", id: 2, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "ConnectionFailed", id: 3, conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "InterfaceTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unspecified", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "WiFi", id: 1, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Ethernet", id: 2, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Cellular", id: 3, conformance: "O" }),
    (0, import_elements.FieldElement)({ name: "Thread", id: 4, conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "BootReasonEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unspecified", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "PowerOnReboot", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "BrownOutReset", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SoftwareWatchdogReset", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "HardwareWatchdogReset", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SoftwareUpdateCompleted", id: 5, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SoftwareReset", id: 6, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "NetworkInterface", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Name", id: 0, type: "string", conformance: "M", constraint: "max 32" }),
    (0, import_elements.FieldElement)({ name: "IsOperational", id: 1, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({
      name: "OffPremiseServicesReachableIPv4",
      id: 2,
      type: "bool",
      conformance: "M",
      default: null,
      quality: "X"
    }),
    (0, import_elements.FieldElement)({
      name: "OffPremiseServicesReachableIPv6",
      id: 3,
      type: "bool",
      conformance: "M",
      default: null,
      quality: "X"
    }),
    (0, import_elements.FieldElement)({ name: "HardwareAddress", id: 4, type: "hwadr", conformance: "M" }),
    (0, import_elements.FieldElement)(
      { name: "IPv4Addresses", id: 5, type: "list", conformance: "M", constraint: "max 4" },
      (0, import_elements.FieldElement)({ name: "entry", type: "ipv4adr" })
    ),
    (0, import_elements.FieldElement)(
      { name: "IPv6Addresses", id: 6, type: "list", conformance: "M", constraint: "max 8" },
      (0, import_elements.FieldElement)({ name: "entry", type: "ipv6adr" })
    ),
    (0, import_elements.FieldElement)({ name: "Type", id: 7, type: "InterfaceTypeEnum", conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(GeneralDiagnostics);
