// @matter/model/dist/cjs/standard/elements/group-key-management.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var group_key_management_element_exports = {};
__export(group_key_management_element_exports, {
  GroupKeyManagement: () => GroupKeyManagement
});
module.exports = __toCommonJS(group_key_management_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const GroupKeyManagement = (0, import_elements.ClusterElement)(
  { name: "GroupKeyManagement", id: 63 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "CS", conformance: "P", constraint: "0", title: "CacheAndSync" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "GroupKeyMap",
      id: 0,
      type: "list",
      access: "RW F VM",
      conformance: "M",
      constraint: "desc",
      default: [],
      quality: "N"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "GroupKeyMapStruct" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "GroupTable",
      id: 1,
      type: "list",
      access: "R F V",
      conformance: "M",
      constraint: "desc",
      default: []
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "GroupInfoMapStruct" })
  ),
  (0, import_elements.AttributeElement)({ name: "MaxGroupsPerFabric", id: 2, type: "uint16", access: "R V", conformance: "M", default: 0, quality: "F" }),
  (0, import_elements.AttributeElement)({
    name: "MaxGroupKeysPerFabric",
    id: 3,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "1 to 65535",
    default: 1,
    quality: "F"
  }),
  (0, import_elements.CommandElement)(
    { name: "KeySetWrite", id: 0, access: "F A", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "GroupKeySet", id: 0, type: "GroupKeySetStruct", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "KeySetRead",
      id: 1,
      access: "F A",
      conformance: "M",
      direction: "request",
      response: "KeySetReadResponse"
    },
    (0, import_elements.FieldElement)({ name: "GroupKeySetId", id: 0, type: "uint16", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "KeySetReadResponse", id: 2, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "GroupKeySet", id: 0, type: "GroupKeySetStruct", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "KeySetRemove", id: 3, access: "F A", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "GroupKeySetId", id: 0, type: "uint16", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "KeySetReadAllIndices",
      id: 4,
      access: "F A",
      conformance: "M",
      direction: "request",
      response: "KeySetReadAllIndicesResponse"
    },
    (0, import_elements.FieldElement)({ name: "DoNotUse", id: 0, conformance: "X" })
  ),
  (0, import_elements.CommandElement)(
    { name: "KeySetReadAllIndicesResponse", id: 5, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)(
      { name: "GroupKeySetIDs", id: 0, type: "list", conformance: "M" },
      (0, import_elements.FieldElement)({ name: "entry", type: "uint16" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "GroupKeySecurityPolicyEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "TrustFirst", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "CacheAndSync", id: 1, conformance: "CS" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "GroupKeyMulticastPolicyEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "PerGroupId", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AllNodes", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "GroupKeyMapStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 1, type: "group-id", access: "F", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "GroupKeySetId", id: 2, type: "uint16", access: "F", conformance: "M", constraint: "1 to 65535" }),
    (0, import_elements.FieldElement)({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "GroupKeySetStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "GroupKeySetId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "GroupKeySecurityPolicy", id: 1, type: "GroupKeySecurityPolicyEnum", access: "S", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EpochKey0", id: 2, type: "octstr", access: "S", conformance: "M", constraint: "16", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochStartTime0", id: 3, type: "epoch-us", access: "S", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochKey1", id: 4, type: "octstr", access: "S", conformance: "M", constraint: "16", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochStartTime1", id: 5, type: "epoch-us", access: "S", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochKey2", id: 6, type: "octstr", access: "S", conformance: "M", constraint: "16", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochStartTime2", id: 7, type: "epoch-us", access: "S", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({
      name: "GroupKeyMulticastPolicy",
      id: 8,
      type: "GroupKeyMulticastPolicyEnum",
      access: "S",
      conformance: "O",
      default: 0
    })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "GroupInfoMapStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 1, type: "group-id", access: "F", conformance: "M" }),
    (0, import_elements.FieldElement)(
      { name: "Endpoints", id: 2, type: "list", access: "F", conformance: "M", constraint: "min 1" },
      (0, import_elements.FieldElement)({ name: "entry", type: "endpoint-no" })
    ),
    (0, import_elements.FieldElement)({ name: "GroupName", id: 3, type: "string", access: "F", conformance: "O", constraint: "max 16" }),
    (0, import_elements.FieldElement)({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(GroupKeyManagement);
