// @matter/model/dist/cjs/standard/elements/icd-management.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var icd_management_element_exports = {};
__export(icd_management_element_exports, {
  IcdManagement: () => IcdManagement
});
module.exports = __toCommonJS(icd_management_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const IcdManagement = (0, import_elements.ClusterElement)(
  { name: "IcdManagement", id: 70 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "CIP", conformance: "LITS, O", constraint: "0", title: "CheckInProtocolSupport" }),
    (0, import_elements.FieldElement)({ name: "UAT", conformance: "LITS, O", constraint: "1", title: "UserActiveModeTrigger" }),
    (0, import_elements.FieldElement)({ name: "LITS", conformance: "O", constraint: "2", title: "LongIdleTimeSupport" }),
    (0, import_elements.FieldElement)({ name: "DSLS", conformance: "[LITS]", constraint: "3", title: "DynamicSitLitSupport" })
  ),
  (0, import_elements.AttributeElement)({
    name: "IdleModeDuration",
    id: 0,
    type: "uint32",
    access: "R V",
    conformance: "M",
    constraint: "1 to 64800",
    default: 1,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "ActiveModeDuration",
    id: 1,
    type: "uint32",
    access: "R V",
    conformance: "M",
    default: 300,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "ActiveModeThreshold",
    id: 2,
    type: "uint16",
    access: "R V",
    conformance: "M",
    default: 300,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "RegisteredClients",
      id: 3,
      type: "list",
      access: "R F A",
      conformance: "CIP",
      constraint: "desc",
      default: [],
      quality: "N"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "MonitoringRegistrationStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "IcdCounter", id: 4, type: "uint32", access: "R A", conformance: "CIP", default: 0, quality: "N C" }
  ),
  (0, import_elements.AttributeElement)({
    name: "ClientsSupportedPerFabric",
    id: 5,
    type: "uint16",
    access: "R V",
    conformance: "CIP",
    constraint: "min 1",
    default: 1,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "UserActiveModeTriggerHint",
    id: 6,
    type: "UserActiveModeTriggerBitmap",
    access: "R V",
    conformance: "UAT",
    constraint: "desc",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "UserActiveModeTriggerInstruction",
    id: 7,
    type: "string",
    access: "R V",
    conformance: "desc",
    constraint: "max 128",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({ name: "OperatingMode", id: 8, type: "OperatingModeEnum", access: "R V", conformance: "LITS" }),
  (0, import_elements.AttributeElement)({
    name: "MaximumCheckInBackoff",
    id: 9,
    type: "uint32",
    access: "R V",
    conformance: "CIP",
    constraint: "idleModeDuration to 64800",
    default: 1,
    quality: "F"
  }),
  (0, import_elements.CommandElement)(
    {
      name: "RegisterClient",
      id: 0,
      access: "F M",
      conformance: "CIP",
      direction: "request",
      response: "RegisterClientResponse"
    },
    (0, import_elements.FieldElement)({ name: "CheckInNodeId", id: 0, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "MonitoredSubject", id: 1, type: "subject-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Key", id: 2, type: "octstr", conformance: "M", constraint: "16" }),
    (0, import_elements.FieldElement)({ name: "VerificationKey", id: 3, type: "octstr", conformance: "O", constraint: "16" }),
    (0, import_elements.FieldElement)({ name: "ClientType", id: 4, type: "ClientTypeEnum", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RegisterClientResponse", id: 1, conformance: "CIP", direction: "response" },
    (0, import_elements.FieldElement)({ name: "IcdCounter", id: 0, type: "uint32", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "UnregisterClient",
      id: 2,
      access: "F M",
      conformance: "CIP",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "CheckInNodeId", id: 0, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "VerificationKey", id: 1, type: "octstr", conformance: "O", constraint: "16" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "StayActiveRequest",
      id: 3,
      access: "O",
      conformance: "LITS, O",
      direction: "request",
      response: "StayActiveResponse"
    },
    (0, import_elements.FieldElement)({ name: "StayActiveDuration", id: 0, type: "uint32", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "StayActiveResponse", id: 4, conformance: "LITS, O", direction: "response" },
    (0, import_elements.FieldElement)({ name: "PromisedActiveDuration", id: 0, type: "uint32", conformance: "M", constraint: "desc" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "UserActiveModeTriggerBitmap", type: "map32" },
    (0, import_elements.FieldElement)({ name: "PowerCycle", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "SettingsMenu", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "CustomInstruction", constraint: "2" }),
    (0, import_elements.FieldElement)({ name: "DeviceManual", constraint: "3" }),
    (0, import_elements.FieldElement)({ name: "ActuateSensor", constraint: "4" }),
    (0, import_elements.FieldElement)({ name: "ActuateSensorSeconds", constraint: "5" }),
    (0, import_elements.FieldElement)({ name: "ActuateSensorTimes", constraint: "6" }),
    (0, import_elements.FieldElement)({ name: "ActuateSensorLightsBlink", constraint: "7" }),
    (0, import_elements.FieldElement)({ name: "ResetButton", constraint: "8" }),
    (0, import_elements.FieldElement)({ name: "ResetButtonLightsBlink", constraint: "9" }),
    (0, import_elements.FieldElement)({ name: "ResetButtonSeconds", constraint: "10" }),
    (0, import_elements.FieldElement)({ name: "ResetButtonTimes", constraint: "11" }),
    (0, import_elements.FieldElement)({ name: "SetupButton", constraint: "12" }),
    (0, import_elements.FieldElement)({ name: "SetupButtonSeconds", constraint: "13" }),
    (0, import_elements.FieldElement)({ name: "SetupButtonLightsBlink", constraint: "14" }),
    (0, import_elements.FieldElement)({ name: "SetupButtonTimes", constraint: "15" }),
    (0, import_elements.FieldElement)({ name: "AppDefinedButton", constraint: "16" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ClientTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Permanent", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Ephemeral", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "OperatingModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Sit", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Lit", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "MonitoringRegistrationStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "CheckInNodeId", id: 1, type: "node-id", access: "S", conformance: "M", quality: "N" }),
    (0, import_elements.FieldElement)({ name: "MonitoredSubject", id: 2, type: "subject-id", access: "S", conformance: "M", quality: "N" }),
    (0, import_elements.FieldElement)({ name: "Key", id: 3, access: "F", conformance: "D" }),
    (0, import_elements.FieldElement)({
      name: "ClientType",
      id: 4,
      type: "ClientTypeEnum",
      access: "S",
      conformance: "M",
      default: 0,
      quality: "N"
    }),
    (0, import_elements.FieldElement)({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(IcdManagement);
