// @matter/model/dist/cjs/standard/elements/illuminance-measurement.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var illuminance_measurement_element_exports = {};
__export(illuminance_measurement_element_exports, {
  IlluminanceMeasurement: () => IlluminanceMeasurement
});
module.exports = __toCommonJS(illuminance_measurement_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const IlluminanceMeasurement = (0, import_elements.ClusterElement)(
  { name: "IlluminanceMeasurement", id: 1024 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  (0, import_elements.AttributeElement)({
    name: "MeasuredValue",
    id: 0,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "0, minMeasuredValue to maxMeasuredValue",
    default: 0,
    quality: "X P"
  }),
  (0, import_elements.AttributeElement)({
    name: "MinMeasuredValue",
    id: 1,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "1 to 65533",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxMeasuredValue",
    id: 2,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "min minMeasuredValue + 1",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({ name: "Tolerance", id: 3, type: "uint16", access: "R V", conformance: "O", constraint: "max 2048" }),
  (0, import_elements.AttributeElement)({ name: "LightSensorType", id: 4, type: "uint8", access: "R V", conformance: "O", default: null, quality: "X" }),
  (0, import_elements.DatatypeElement)(
    { name: "LightSensorTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Photodiode", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Cmos", id: 1, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(IlluminanceMeasurement);
