// @matter/model/dist/cjs/standard/elements/joint-fabric-pki.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var joint_fabric_pki_element_exports = {};
__export(joint_fabric_pki_element_exports, {
  JointFabricPki: () => JointFabricPki
});
module.exports = __toCommonJS(joint_fabric_pki_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const JointFabricPki = (0, import_elements.ClusterElement)(
  { name: "JointFabricPki", id: 1875 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.CommandElement)(
    {
      name: "IcacsrRequest",
      id: 0,
      access: "A",
      conformance: "M",
      direction: "request",
      response: "IcacsrResponse"
    },
    (0, import_elements.FieldElement)({ name: "Icacsr", id: 0, type: "octstr", conformance: "M", constraint: "max 400" })
  ),
  (0, import_elements.CommandElement)(
    { name: "IcacsrResponse", id: 1, access: "A", conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "StatusCode", id: 0, type: "IcacsrRequestStatusEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Icac", id: 1, type: "octstr", conformance: "O", constraint: "max 400" })
  ),
  (0, import_elements.CommandElement)({
    name: "TransferAnchorRequest",
    id: 2,
    access: "A",
    conformance: "M",
    direction: "request",
    response: "TransferAnchorResponse"
  }),
  (0, import_elements.CommandElement)({ name: "TransferAnchorResponse", id: 3, access: "A", conformance: "M", direction: "response" }),
  (0, import_elements.CommandElement)({
    name: "TransferAnchorComplete",
    id: 4,
    access: "A",
    conformance: "M",
    direction: "request",
    response: "status"
  }),
  (0, import_elements.DatatypeElement)(
    { name: "IcacsrRequestStatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Ok", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvalidIcaCsrFormat", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvalidIcaCsrSignature", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FailedDclVendorIdValidation", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NotAnIcac", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "BusyAnchorTransfer", id: 5, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "IcaCsrSigningFailed", id: 6, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "IcaCsrRequestNoUserConsent", id: 7, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "TransferAnchorResponseStatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Ok", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransferAnchorStatusDatastoreBusy", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransferAnchorStatusNoUserConsent", id: 2, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(JointFabricPki);
