// @matter/model/dist/cjs/standard/elements/laundry-dryer-controls.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var laundry_dryer_controls_element_exports = {};
__export(laundry_dryer_controls_element_exports, {
  LaundryDryerControls: () => LaundryDryerControls
});
module.exports = __toCommonJS(laundry_dryer_controls_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LaundryDryerControls = (0, import_elements.ClusterElement)(
  { name: "LaundryDryerControls", id: 74 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    {
      name: "SupportedDrynessLevels",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "1 to 4"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "DrynessLevelEnum" })
  ),
  (0, import_elements.AttributeElement)({
    name: "SelectedDrynessLevel",
    id: 1,
    type: "DrynessLevelEnum",
    access: "RW VO",
    conformance: "M",
    constraint: "desc",
    quality: "X"
  }),
  (0, import_elements.DatatypeElement)(
    { name: "DrynessLevelEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Low", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Normal", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Extra", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Max", id: 3, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(LaundryDryerControls);
