// @matter/model/dist/cjs/standard/elements/laundry-washer-controls.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var laundry_washer_controls_element_exports = {};
__export(laundry_washer_controls_element_exports, {
  LaundryWasherControls: () => LaundryWasherControls
});
module.exports = __toCommonJS(laundry_washer_controls_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LaundryWasherControls = (0, import_elements.ClusterElement)(
  { name: "LaundryWasherControls", id: 83 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "SPIN", conformance: "O.a+", constraint: "0", title: "Spin" }),
    (0, import_elements.FieldElement)({ name: "RINSE", conformance: "O.a+", constraint: "1", title: "Rinse" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "SpinSpeeds", id: 0, type: "list", access: "R V", conformance: "SPIN", constraint: "max 16[max 64]" },
    (0, import_elements.FieldElement)({ name: "entry", type: "string" })
  ),
  (0, import_elements.AttributeElement)({
    name: "SpinSpeedCurrent",
    id: 1,
    type: "uint8",
    access: "RW VO",
    conformance: "SPIN",
    constraint: "max 15",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "NumberOfRinses",
    id: 2,
    type: "NumberOfRinsesEnum",
    access: "RW VO",
    conformance: "RINSE",
    constraint: "desc",
    default: 1
  }),
  (0, import_elements.AttributeElement)(
    { name: "SupportedRinses", id: 3, type: "list", access: "R V", conformance: "RINSE", constraint: "max 4" },
    (0, import_elements.FieldElement)({ name: "entry", type: "NumberOfRinsesEnum" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "NumberOfRinsesEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "None", id: 0, conformance: "RINSE" }),
    (0, import_elements.FieldElement)({ name: "Normal", id: 1, conformance: "RINSE" }),
    (0, import_elements.FieldElement)({ name: "Extra", id: 2, conformance: "RINSE" }),
    (0, import_elements.FieldElement)({ name: "Max", id: 3, conformance: "RINSE" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(LaundryWasherControls);
