// @matter/model/dist/cjs/standard/elements/localization-configuration.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var localization_configuration_element_exports = {};
__export(localization_configuration_element_exports, {
  LocalizationConfiguration: () => LocalizationConfiguration
});
module.exports = __toCommonJS(localization_configuration_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LocalizationConfiguration = (0, import_elements.ClusterElement)(
  { name: "LocalizationConfiguration", id: 43 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)({
    name: "ActiveLocale",
    id: 0,
    type: "string",
    access: "RW VM",
    conformance: "M",
    constraint: "in SupportedLocales",
    quality: "N"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "SupportedLocales",
      id: 1,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "max 32[max 35]",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "string" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(LocalizationConfiguration);
