// @matter/model/dist/cjs/standard/elements/measurement-accuracy-range-struct.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var measurement_accuracy_range_struct_element_exports = {};
__export(measurement_accuracy_range_struct_element_exports, {
  MeasurementAccuracyRangeStruct: () => MeasurementAccuracyRangeStruct
});
module.exports = __toCommonJS(measurement_accuracy_range_struct_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const MeasurementAccuracyRangeStruct = (0, import_elements.DatatypeElement)(
  { name: "MeasurementAccuracyRangeStruct", type: "struct" },
  (0, import_elements.FieldElement)({ name: "RangeMin", id: 0, type: "int64", conformance: "M", quality: "F" }),
  (0, import_elements.FieldElement)({ name: "RangeMax", id: 1, type: "int64", conformance: "M", quality: "F" }),
  (0, import_elements.FieldElement)({ name: "PercentMax", id: 2, type: "percent100ths", conformance: "O.a+", quality: "F" }),
  (0, import_elements.FieldElement)({
    name: "PercentMin",
    id: 3,
    type: "percent100ths",
    conformance: "[PercentMax]",
    constraint: "max percentTypical",
    quality: "F"
  }),
  (0, import_elements.FieldElement)({
    name: "PercentTypical",
    id: 4,
    type: "percent100ths",
    conformance: "[PercentMin]",
    constraint: "percentMin to percentMax",
    quality: "F"
  }),
  (0, import_elements.FieldElement)({ name: "FixedMax", id: 5, type: "uint64", conformance: "O.a+", quality: "F" }),
  (0, import_elements.FieldElement)({ name: "FixedMin", id: 6, type: "uint64", conformance: "[FixedMax]", constraint: "max fixedMax", quality: "F" }),
  (0, import_elements.FieldElement)({
    name: "FixedTypical",
    id: 7,
    type: "uint64",
    conformance: "[FixedMin]",
    constraint: "fixedMin to fixedMax",
    quality: "F"
  })
);
import_MatterDefinition.MatterDefinition.children.push(MeasurementAccuracyRangeStruct);
