// @matter/model/dist/cjs/standard/elements/microwave-oven-control.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var microwave_oven_control_element_exports = {};
__export(microwave_oven_control_element_exports, {
  MicrowaveOvenControl: () => MicrowaveOvenControl
});
module.exports = __toCommonJS(microwave_oven_control_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const MicrowaveOvenControl = (0, import_elements.ClusterElement)(
  { name: "MicrowaveOvenControl", id: 95 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "PWRNUM", conformance: "O.a", constraint: "0", title: "PowerAsNumber" }),
    (0, import_elements.FieldElement)({ name: "WATTS", conformance: "P, O.a", constraint: "1", title: "PowerInWatts" }),
    (0, import_elements.FieldElement)({ name: "PWRLMTS", conformance: "[PWRNUM]", constraint: "2", title: "PowerNumberLimits" })
  ),
  (0, import_elements.AttributeElement)({
    name: "CookTime",
    id: 0,
    type: "elapsed-s",
    access: "R V",
    conformance: "M",
    constraint: "1 to maxCookTime",
    default: 30
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxCookTime",
    id: 1,
    type: "elapsed-s",
    access: "R V",
    conformance: "M",
    constraint: "1 to 86400",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({ name: "PowerSetting", id: 2, type: "uint8", access: "R V", conformance: "PWRNUM", constraint: "desc" }),
  (0, import_elements.AttributeElement)({
    name: "MinPower",
    id: 3,
    type: "uint8",
    access: "R V",
    conformance: "PWRLMTS",
    constraint: "1 to 99",
    default: 10,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxPower",
    id: 4,
    type: "uint8",
    access: "R V",
    conformance: "PWRLMTS",
    constraint: "minPower + 1 to 100",
    default: 100,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "PowerStep",
    id: 5,
    type: "uint8",
    access: "R V",
    conformance: "PWRLMTS",
    constraint: "desc",
    default: 10,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "SupportedWatts",
      id: 6,
      type: "list",
      access: "R V",
      conformance: "P, WATTS",
      constraint: "1 to 10",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "uint16" })
  ),
  (0, import_elements.AttributeElement)({ name: "SelectedWattIndex", id: 7, type: "uint8", access: "R V", conformance: "P, WATTS", constraint: "desc" }),
  (0, import_elements.AttributeElement)({ name: "WattRating", id: 8, type: "uint16", access: "R V", conformance: "O", quality: "F" }),
  (0, import_elements.CommandElement)(
    {
      name: "SetCookingParameters",
      id: 0,
      access: "O",
      conformance: "M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "CookMode", id: 0, type: "uint8", conformance: "O.b+", constraint: "desc" }),
    (0, import_elements.FieldElement)({
      name: "CookTime",
      id: 1,
      type: "elapsed-s",
      conformance: "O.b+",
      constraint: "1 to maxCookTime",
      default: 30
    }),
    (0, import_elements.FieldElement)({
      name: "PowerSetting",
      id: 2,
      type: "uint8",
      conformance: "[PWRNUM].b+",
      constraint: "minPower to maxPower",
      default: { type: "reference", name: "MaxPower" }
    }),
    (0, import_elements.FieldElement)({ name: "WattSettingIndex", id: 3, type: "uint8", conformance: "[WATTS].b+", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "StartAfterSetting", id: 4, type: "bool", conformance: "O", default: false })
  ),
  (0, import_elements.CommandElement)(
    { name: "AddMoreTime", id: 1, access: "O", conformance: "O", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "TimeToAdd", id: 0, type: "elapsed-s", conformance: "M", constraint: "1 to maxCookTime" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(MicrowaveOvenControl);
