// @matter/model/dist/cjs/standard/elements/mode-base.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var mode_base_element_exports = {};
__export(mode_base_element_exports, {
  ModeBase: () => ModeBase
});
module.exports = __toCommonJS(mode_base_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ModeBase = (0, import_elements.ClusterElement)(
  { name: "ModeBase" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "DEPONOFF", constraint: "0", title: "OnOff" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "SupportedModes",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "2 to 255",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "ModeOptionStruct" })
  ),
  (0, import_elements.AttributeElement)({ name: "CurrentMode", id: 1, type: "uint8", access: "R V", conformance: "M", constraint: "desc", quality: "N" }),
  (0, import_elements.AttributeElement)({
    name: "StartUpMode",
    id: 2,
    type: "uint8",
    access: "RW VO",
    conformance: "O",
    constraint: "desc",
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)({
    name: "OnMode",
    id: 3,
    type: "uint8",
    access: "RW VO",
    conformance: "DEPONOFF",
    constraint: "desc",
    default: null,
    quality: "X N"
  }),
  (0, import_elements.CommandElement)(
    {
      name: "ChangeToMode",
      id: 0,
      access: "O",
      conformance: "M",
      direction: "request",
      response: "ChangeToModeResponse"
    },
    (0, import_elements.FieldElement)({ name: "NewMode", id: 0, type: "uint8", conformance: "M", constraint: "desc" })
  ),
  (0, import_elements.CommandElement)(
    { name: "ChangeToModeResponse", id: 1, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "ModeChangeStatus", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "StatusText", id: 1, type: "string", conformance: "[Status == Success], M", constraint: "max 64" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ModeTagStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "MfgCode", id: 0, type: "vendor-id", conformance: "O", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "Value", id: 1, type: "ModeTag", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ModeOptionStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Label", id: 0, type: "string", conformance: "M", constraint: "max 64", quality: "F" }),
    (0, import_elements.FieldElement)({ name: "Mode", id: 1, type: "uint8", conformance: "M", quality: "F" }),
    (0, import_elements.FieldElement)(
      { name: "ModeTags", id: 2, type: "list", conformance: "M", constraint: "max 8", quality: "F" },
      (0, import_elements.FieldElement)({ name: "entry", type: "ModeTagStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ModeChangeStatus", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Success", id: 0 }),
    (0, import_elements.FieldElement)({ name: "UnsupportedMode", id: 1 }),
    (0, import_elements.FieldElement)({ name: "GenericFailure", id: 2 }),
    (0, import_elements.FieldElement)({ name: "InvalidInMode", id: 3 })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ModeTag", type: "enum16" },
    (0, import_elements.FieldElement)({ name: "Auto", id: 0 }),
    (0, import_elements.FieldElement)({ name: "Quick", id: 1 }),
    (0, import_elements.FieldElement)({ name: "Quiet", id: 2 }),
    (0, import_elements.FieldElement)({ name: "LowNoise", id: 3 }),
    (0, import_elements.FieldElement)({ name: "LowEnergy", id: 4 }),
    (0, import_elements.FieldElement)({ name: "Vacation", id: 5 }),
    (0, import_elements.FieldElement)({ name: "Min", id: 6 }),
    (0, import_elements.FieldElement)({ name: "Max", id: 7 }),
    (0, import_elements.FieldElement)({ name: "Night", id: 8 }),
    (0, import_elements.FieldElement)({ name: "Day", id: 9 })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ModeBase);
