// @matter/model/dist/cjs/standard/elements/mode-select-cluster.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var mode_select_cluster_element_exports = {};
__export(mode_select_cluster_element_exports, {
  ModeSelect: () => ModeSelect
});
module.exports = __toCommonJS(mode_select_cluster_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ModeSelect = (0, import_elements.ClusterElement)(
  { name: "ModeSelect", id: 80 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "DEPONOFF", constraint: "0", title: "OnOff" })
  ),
  (0, import_elements.AttributeElement)({
    name: "Description",
    id: 0,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "max 64",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "StandardNamespace",
    id: 1,
    type: "namespace",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "SupportedModes",
      id: 2,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "max 255",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "ModeOptionStruct" })
  ),
  (0, import_elements.AttributeElement)({ name: "CurrentMode", id: 3, type: "uint8", access: "R V", conformance: "M", constraint: "desc", quality: "N" }),
  (0, import_elements.AttributeElement)({
    name: "StartUpMode",
    id: 4,
    type: "uint8",
    access: "RW VO",
    conformance: "O",
    constraint: "desc",
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)({
    name: "OnMode",
    id: 5,
    type: "uint8",
    access: "RW VO",
    conformance: "DEPONOFF",
    constraint: "desc",
    default: null,
    quality: "X N"
  }),
  (0, import_elements.CommandElement)(
    { name: "ChangeToMode", id: 0, access: "O", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "NewMode", id: 0, type: "uint8", conformance: "M", constraint: "desc" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "SemanticTagStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "MfgCode", id: 0, type: "vendor-id", conformance: "M", constraint: "desc", quality: "F" }),
    (0, import_elements.FieldElement)({ name: "Value", id: 1, type: "uint16", conformance: "M", quality: "F" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ModeOptionStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Label", id: 0, type: "string", conformance: "M", constraint: "max 64", quality: "F" }),
    (0, import_elements.FieldElement)({ name: "Mode", id: 1, type: "uint8", conformance: "M", quality: "F" }),
    (0, import_elements.FieldElement)(
      { name: "SemanticTags", id: 2, type: "list", conformance: "M", constraint: "max 64", quality: "F" },
      (0, import_elements.FieldElement)({ name: "entry", type: "SemanticTagStruct" })
    )
  )
);
import_MatterDefinition.MatterDefinition.children.push(ModeSelect);
