// @matter/model/dist/cjs/standard/elements/power-topology.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var power_topology_element_exports = {};
__export(power_topology_element_exports, {
  PowerTopology: () => PowerTopology
});
module.exports = __toCommonJS(power_topology_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const PowerTopology = (0, import_elements.ClusterElement)(
  { name: "PowerTopology", id: 156 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "NODE", conformance: "O.a", constraint: "0", title: "NodeTopology" }),
    (0, import_elements.FieldElement)({ name: "TREE", conformance: "O.a", constraint: "1", title: "TreeTopology" }),
    (0, import_elements.FieldElement)({ name: "SET", conformance: "O.a", constraint: "2", title: "SetTopology" }),
    (0, import_elements.FieldElement)({ name: "DYPF", conformance: "[SET]", constraint: "3", title: "DynamicPowerFlow" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "AvailableEndpoints",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "SET",
      constraint: "max 20",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "endpoint-no" })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "ActiveEndpoints",
      id: 1,
      type: "list",
      access: "R V",
      conformance: "DYPF",
      constraint: "max 20",
      quality: "N"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "endpoint-no" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(PowerTopology);
