// @matter/model/dist/cjs/standard/elements/pump-configuration-and-control.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pump_configuration_and_control_element_exports = {};
__export(pump_configuration_and_control_element_exports, {
  PumpConfigurationAndControl: () => PumpConfigurationAndControl
});
module.exports = __toCommonJS(pump_configuration_and_control_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const PumpConfigurationAndControl = (0, import_elements.ClusterElement)(
  { name: "PumpConfigurationAndControl", id: 512 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 4 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "PRSCONST", conformance: "O.a+", constraint: "0", title: "ConstantPressure" }),
    (0, import_elements.FieldElement)({ name: "PRSCOMP", conformance: "O.a+", constraint: "1", title: "CompensatedPressure" }),
    (0, import_elements.FieldElement)({ name: "FLW", conformance: "O.a+", constraint: "2", title: "ConstantFlow" }),
    (0, import_elements.FieldElement)({ name: "SPD", conformance: "O.a+", constraint: "3", title: "ConstantSpeed" }),
    (0, import_elements.FieldElement)({ name: "TEMP", conformance: "O.a+", constraint: "4", title: "ConstantTemperature" }),
    (0, import_elements.FieldElement)({ name: "AUTO", conformance: "O", constraint: "5", title: "Automatic" }),
    (0, import_elements.FieldElement)({ name: "LOCAL", conformance: "O", constraint: "6", title: "LocalOperation" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "MaxPressure", id: 0, type: "int16", access: "R V", conformance: "M", default: null, quality: "X F" }
  ),
  (0, import_elements.AttributeElement)(
    { name: "MaxSpeed", id: 1, type: "uint16", access: "R V", conformance: "M", default: null, quality: "X F" }
  ),
  (0, import_elements.AttributeElement)({ name: "MaxFlow", id: 2, type: "uint16", access: "R V", conformance: "M", default: null, quality: "X F" }),
  (0, import_elements.AttributeElement)({
    name: "MinConstPressure",
    id: 3,
    type: "int16",
    access: "R V",
    conformance: "PRSCONST, [AUTO]",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxConstPressure",
    id: 4,
    type: "int16",
    access: "R V",
    conformance: "PRSCONST, [AUTO]",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MinCompPressure",
    id: 5,
    type: "int16",
    access: "R V",
    conformance: "PRSCOMP, [AUTO]",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxCompPressure",
    id: 6,
    type: "int16",
    access: "R V",
    conformance: "PRSCOMP, [AUTO]",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MinConstSpeed",
    id: 7,
    type: "uint16",
    access: "R V",
    conformance: "SPD, [AUTO]",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxConstSpeed",
    id: 8,
    type: "uint16",
    access: "R V",
    conformance: "SPD, [AUTO]",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MinConstFlow",
    id: 9,
    type: "uint16",
    access: "R V",
    conformance: "FLW, [AUTO]",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxConstFlow",
    id: 10,
    type: "uint16",
    access: "R V",
    conformance: "FLW, [AUTO]",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MinConstTemp",
    id: 11,
    type: "int16",
    access: "R V",
    conformance: "TEMP, [AUTO]",
    constraint: "min -27315",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxConstTemp",
    id: 12,
    type: "int16",
    access: "R V",
    conformance: "TEMP, [AUTO]",
    constraint: "min -27315",
    default: null,
    quality: "X F"
  }),
  (0, import_elements.AttributeElement)({
    name: "PumpStatus",
    id: 16,
    type: "PumpStatusBitmap",
    access: "R V",
    conformance: "O",
    constraint: "desc",
    default: 0,
    quality: "P"
  }),
  (0, import_elements.AttributeElement)({
    name: "EffectiveOperationMode",
    id: 17,
    type: "OperationModeEnum",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "EffectiveControlMode",
    id: 18,
    type: "ControlModeEnum",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    quality: "N"
  }),
  (0, import_elements.AttributeElement)(
    { name: "Capacity", id: 19, type: "int16", access: "R V", conformance: "M", default: null, quality: "X P" }
  ),
  (0, import_elements.AttributeElement)({ name: "Speed", id: 20, type: "uint16", access: "R V", conformance: "O", default: null, quality: "X" }),
  (0, import_elements.AttributeElement)({
    name: "LifetimeRunningHours",
    id: 21,
    type: "uint24",
    access: "RW VM",
    conformance: "O",
    default: 0,
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)({ name: "Power", id: 22, type: "uint24", access: "R V", conformance: "O", default: null, quality: "X" }),
  (0, import_elements.AttributeElement)({
    name: "LifetimeEnergyConsumed",
    id: 23,
    type: "uint32",
    access: "RW VM",
    conformance: "O",
    default: 0,
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)({
    name: "OperationMode",
    id: 32,
    type: "OperationModeEnum",
    access: "RW VM",
    conformance: "M",
    constraint: "desc",
    default: 0,
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "ControlMode",
    id: 33,
    type: "ControlModeEnum",
    access: "RW VM",
    conformance: "O",
    constraint: "desc",
    default: 0,
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({ name: "AlarmMask", id: 34, type: "uint16", conformance: "D" }),
  (0, import_elements.EventElement)({ name: "SupplyVoltageLow", id: 0, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "SupplyVoltageHigh", id: 1, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "PowerMissingPhase", id: 2, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "SystemPressureLow", id: 3, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "SystemPressureHigh", id: 4, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "DryRunning", id: 5, access: "V", conformance: "O", priority: "critical" }),
  (0, import_elements.EventElement)({ name: "MotorTemperatureHigh", id: 6, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "PumpMotorFatalFailure", id: 7, access: "V", conformance: "O", priority: "critical" }),
  (0, import_elements.EventElement)({ name: "ElectronicTemperatureHigh", id: 8, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "PumpBlocked", id: 9, access: "V", conformance: "O", priority: "critical" }),
  (0, import_elements.EventElement)({ name: "SensorFailure", id: 10, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "ElectronicNonFatalFailure", id: 11, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "ElectronicFatalFailure", id: 12, access: "V", conformance: "O", priority: "critical" }),
  (0, import_elements.EventElement)({ name: "GeneralFault", id: 13, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "Leakage", id: 14, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "AirDetection", id: 15, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.EventElement)({ name: "TurbineOperation", id: 16, access: "V", conformance: "O", priority: "info" }),
  (0, import_elements.DatatypeElement)(
    { name: "PumpStatusBitmap", type: "map16" },
    (0, import_elements.FieldElement)({ name: "DeviceFault", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "SupplyFault", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "SpeedLow", constraint: "2" }),
    (0, import_elements.FieldElement)({ name: "SpeedHigh", constraint: "3" }),
    (0, import_elements.FieldElement)({ name: "LocalOverride", constraint: "4" }),
    (0, import_elements.FieldElement)({ name: "Running", constraint: "5" }),
    (0, import_elements.FieldElement)({ name: "RemotePressure", constraint: "6" }),
    (0, import_elements.FieldElement)({ name: "RemoteFlow", constraint: "7" }),
    (0, import_elements.FieldElement)({ name: "RemoteTemperature", constraint: "8" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "OperationModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Normal", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Minimum", id: 1, conformance: "SPD" }),
    (0, import_elements.FieldElement)({ name: "Maximum", id: 2, conformance: "SPD" }),
    (0, import_elements.FieldElement)({ name: "Local", id: 3, conformance: "LOCAL" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ControlModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "ConstantSpeed", id: 0, conformance: "SPD" }),
    (0, import_elements.FieldElement)({ name: "ConstantPressure", id: 1, conformance: "PRSCONST" }),
    (0, import_elements.FieldElement)({ name: "ProportionalPressure", id: 2, conformance: "PRSCOMP" }),
    (0, import_elements.FieldElement)({ name: "ConstantFlow", id: 3, conformance: "FLW" }),
    (0, import_elements.FieldElement)({ name: "ConstantTemperature", id: 5, conformance: "TEMP" }),
    (0, import_elements.FieldElement)({ name: "Automatic", id: 7, conformance: "AUTO" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(PumpConfigurationAndControl);
