// @matter/model/dist/cjs/standard/elements/resource-monitoring.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var resource_monitoring_element_exports = {};
__export(resource_monitoring_element_exports, {
  ResourceMonitoring: () => ResourceMonitoring
});
module.exports = __toCommonJS(resource_monitoring_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ResourceMonitoring = (0, import_elements.ClusterElement)(
  { name: "ResourceMonitoring" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "CON", conformance: "O", constraint: "0", title: "Condition" }),
    (0, import_elements.FieldElement)({ name: "WRN", conformance: "O", constraint: "1", title: "Warning" }),
    (0, import_elements.FieldElement)({ name: "REP", conformance: "O", constraint: "2", title: "ReplacementProductList" })
  ),
  (0, import_elements.AttributeElement)({ name: "Condition", id: 0, type: "percent", access: "R V", conformance: "CON" }),
  (0, import_elements.AttributeElement)({
    name: "DegradationDirection",
    id: 1,
    type: "DegradationDirectionEnum",
    access: "R V",
    conformance: "CON",
    constraint: "desc",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({ name: "ChangeIndication", id: 2, type: "ChangeIndicationEnum", access: "R V", conformance: "M", default: 0 }),
  (0, import_elements.AttributeElement)({ name: "InPlaceIndicator", id: 3, type: "bool", access: "R V", conformance: "O" }),
  (0, import_elements.AttributeElement)({
    name: "LastChangedTime",
    id: 4,
    type: "epoch-s",
    access: "RW VO",
    conformance: "O",
    default: null,
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "ReplacementProductList",
      id: 5,
      type: "list",
      access: "R V",
      conformance: "REP",
      constraint: "max 5",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "ReplacementProductStruct" })
  ),
  (0, import_elements.CommandElement)({ name: "ResetCondition", id: 0, access: "O", conformance: "O", direction: "request", response: "status" }),
  (0, import_elements.DatatypeElement)(
    { name: "DegradationDirectionEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Up", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Down", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ChangeIndicationEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Ok", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Warning", id: 1, conformance: "WRN" }),
    (0, import_elements.FieldElement)({ name: "Critical", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ProductIdentifierTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Upc", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Gtin8", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Ean", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Gtin14", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Oem", id: 4, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ReplacementProductStruct", type: "struct" },
    (0, import_elements.FieldElement)({
      name: "ProductIdentifierType",
      id: 0,
      type: "ProductIdentifierTypeEnum",
      conformance: "M",
      constraint: "desc"
    }),
    (0, import_elements.FieldElement)({ name: "ProductIdentifierValue", id: 1, type: "string", conformance: "M", constraint: "max 20" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ResourceMonitoring);
