// @matter/model/dist/cjs/standard/elements/root-node.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var root_node_element_exports = {};
__export(root_node_element_exports, {
  RootNodeDt: () => RootNodeDt
});
module.exports = __toCommonJS(root_node_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const RootNodeDt = (0, import_elements.DeviceTypeElement)(
  { name: "RootNode", id: 22 },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 22, revision: 3 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({ name: "BasicInformation", id: 40, conformance: "M", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)(
    { name: "AccessControl", id: 31, conformance: "M", element: "serverCluster", quality: "I" },
    (0, import_elements.RequirementElement)({ name: "MNGD", conformance: "[ManagedAcl, Allowed]", constraint: "desc", element: "feature" })
  ),
  (0, import_elements.RequirementElement)(
    { name: "PowerSourceConfiguration", id: 46, conformance: "O, D", element: "serverCluster", quality: "I" }
  ),
  (0, import_elements.RequirementElement)({ name: "TimeSynchronization", id: 56, conformance: "O", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)({ name: "GroupKeyManagement", id: 63, conformance: "M", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)({ name: "GeneralCommissioning", id: 48, conformance: "M", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)({ name: "NetworkCommissioning", id: 49, conformance: "!CustomNetworkConfig", element: "serverCluster" }),
  (0, import_elements.RequirementElement)({ name: "AdministratorCommissioning", id: 60, conformance: "M", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)({ name: "OperationalCredentials", id: 62, conformance: "M", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)({
    name: "LocalizationConfiguration",
    id: 43,
    conformance: "LanguageLocale",
    element: "serverCluster",
    quality: "I"
  }),
  (0, import_elements.RequirementElement)(
    { name: "TimeFormatLocalization", id: 44, conformance: "TimeLocale", element: "serverCluster", quality: "I" }
  ),
  (0, import_elements.RequirementElement)({ name: "UnitLocalization", id: 45, conformance: "UnitLocale", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)({ name: "GeneralDiagnostics", id: 51, conformance: "M", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)({ name: "DiagnosticLogs", id: 50, conformance: "O", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)({ name: "SoftwareDiagnostics", id: 52, conformance: "O", element: "serverCluster", quality: "I" }),
  (0, import_elements.RequirementElement)({ name: "EthernetNetworkDiagnostics", id: 55, conformance: "[Ethernet]", element: "serverCluster" }),
  (0, import_elements.RequirementElement)({ name: "WiFiNetworkDiagnostics", id: 54, conformance: "[Wi, Fi]", element: "serverCluster" }),
  (0, import_elements.RequirementElement)({ name: "ThreadNetworkDiagnostics", id: 53, conformance: "[Thread]", element: "serverCluster" }),
  (0, import_elements.RequirementElement)(
    { name: "IcdManagement", id: 70, conformance: "SIT | LIT", element: "serverCluster", quality: "I" },
    (0, import_elements.RequirementElement)({ name: "LONGIDLETIMESUPPORT", conformance: "LIT", element: "feature" })
  ),
  (0, import_elements.FieldElement)(
    { name: "conditions", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "CustomNetworkConfig" }),
    (0, import_elements.FieldElement)({ name: "ManagedAclAllowed" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(RootNodeDt);
