// @matter/model/dist/cjs/standard/elements/rvc-operational-state.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var rvc_operational_state_element_exports = {};
__export(rvc_operational_state_element_exports, {
  RvcOperationalState: () => RvcOperationalState
});
module.exports = __toCommonJS(rvc_operational_state_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const RvcOperationalState = (0, import_elements.ClusterElement)(
  { name: "RvcOperationalState", id: 97, type: "OperationalState" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.CommandElement)({ name: "Pause", id: 0 }),
  (0, import_elements.CommandElement)({ name: "Stop", id: 1, conformance: "X" }),
  (0, import_elements.CommandElement)({ name: "Start", id: 2, conformance: "X" }),
  (0, import_elements.CommandElement)({ name: "Resume", id: 3 }),
  (0, import_elements.CommandElement)({ name: "OperationalCommandResponse", id: 4 }),
  (0, import_elements.CommandElement)({
    name: "GoHome",
    id: 128,
    access: "O",
    conformance: "O",
    direction: "request",
    response: "OperationalCommandResponse"
  }),
  (0, import_elements.DatatypeElement)(
    { name: "OperationalStateEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Stopped", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Running", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Paused", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Error", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SeekingCharger", id: 64, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Charging", id: 65, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Docked", id: 66, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ErrorStateEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "NoError", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "UnableToStartOrResume", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "UnableToCompleteOperation", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "CommandInvalidInState", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FailedToFindChargingDock", id: 64, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Stuck", id: 65, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "DustBinMissing", id: 66, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "DustBinFull", id: 67, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "WaterTankEmpty", id: 68, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "WaterTankMissing", id: 69, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "WaterTankLidOpen", id: 70, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "MopCleaningPadMissing", id: 71, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(RvcOperationalState);
