// @matter/model/dist/cjs/standard/elements/scenes-management.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var scenes_management_element_exports = {};
__export(scenes_management_element_exports, {
  ScenesManagement: () => ScenesManagement
});
module.exports = __toCommonJS(scenes_management_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ScenesManagement = (0, import_elements.ClusterElement)(
  { name: "ScenesManagement", id: 98 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "SN", conformance: "O", constraint: "0", title: "SceneNames" })
  ),
  (0, import_elements.AttributeElement)({
    name: "LastConfiguredBy",
    id: 0,
    type: "node-id",
    access: "R V",
    conformance: "O",
    default: null,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "SceneTableSize",
    id: 1,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 16,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)(
    { name: "FabricSceneInfo", id: 2, type: "list", access: "R F V", conformance: "M", constraint: "desc" },
    (0, import_elements.FieldElement)({ name: "entry", type: "SceneInfoStruct" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "AddScene",
      id: 0,
      access: "F M",
      conformance: "M",
      direction: "request",
      response: "AddSceneResponse"
    },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 1, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint32", conformance: "M", constraint: "max 60000000" }),
    (0, import_elements.FieldElement)({ name: "SceneName", id: 3, type: "string", conformance: "M", constraint: "max 16" }),
    (0, import_elements.FieldElement)(
      { name: "ExtensionFieldSetStructs", id: 4, type: "list", conformance: "M", constraint: "desc" },
      (0, import_elements.FieldElement)({ name: "entry", type: "ExtensionFieldSetStruct" })
    )
  ),
  (0, import_elements.CommandElement)(
    { name: "AddSceneResponse", id: 0, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "status", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "GroupId", id: 1, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 2, type: "uint8", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "ViewScene",
      id: 1,
      access: "F O",
      conformance: "M",
      direction: "request",
      response: "ViewSceneResponse"
    },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 1, type: "uint8", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.CommandElement)(
    { name: "ViewSceneResponse", id: 1, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "status", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "GroupId", id: 1, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 2, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 3, type: "uint32", conformance: "desc", constraint: "max 60000000" }),
    (0, import_elements.FieldElement)({ name: "SceneName", id: 4, type: "string", conformance: "desc", constraint: "max 16" }),
    (0, import_elements.FieldElement)(
      { name: "ExtensionFieldSetStructs", id: 5, type: "list", conformance: "desc" },
      (0, import_elements.FieldElement)({ name: "entry", type: "ExtensionFieldSetStruct" })
    )
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RemoveScene",
      id: 2,
      access: "F M",
      conformance: "M",
      direction: "request",
      response: "RemoveSceneResponse"
    },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 1, type: "uint8", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RemoveSceneResponse", id: 2, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "status", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "GroupId", id: 1, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 2, type: "uint8", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RemoveAllScenes",
      id: 3,
      access: "F M",
      conformance: "M",
      direction: "request",
      response: "RemoveAllScenesResponse"
    },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RemoveAllScenesResponse", id: 3, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "status", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "GroupId", id: 1, type: "group-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "StoreScene",
      id: 4,
      access: "F M",
      conformance: "M",
      direction: "request",
      response: "StoreSceneResponse"
    },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 1, type: "uint8", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.CommandElement)(
    { name: "StoreSceneResponse", id: 4, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "status", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "GroupId", id: 1, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 2, type: "uint8", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RecallScene", id: 5, access: "F O", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 1, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "TransitionTime", id: 2, type: "uint32", conformance: "O", constraint: "max 60000000", quality: "X" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "GetSceneMembership",
      id: 6,
      access: "F O",
      conformance: "M",
      direction: "request",
      response: "GetSceneMembershipResponse"
    },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "GetSceneMembershipResponse", id: 6, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "status", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "Capacity", id: 1, type: "uint8", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "GroupId", id: 2, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)(
      { name: "SceneList", id: 3, type: "list", conformance: "Status == Success" },
      (0, import_elements.FieldElement)({ name: "entry", type: "uint8" })
    )
  ),
  (0, import_elements.CommandElement)(
    {
      name: "CopyScene",
      id: 64,
      access: "F M",
      conformance: "O",
      direction: "request",
      response: "CopySceneResponse"
    },
    (0, import_elements.FieldElement)({ name: "Mode", id: 0, type: "CopyModeBitmap", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "GroupIdentifierFrom", id: 1, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneIdentifierFrom", id: 2, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "GroupIdentifierTo", id: 3, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneIdentifierTo", id: 4, type: "uint8", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.CommandElement)(
    { name: "CopySceneResponse", id: 64, conformance: "CopyScene", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "status", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "GroupIdentifierFrom", id: 1, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneIdentifierFrom", id: 2, type: "uint8", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.DatatypeElement)({ name: "CopyModeBitmap", type: "map8" }, (0, import_elements.FieldElement)({ name: "CopyAllScenes", constraint: "0" })),
  (0, import_elements.DatatypeElement)(
    { name: "SceneInfoStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "SceneCount", id: 0, type: "uint8", access: "F", conformance: "M", default: 0 }),
    (0, import_elements.FieldElement)({
      name: "CurrentScene",
      id: 1,
      type: "uint8",
      access: "S",
      conformance: "M",
      constraint: "desc",
      default: 255
    }),
    (0, import_elements.FieldElement)({ name: "CurrentGroup", id: 2, type: "group-id", access: "S", conformance: "M", default: 0 }),
    (0, import_elements.FieldElement)({ name: "SceneValid", id: 3, type: "bool", access: "S", conformance: "M", default: false }),
    (0, import_elements.FieldElement)({ name: "RemainingCapacity", id: 4, type: "uint8", access: "F", conformance: "M", constraint: "max 253" }),
    (0, import_elements.FieldElement)({ name: "FabricIndex", id: 254, type: "FabricIndex" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "AttributeValuePairStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "AttributeId", id: 0, type: "attrib-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ValueUnsigned8", id: 1, type: "uint8", conformance: "O.a" }),
    (0, import_elements.FieldElement)({ name: "ValueSigned8", id: 2, type: "int8", conformance: "O.a" }),
    (0, import_elements.FieldElement)({ name: "ValueUnsigned16", id: 3, type: "uint16", conformance: "O.a" }),
    (0, import_elements.FieldElement)({ name: "ValueSigned16", id: 4, type: "int16", conformance: "O.a" }),
    (0, import_elements.FieldElement)({ name: "ValueUnsigned32", id: 5, type: "uint32", conformance: "O.a" }),
    (0, import_elements.FieldElement)({ name: "ValueSigned32", id: 6, type: "int32", conformance: "O.a" }),
    (0, import_elements.FieldElement)({ name: "ValueUnsigned64", id: 7, type: "uint64", conformance: "O.a" }),
    (0, import_elements.FieldElement)({ name: "ValueSigned64", id: 8, type: "int64", conformance: "O.a" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ExtensionFieldSetStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "ClusterId", id: 0, type: "cluster-id", conformance: "M" }),
    (0, import_elements.FieldElement)(
      { name: "AttributeValueList", id: 1, type: "list", conformance: "M", constraint: "desc" },
      (0, import_elements.FieldElement)({ name: "entry", type: "AttributeValuePairStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "LogicalSceneTable", type: "struct" },
    (0, import_elements.FieldElement)({ name: "SceneGroupId", id: 0, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SceneId", id: 1, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "SceneName", id: 2, type: "string", conformance: "SN", constraint: "max 16" }),
    (0, import_elements.FieldElement)({
      name: "SceneTransitionTime",
      id: 3,
      type: "uint32",
      conformance: "M",
      constraint: "max 60000000",
      default: 0
    }),
    (0, import_elements.FieldElement)(
      { name: "ExtensionFields", id: 4, type: "list", conformance: "M", default: [] },
      (0, import_elements.FieldElement)({ name: "entry", type: "ExtensionFieldSetStruct" })
    )
  )
);
import_MatterDefinition.MatterDefinition.children.push(ScenesManagement);
