// @matter/model/dist/cjs/standard/elements/target-navigator.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var target_navigator_element_exports = {};
__export(target_navigator_element_exports, {
  TargetNavigator: () => TargetNavigator
});
module.exports = __toCommonJS(target_navigator_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TargetNavigator = (0, import_elements.ClusterElement)(
  { name: "TargetNavigator", id: 1285 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "TargetList", id: 0, type: "list", access: "R V", conformance: "M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "TargetInfoStruct" })
  ),
  (0, import_elements.AttributeElement)({
    name: "CurrentTarget",
    id: 1,
    type: "uint8",
    access: "R V",
    conformance: "O",
    constraint: "desc",
    default: 255
  }),
  (0, import_elements.EventElement)(
    { name: "TargetUpdated", id: 0, access: "V", conformance: "O", priority: "info" },
    (0, import_elements.FieldElement)(
      { name: "TargetList", id: 0, type: "list", conformance: "O" },
      (0, import_elements.FieldElement)({ name: "entry", type: "TargetInfoStruct" })
    ),
    (0, import_elements.FieldElement)({ name: "CurrentTarget", id: 1, type: "uint8", conformance: "O", constraint: "desc", default: 255 }),
    (0, import_elements.FieldElement)({ name: "Data", id: 2, type: "octstr", conformance: "O", constraint: "max 900" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "NavigateTarget",
      id: 0,
      access: "O",
      conformance: "M",
      direction: "request",
      response: "NavigateTargetResponse"
    },
    (0, import_elements.FieldElement)({ name: "Target", id: 0, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Data", id: 1, type: "string", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    { name: "NavigateTargetResponse", id: 1, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Status", id: 0, type: "StatusEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Data", id: 1, type: "string", conformance: "O", constraint: "any" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Success", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TargetNotFound", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NotAllowed", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "TargetInfoStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Identifier", id: 0, type: "uint8", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "Name", id: 1, type: "string", conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(TargetNavigator);
