// @matter/model/dist/cjs/standard/elements/temperature-control.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var temperature_control_element_exports = {};
__export(temperature_control_element_exports, {
  TemperatureControl: () => TemperatureControl
});
module.exports = __toCommonJS(temperature_control_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TemperatureControl = (0, import_elements.ClusterElement)(
  { name: "TemperatureControl", id: 86 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "TN", conformance: "O.a", constraint: "0", title: "TemperatureNumber" }),
    (0, import_elements.FieldElement)({ name: "TL", conformance: "O.a", constraint: "1", title: "TemperatureLevel" }),
    (0, import_elements.FieldElement)({ name: "STEP", conformance: "[TN]", constraint: "2", title: "TemperatureStep" })
  ),
  (0, import_elements.AttributeElement)({
    name: "TemperatureSetpoint",
    id: 0,
    type: "temperature",
    access: "R V",
    conformance: "TN",
    constraint: "minTemperature to maxTemperature"
  }),
  (0, import_elements.AttributeElement)({
    name: "MinTemperature",
    id: 1,
    type: "temperature",
    access: "R V",
    conformance: "TN",
    constraint: "max maxTemperature - 1",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxTemperature",
    id: 2,
    type: "temperature",
    access: "R V",
    conformance: "TN",
    constraint: "desc",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "Step",
    id: 3,
    type: "temperature",
    access: "R V",
    conformance: "STEP",
    constraint: "max maxTemperature - minTemperature",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "SelectedTemperatureLevel",
    id: 4,
    type: "uint8",
    access: "R V",
    conformance: "TL",
    constraint: "max 31"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "SupportedTemperatureLevels",
      id: 5,
      type: "list",
      access: "R V",
      conformance: "TL",
      constraint: "max 32[max 16]"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "string" })
  ),
  (0, import_elements.CommandElement)(
    { name: "SetTemperature", id: 0, access: "O", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "TargetTemperature", id: 0, type: "temperature", conformance: "TN", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "TargetTemperatureLevel", id: 1, type: "uint8", conformance: "TL", constraint: "desc" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(TemperatureControl);
