// @matter/model/dist/cjs/standard/elements/thermostat-user-interface-configuration.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var thermostat_user_interface_configuration_element_exports = {};
__export(thermostat_user_interface_configuration_element_exports, {
  ThermostatUserInterfaceConfiguration: () => ThermostatUserInterfaceConfiguration
});
module.exports = __toCommonJS(thermostat_user_interface_configuration_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ThermostatUserInterfaceConfiguration = (0, import_elements.ClusterElement)(
  { name: "ThermostatUserInterfaceConfiguration", id: 516 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)({
    name: "TemperatureDisplayMode",
    id: 0,
    type: "TemperatureDisplayModeEnum",
    access: "RW VO",
    conformance: "M",
    default: 0
  }),
  (0, import_elements.AttributeElement)({ name: "KeypadLockout", id: 1, type: "KeypadLockoutEnum", access: "RW VM", conformance: "M", default: 0 }),
  (0, import_elements.AttributeElement)({
    name: "ScheduleProgrammingVisibility",
    id: 2,
    type: "ScheduleProgrammingVisibilityEnum",
    access: "RW VM",
    conformance: "O",
    default: 0
  }),
  (0, import_elements.DatatypeElement)(
    { name: "TemperatureDisplayModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Celsius", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Fahrenheit", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "KeypadLockoutEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "NoLockout", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Lockout1", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Lockout2", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Lockout3", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Lockout4", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Lockout5", id: 5, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ScheduleProgrammingVisibilityEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "ScheduleProgrammingPermitted", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ScheduleProgrammingDenied", id: 1, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ThermostatUserInterfaceConfiguration);
