// @matter/model/dist/cjs/standard/elements/thread-network-diagnostics.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var thread_network_diagnostics_element_exports = {};
__export(thread_network_diagnostics_element_exports, {
  ThreadNetworkDiagnostics: () => ThreadNetworkDiagnostics
});
module.exports = __toCommonJS(thread_network_diagnostics_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ThreadNetworkDiagnostics = (0, import_elements.ClusterElement)(
  { name: "ThreadNetworkDiagnostics", id: 53, quality: "K" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "PKTCNT", constraint: "0", title: "PacketCounts" }),
    (0, import_elements.FieldElement)({ name: "ERRCNT", constraint: "1", title: "ErrorCounts" }),
    (0, import_elements.FieldElement)({ name: "MLECNT", constraint: "2", title: "MleCounts" }),
    (0, import_elements.FieldElement)({ name: "MACCNT", constraint: "3", title: "MacCounts" })
  ),
  (0, import_elements.AttributeElement)({ name: "Channel", id: 0, type: "uint16", access: "R V", conformance: "M", quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "RoutingRole", id: 1, type: "RoutingRoleEnum", access: "R V", conformance: "M", quality: "X" }),
  (0, import_elements.AttributeElement)({
    name: "NetworkName",
    id: 2,
    type: "string",
    access: "R V",
    conformance: "M",
    constraint: "max 16",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({ name: "PanId", id: 3, type: "uint16", access: "R V", conformance: "M", quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "ExtendedPanId", id: 4, type: "uint64", access: "R V", conformance: "M", quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "MeshLocalPrefix", id: 5, type: "ipv6pre", access: "R V", conformance: "M", quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "OverrunCount", id: 6, type: "uint64", access: "R V", conformance: "ERRCNT", default: 0, quality: "C" }),
  (0, import_elements.AttributeElement)(
    { name: "NeighborTable", id: 7, type: "list", access: "R V", conformance: "M", default: [] },
    (0, import_elements.FieldElement)({ name: "entry", type: "NeighborTableStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "RouteTable", id: 8, type: "list", access: "R V", conformance: "M", default: [] },
    (0, import_elements.FieldElement)({ name: "entry", type: "RouteTableStruct" })
  ),
  (0, import_elements.AttributeElement)({ name: "PartitionId", id: 9, type: "uint32", access: "R V", conformance: "M", quality: "X" }),
  (0, import_elements.AttributeElement)({
    name: "Weighting",
    id: 10,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "max 255",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "DataVersion",
    id: 11,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "max 255",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "StableDataVersion",
    id: 12,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "max 255",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "LeaderRouterId",
    id: 13,
    type: "uint8",
    access: "R V",
    conformance: "M",
    constraint: "max 62",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "DetachedRoleCount",
    id: 14,
    type: "uint16",
    access: "R V",
    conformance: "[MLECNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "ChildRoleCount",
    id: 15,
    type: "uint16",
    access: "R V",
    conformance: "[MLECNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RouterRoleCount",
    id: 16,
    type: "uint16",
    access: "R V",
    conformance: "[MLECNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "LeaderRoleCount",
    id: 17,
    type: "uint16",
    access: "R V",
    conformance: "[MLECNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "AttachAttemptCount",
    id: 18,
    type: "uint16",
    access: "R V",
    conformance: "[MLECNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "PartitionIdChangeCount",
    id: 19,
    type: "uint16",
    access: "R V",
    conformance: "[MLECNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "BetterPartitionAttachAttemptCount",
    id: 20,
    type: "uint16",
    access: "R V",
    conformance: "[MLECNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "ParentChangeCount",
    id: 21,
    type: "uint16",
    access: "R V",
    conformance: "[MLECNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxTotalCount",
    id: 22,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxUnicastCount",
    id: 23,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxBroadcastCount",
    id: 24,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxAckRequestedCount",
    id: 25,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxAckedCount",
    id: 26,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxNoAckRequestedCount",
    id: 27,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({ name: "TxDataCount", id: 28, type: "uint32", access: "R V", conformance: "[MACCNT]", default: 0, quality: "C" }),
  (0, import_elements.AttributeElement)({
    name: "TxDataPollCount",
    id: 29,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxBeaconCount",
    id: 30,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxBeaconRequestCount",
    id: 31,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxOtherCount",
    id: 32,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxRetryCount",
    id: 33,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxDirectMaxRetryExpiryCount",
    id: 34,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxIndirectMaxRetryExpiryCount",
    id: 35,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxErrCcaCount",
    id: 36,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxErrAbortCount",
    id: 37,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "TxErrBusyChannelCount",
    id: 38,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxTotalCount",
    id: 39,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxUnicastCount",
    id: 40,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxBroadcastCount",
    id: 41,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({ name: "RxDataCount", id: 42, type: "uint32", access: "R V", conformance: "[MACCNT]", default: 0, quality: "C" }),
  (0, import_elements.AttributeElement)({
    name: "RxDataPollCount",
    id: 43,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxBeaconCount",
    id: 44,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxBeaconRequestCount",
    id: 45,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxOtherCount",
    id: 46,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxAddressFilteredCount",
    id: 47,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxDestAddrFilteredCount",
    id: 48,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxDuplicatedCount",
    id: 49,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxErrNoFrameCount",
    id: 50,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxErrUnknownNeighborCount",
    id: 51,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxErrInvalidSrcAddrCount",
    id: 52,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxErrSecCount",
    id: 53,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxErrFcsCount",
    id: 54,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({
    name: "RxErrOtherCount",
    id: 55,
    type: "uint32",
    access: "R V",
    conformance: "[MACCNT]",
    default: 0,
    quality: "C"
  }),
  (0, import_elements.AttributeElement)({ name: "ActiveTimestamp", id: 56, type: "uint64", access: "R V", conformance: "O", default: 0, quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "PendingTimestamp", id: 57, type: "uint64", access: "R V", conformance: "O", default: 0, quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "Delay", id: 58, type: "uint32", access: "R V", conformance: "O", default: 0, quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "SecurityPolicy", id: 59, type: "SecurityPolicy", access: "R V", conformance: "M", quality: "X" }),
  (0, import_elements.AttributeElement)({
    name: "ChannelPage0Mask",
    id: 60,
    type: "octstr",
    access: "R V",
    conformance: "M",
    constraint: "4",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "OperationalDatasetComponents",
    id: 61,
    type: "OperationalDatasetComponents",
    access: "R V",
    conformance: "M",
    quality: "X"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "ActiveNetworkFaultsList",
      id: 62,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "max 4"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "NetworkFaultEnum" })
  ),
  (0, import_elements.AttributeElement)({ name: "ExtAddress", id: 63, type: "uint64", access: "R V", conformance: "P, M", quality: "X" }),
  (0, import_elements.AttributeElement)({ name: "Rloc16", id: 64, type: "uint16", access: "R V", conformance: "P, M", quality: "X" }),
  (0, import_elements.EventElement)(
    { name: "ConnectionStatus", id: 0, access: "V", conformance: "O", priority: "info" },
    (0, import_elements.FieldElement)({ name: "ConnectionStatus", id: 0, type: "ConnectionStatusEnum", conformance: "M" })
  ),
  (0, import_elements.EventElement)(
    { name: "NetworkFaultChange", id: 1, access: "V", conformance: "O", priority: "info" },
    (0, import_elements.FieldElement)(
      { name: "Current", id: 0, type: "list", conformance: "M", constraint: "max 4" },
      (0, import_elements.FieldElement)({ name: "entry", type: "NetworkFaultEnum" })
    ),
    (0, import_elements.FieldElement)(
      { name: "Previous", id: 1, type: "list", conformance: "M", constraint: "max 4" },
      (0, import_elements.FieldElement)({ name: "entry", type: "NetworkFaultEnum" })
    )
  ),
  (0, import_elements.CommandElement)({ name: "ResetCounts", id: 0, access: "M", conformance: "ERRCNT", direction: "request", response: "status" }),
  (0, import_elements.DatatypeElement)(
    { name: "NetworkFaultEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unspecified", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "LinkDown", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "HardwareFailure", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NetworkJammed", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ConnectionStatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Connected", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NotConnected", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "RoutingRoleEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unspecified", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Unassigned", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SleepyEndDevice", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndDevice", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Reed", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Router", id: 5, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Leader", id: 6, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "NeighborTableStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "ExtAddress", id: 0, type: "uint64", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Age", id: 1, type: "uint32", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rloc16", id: 2, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "LinkFrameCounter", id: 3, type: "uint32", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "MleFrameCounter", id: 4, type: "uint32", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Lqi", id: 5, type: "uint8", conformance: "M", constraint: "0 to 255" }),
    (0, import_elements.FieldElement)({
      name: "AverageRssi",
      id: 6,
      type: "int8",
      conformance: "M",
      constraint: "-128 to 0",
      default: null,
      quality: "X"
    }),
    (0, import_elements.FieldElement)({
      name: "LastRssi",
      id: 7,
      type: "int8",
      conformance: "M",
      constraint: "-128 to 0",
      default: null,
      quality: "X"
    }),
    (0, import_elements.FieldElement)({ name: "FrameErrorRate", id: 8, type: "uint8", conformance: "M", constraint: "0 to 100", default: 0 }),
    (0, import_elements.FieldElement)({ name: "MessageErrorRate", id: 9, type: "uint8", conformance: "M", constraint: "0 to 100", default: 0 }),
    (0, import_elements.FieldElement)({ name: "RxOnWhenIdle", id: 10, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FullThreadDevice", id: 11, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FullNetworkData", id: 12, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "IsChild", id: 13, type: "bool", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "RouteTableStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "ExtAddress", id: 0, type: "uint64", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Rloc16", id: 1, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "RouterId", id: 2, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NextHop", id: 3, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "PathCost", id: 4, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "LqiIn", id: 5, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "LqiOut", id: 6, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Age", id: 7, type: "uint8", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Allocated", id: 8, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "LinkEstablished", id: 9, type: "bool", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "SecurityPolicy", type: "struct" },
    (0, import_elements.FieldElement)({ name: "RotationTime", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Flags", id: 1, type: "uint16", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "OperationalDatasetComponents", type: "struct" },
    (0, import_elements.FieldElement)({ name: "ActiveTimestampPresent", id: 0, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "PendingTimestampPresent", id: 1, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "MasterKeyPresent", id: 2, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NetworkNamePresent", id: 3, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ExtendedPanIdPresent", id: 4, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "MeshLocalPrefixPresent", id: 5, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "DelayPresent", id: 6, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "PanIdPresent", id: 7, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ChannelPresent", id: 8, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "PskcPresent", id: 9, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SecurityPolicyPresent", id: 10, type: "bool", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ChannelMaskPresent", id: 11, type: "bool", conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ThreadNetworkDiagnostics);
