// @matter/model/dist/cjs/standard/elements/thread-network-directory.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var thread_network_directory_element_exports = {};
__export(thread_network_directory_element_exports, {
  ThreadNetworkDirectory: () => ThreadNetworkDirectory
});
module.exports = __toCommonJS(thread_network_directory_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ThreadNetworkDirectory = (0, import_elements.ClusterElement)(
  { name: "ThreadNetworkDirectory", id: 1107 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)({
    name: "PreferredExtendedPanId",
    id: 0,
    type: "octstr",
    access: "RW VM",
    conformance: "M",
    constraint: "8",
    default: null,
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "ThreadNetworks",
      id: 1,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "max threadNetworkTableSize",
      quality: "N"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "ThreadNetworkStruct" })
  ),
  (0, import_elements.AttributeElement)({
    name: "ThreadNetworkTableSize",
    id: 2,
    type: "uint8",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 10,
    quality: "F"
  }),
  (0, import_elements.CommandElement)(
    { name: "AddNetwork", id: 0, access: "M T", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "OperationalDataset", id: 0, type: "octstr", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RemoveNetwork", id: 1, access: "M T", conformance: "M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "ExtendedPanId", id: 0, type: "octstr", conformance: "M", constraint: "8" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "GetOperationalDataset",
      id: 2,
      access: "M",
      conformance: "M",
      direction: "request",
      response: "OperationalDatasetResponse"
    },
    (0, import_elements.FieldElement)({ name: "ExtendedPanId", id: 0, type: "octstr", conformance: "M", constraint: "8" })
  ),
  (0, import_elements.CommandElement)(
    { name: "OperationalDatasetResponse", id: 3, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "OperationalDataset", id: 0, type: "octstr", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ThreadNetworkStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "ExtendedPanId", id: 0, type: "octstr", conformance: "M", constraint: "8" }),
    (0, import_elements.FieldElement)({ name: "NetworkName", id: 1, type: "string", conformance: "M", constraint: "1 to 16" }),
    (0, import_elements.FieldElement)({ name: "Channel", id: 2, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ActiveTimestamp", id: 3, type: "uint64", conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ThreadNetworkDirectory);
