// @matter/model/dist/cjs/standard/elements/thread-border-router-management.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var thread_border_router_management_element_exports = {};
__export(thread_border_router_management_element_exports, {
  ThreadBorderRouterManagement: () => ThreadBorderRouterManagement
});
module.exports = __toCommonJS(thread_border_router_management_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ThreadBorderRouterManagement = (0, import_elements.ClusterElement)(
  { name: "ThreadBorderRouterManagement", id: 1106 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "PC", conformance: "O", constraint: "0", title: "PanChange" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "BorderRouterName", id: 0, type: "string", access: "R V", conformance: "M", constraint: "1 to 63" }
  ),
  (0, import_elements.AttributeElement)({ name: "BorderAgentId", id: 1, type: "octstr", access: "R V", conformance: "M", constraint: "16" }),
  (0, import_elements.AttributeElement)({ name: "ThreadVersion", id: 2, type: "uint16", access: "R V", conformance: "M", quality: "F" }),
  (0, import_elements.AttributeElement)({ name: "InterfaceEnabled", id: 3, type: "bool", access: "R V", conformance: "M", default: false, quality: "N" }),
  (0, import_elements.AttributeElement)({
    name: "ActiveDatasetTimestamp",
    id: 4,
    type: "uint64",
    access: "R V",
    conformance: "M",
    default: 0,
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)({
    name: "PendingDatasetTimestamp",
    id: 5,
    type: "uint64",
    access: "R V",
    conformance: "M",
    default: 0,
    quality: "X N"
  }),
  (0, import_elements.CommandElement)({
    name: "GetActiveDatasetRequest",
    id: 0,
    access: "M",
    conformance: "M",
    direction: "request",
    response: "DatasetResponse"
  }),
  (0, import_elements.CommandElement)({
    name: "GetPendingDatasetRequest",
    id: 1,
    access: "M",
    conformance: "M",
    direction: "request",
    response: "DatasetResponse"
  }),
  (0, import_elements.CommandElement)(
    { name: "DatasetResponse", id: 2, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Dataset", id: 0, type: "octstr", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "SetActiveDatasetRequest",
      id: 3,
      access: "M T",
      conformance: "M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ActiveDataset", id: 0, type: "octstr", conformance: "M", constraint: "max 254" }),
    (0, import_elements.FieldElement)({ name: "Breadcrumb", id: 1, type: "uint64", conformance: "O" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "SetPendingDatasetRequest",
      id: 4,
      access: "M T",
      conformance: "PC",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "PendingDataset", id: 0, type: "octstr", conformance: "M", constraint: "max 254" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ThreadBorderRouterManagement);
