// @matter/model/dist/cjs/standard/elements/time-format-localization.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var time_format_localization_element_exports = {};
__export(time_format_localization_element_exports, {
  TimeFormatLocalization: () => TimeFormatLocalization
});
module.exports = __toCommonJS(time_format_localization_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TimeFormatLocalization = (0, import_elements.ClusterElement)(
  { name: "TimeFormatLocalization", id: 44 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "CALFMT", constraint: "0", title: "CalendarFormat" })
  ),
  (0, import_elements.AttributeElement)({ name: "HourFormat", id: 0, type: "HourFormatEnum", access: "RW VM", conformance: "M", quality: "N" }),
  (0, import_elements.AttributeElement)({
    name: "ActiveCalendarType",
    id: 1,
    type: "CalendarTypeEnum",
    access: "RW VM",
    conformance: "CALFMT",
    constraint: "in SupportedCalendarTypes",
    quality: "N"
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "SupportedCalendarTypes",
      id: 2,
      type: "list",
      access: "R V",
      conformance: "CALFMT",
      constraint: "desc",
      quality: "F"
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "CalendarTypeEnum" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "HourFormatEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "12Hr", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "24Hr", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "UseActiveLocale", id: 255, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "CalendarTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Buddhist", id: 0, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Chinese", id: 1, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Coptic", id: 2, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Ethiopian", id: 3, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Gregorian", id: 4, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Hebrew", id: 5, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Indian", id: 6, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Islamic", id: 7, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Japanese", id: 8, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Korean", id: 9, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Persian", id: 10, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "Taiwanese", id: 11, conformance: "O.a+" }),
    (0, import_elements.FieldElement)({ name: "UseActiveLocale", id: 255, conformance: "O.a+" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(TimeFormatLocalization);
