// @matter/model/dist/cjs/standard/elements/wake-on-lan.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var wake_on_lan_element_exports = {};
__export(wake_on_lan_element_exports, {
  WakeOnLan: () => WakeOnLan
});
module.exports = __toCommonJS(wake_on_lan_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const WakeOnLan = (0, import_elements.ClusterElement)(
  { name: "WakeOnLan", id: 1283 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)({
    name: "MacAddress",
    id: 0,
    type: "string",
    access: "R V",
    conformance: "O",
    constraint: "max 12",
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "LinkLocalAddress",
    id: 1,
    type: "ipv6adr",
    access: "R V",
    conformance: "O",
    constraint: "desc",
    quality: "F"
  })
);
import_MatterDefinition.MatterDefinition.children.push(WakeOnLan);
