// @matter/model/dist/cjs/standard/elements/wi-fi-network-diagnostics.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var wi_fi_network_diagnostics_element_exports = {};
__export(wi_fi_network_diagnostics_element_exports, {
  WiFiNetworkDiagnostics: () => WiFiNetworkDiagnostics
});
module.exports = __toCommonJS(wi_fi_network_diagnostics_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const WiFiNetworkDiagnostics = (0, import_elements.ClusterElement)(
  { name: "WiFiNetworkDiagnostics", id: 54, quality: "K" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "PKTCNT", constraint: "0", title: "PacketCounts" }),
    (0, import_elements.FieldElement)({ name: "ERRCNT", constraint: "1", title: "ErrorCounts" })
  ),
  (0, import_elements.AttributeElement)({
    name: "Bssid",
    id: 0,
    type: "octstr",
    access: "R V",
    conformance: "M",
    constraint: "6",
    default: null,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "SecurityType",
    id: 1,
    type: "SecurityTypeEnum",
    access: "R V",
    conformance: "M",
    default: null,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({
    name: "WiFiVersion",
    id: 2,
    type: "WiFiVersionEnum",
    access: "R V",
    conformance: "M",
    default: null,
    quality: "X"
  }),
  (0, import_elements.AttributeElement)({ name: "ChannelNumber", id: 3, type: "uint16", access: "R V", conformance: "M", default: null, quality: "X" }),
  (0, import_elements.AttributeElement)({
    name: "Rssi",
    id: 4,
    type: "int8",
    access: "R V",
    conformance: "M",
    constraint: "-120 to 0",
    default: null,
    quality: "X C"
  }),
  (0, import_elements.AttributeElement)({
    name: "BeaconLostCount",
    id: 5,
    type: "uint32",
    access: "R V",
    conformance: "ERRCNT",
    default: 0,
    quality: "X C"
  }),
  (0, import_elements.AttributeElement)({
    name: "BeaconRxCount",
    id: 6,
    type: "uint32",
    access: "R V",
    conformance: "PKTCNT",
    default: 0,
    quality: "X C"
  }),
  (0, import_elements.AttributeElement)({
    name: "PacketMulticastRxCount",
    id: 7,
    type: "uint32",
    access: "R V",
    conformance: "PKTCNT",
    default: 0,
    quality: "X C"
  }),
  (0, import_elements.AttributeElement)({
    name: "PacketMulticastTxCount",
    id: 8,
    type: "uint32",
    access: "R V",
    conformance: "PKTCNT",
    default: 0,
    quality: "X C"
  }),
  (0, import_elements.AttributeElement)({
    name: "PacketUnicastRxCount",
    id: 9,
    type: "uint32",
    access: "R V",
    conformance: "PKTCNT",
    default: 0,
    quality: "X C"
  }),
  (0, import_elements.AttributeElement)({
    name: "PacketUnicastTxCount",
    id: 10,
    type: "uint32",
    access: "R V",
    conformance: "PKTCNT",
    default: 0,
    quality: "X C"
  }),
  (0, import_elements.AttributeElement)({ name: "CurrentMaxRate", id: 11, type: "uint64", access: "R V", conformance: "O", default: 0, quality: "X C" }),
  (0, import_elements.AttributeElement)({ name: "OverrunCount", id: 12, type: "uint64", access: "R V", conformance: "ERRCNT", default: 0, quality: "X C" }),
  (0, import_elements.EventElement)(
    { name: "Disconnection", id: 0, access: "V", conformance: "O", priority: "info" },
    (0, import_elements.FieldElement)({ name: "ReasonCode", id: 0, type: "uint16", conformance: "M" })
  ),
  (0, import_elements.EventElement)(
    { name: "AssociationFailure", id: 1, access: "V", conformance: "O", priority: "info" },
    (0, import_elements.FieldElement)({ name: "AssociationFailureCause", id: 0, type: "AssociationFailureCauseEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Status", id: 1, type: "uint16", conformance: "M" })
  ),
  (0, import_elements.EventElement)(
    { name: "ConnectionStatus", id: 2, access: "V", conformance: "O", priority: "info" },
    (0, import_elements.FieldElement)({ name: "ConnectionStatus", id: 0, type: "ConnectionStatusEnum", conformance: "M" })
  ),
  (0, import_elements.CommandElement)({ name: "ResetCounts", id: 0, access: "O", conformance: "ERRCNT", direction: "request", response: "status" }),
  (0, import_elements.DatatypeElement)(
    { name: "SecurityTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unspecified", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "None", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Wep", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Wpa", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Wpa2", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Wpa3", id: 5, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "WiFiVersionEnum", type: "enum8" },
    (0, import_elements.FieldElement)({
      name: "A",
      id: 0,
      conformance: "M",
      title: "Indicate the network interface is currently using 802.11a against the wireless access point."
    }),
    (0, import_elements.FieldElement)({
      name: "B",
      id: 1,
      conformance: "M",
      title: "Indicate the network interface is currently using 802.11b against the wireless access point."
    }),
    (0, import_elements.FieldElement)({
      name: "G",
      id: 2,
      conformance: "M",
      title: "Indicate the network interface is currently using 802.11g against the wireless access point."
    }),
    (0, import_elements.FieldElement)({
      name: "N",
      id: 3,
      conformance: "M",
      title: "Indicate the network interface is currently using 802.11n against the wireless access point."
    }),
    (0, import_elements.FieldElement)({ name: "Ac", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Ax", id: 5, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Ah", id: 6, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "AssociationFailureCauseEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Unknown", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AssociationFailed", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AuthenticationFailed", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "SsidNotFound", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ConnectionStatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Connected", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NotConnected", id: 1, conformance: "M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(WiFiNetworkDiagnostics);
