// @matter/model/dist/cjs/standard/elements/wi-fi-network-management.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var wi_fi_network_management_element_exports = {};
__export(wi_fi_network_management_element_exports, {
  WiFiNetworkManagement: () => WiFiNetworkManagement
});
module.exports = __toCommonJS(wi_fi_network_management_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const WiFiNetworkManagement = (0, import_elements.ClusterElement)(
  { name: "WiFiNetworkManagement", id: 1105 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)({
    name: "Ssid",
    id: 0,
    type: "octstr",
    access: "R V",
    conformance: "M",
    constraint: "1 to 32",
    default: null,
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)({
    name: "PassphraseSurrogate",
    id: 1,
    type: "uint64",
    access: "R M",
    conformance: "M",
    default: null,
    quality: "X N"
  }),
  (0, import_elements.CommandElement)({
    name: "NetworkPassphraseRequest",
    id: 0,
    access: "M",
    conformance: "M",
    direction: "request",
    response: "NetworkPassphraseResponse"
  }),
  (0, import_elements.CommandElement)(
    { name: "NetworkPassphraseResponse", id: 1, conformance: "M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Passphrase", id: 0, type: "octstr", conformance: "M", constraint: "max 64" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(WiFiNetworkManagement);
