// @matter/model/dist/cjs/standard/elements/window-covering-cluster.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var window_covering_cluster_element_exports = {};
__export(window_covering_cluster_element_exports, {
  WindowCovering: () => WindowCovering
});
module.exports = __toCommonJS(window_covering_cluster_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const WindowCovering = (0, import_elements.ClusterElement)(
  { name: "WindowCovering", id: 258 },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 5 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    (0, import_elements.FieldElement)({ name: "LF", conformance: "O.a+", constraint: "0", title: "Lift" }),
    (0, import_elements.FieldElement)({ name: "TL", conformance: "O.a+", constraint: "1", title: "Tilt" }),
    (0, import_elements.FieldElement)({ name: "PA_LF", conformance: "[LF]", constraint: "2", title: "PositionAwareLift" }),
    (0, import_elements.FieldElement)({ name: "ABS", conformance: "O", constraint: "3", title: "AbsolutePosition" }),
    (0, import_elements.FieldElement)({ name: "PA_TL", conformance: "[TL]", constraint: "4", title: "PositionAwareTilt" })
  ),
  (0, import_elements.AttributeElement)({
    name: "Type",
    id: 0,
    type: "TypeEnum",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "PhysicalClosedLimitLift",
    id: 1,
    type: "uint16",
    access: "R V",
    conformance: "[LF & PA_LF & ABS]",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "PhysicalClosedLimitTilt",
    id: 2,
    type: "uint16",
    access: "R V",
    conformance: "[TL & PA_TL & ABS]",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentPositionLift",
    id: 3,
    type: "uint16",
    access: "R V",
    conformance: "[LF & PA_LF & ABS]",
    constraint: "installedOpenLimitLift to installedClosedLimitLift",
    default: null,
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentPositionTilt",
    id: 4,
    type: "uint16",
    access: "R V",
    conformance: "[TL & PA_TL & ABS]",
    constraint: "installedOpenLimitTilt to installedClosedLimitTilt",
    default: null,
    quality: "X N"
  }),
  (0, import_elements.AttributeElement)({
    name: "NumberOfActuationsLift",
    id: 5,
    type: "uint16",
    access: "R V",
    conformance: "[LF]",
    default: 0,
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "NumberOfActuationsTilt",
    id: 6,
    type: "uint16",
    access: "R V",
    conformance: "[TL]",
    default: 0,
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "ConfigStatus",
    id: 7,
    type: "ConfigStatusBitmap",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentPositionLiftPercentage",
    id: 8,
    type: "percent",
    access: "R V",
    conformance: "[LF & PA_LF]",
    default: null,
    quality: "X N P"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentPositionTiltPercentage",
    id: 9,
    type: "percent",
    access: "R V",
    conformance: "[TL & PA_TL]",
    default: null,
    quality: "X N P"
  }),
  (0, import_elements.AttributeElement)({
    name: "OperationalStatus",
    id: 10,
    type: "OperationalStatusBitmap",
    access: "R V",
    conformance: "M",
    default: 0,
    quality: "P"
  }),
  (0, import_elements.AttributeElement)({
    name: "TargetPositionLiftPercent100ths",
    id: 11,
    type: "percent100ths",
    access: "R V",
    conformance: "LF & PA_LF",
    default: null,
    quality: "X P"
  }),
  (0, import_elements.AttributeElement)({
    name: "TargetPositionTiltPercent100ths",
    id: 12,
    type: "percent100ths",
    access: "R V",
    conformance: "TL & PA_TL",
    default: null,
    quality: "X P"
  }),
  (0, import_elements.AttributeElement)({
    name: "EndProductType",
    id: 13,
    type: "EndProductTypeEnum",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 0,
    quality: "F"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentPositionLiftPercent100ths",
    id: 14,
    type: "percent100ths",
    access: "R V",
    conformance: "LF & PA_LF",
    constraint: "max 10000",
    default: null,
    quality: "X N P"
  }),
  (0, import_elements.AttributeElement)({
    name: "CurrentPositionTiltPercent100ths",
    id: 15,
    type: "percent100ths",
    access: "R V",
    conformance: "TL & PA_TL",
    constraint: "max 10000",
    default: null,
    quality: "X N P"
  }),
  (0, import_elements.AttributeElement)({
    name: "InstalledOpenLimitLift",
    id: 16,
    type: "uint16",
    access: "R V",
    conformance: "LF & PA_LF & ABS",
    constraint: "max 65534",
    default: 0,
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "InstalledClosedLimitLift",
    id: 17,
    type: "uint16",
    access: "R V",
    conformance: "LF & PA_LF & ABS",
    constraint: "max 65534",
    default: 65534,
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "InstalledOpenLimitTilt",
    id: 18,
    type: "uint16",
    access: "R V",
    conformance: "TL & PA_TL & ABS",
    constraint: "max 65534",
    default: 0,
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({
    name: "InstalledClosedLimitTilt",
    id: 19,
    type: "uint16",
    access: "R V",
    conformance: "TL & PA_TL & ABS",
    constraint: "max 65534",
    default: 65534,
    quality: "N"
  }),
  (0, import_elements.AttributeElement)({ name: "VelocityLift", id: 20, conformance: "D" }),
  (0, import_elements.AttributeElement)({ name: "AccelerationTimeLift", id: 21, conformance: "D" }),
  (0, import_elements.AttributeElement)({ name: "DecelerationTimeLift", id: 22, conformance: "D" }),
  (0, import_elements.AttributeElement)({ name: "Mode", id: 23, type: "ModeBitmap", access: "RW VM", conformance: "M", default: 0, quality: "N" }),
  (0, import_elements.AttributeElement)({ name: "IntermediateSetpointsLift", id: 24, conformance: "D" }),
  (0, import_elements.AttributeElement)({ name: "IntermediateSetpointsTilt", id: 25, conformance: "D" }),
  (0, import_elements.AttributeElement)({
    name: "SafetyStatus",
    id: 26,
    type: "SafetyStatusBitmap",
    access: "R V",
    conformance: "O",
    constraint: "desc",
    default: 0,
    quality: "P"
  }),
  (0, import_elements.CommandElement)({ name: "UpOrOpen", id: 0, access: "O", conformance: "M", direction: "request", response: "status" }),
  (0, import_elements.CommandElement)({ name: "DownOrClose", id: 1, access: "O", conformance: "M", direction: "request", response: "status" }),
  (0, import_elements.CommandElement)({ name: "StopMotion", id: 2, access: "O", conformance: "M", direction: "request", response: "status" }),
  (0, import_elements.CommandElement)(
    {
      name: "GoToLiftValue",
      id: 4,
      access: "O",
      conformance: "[LF & ABS]",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({
      name: "LiftValue",
      id: 0,
      type: "uint16",
      conformance: "M",
      constraint: "installedOpenLimitLift to installedClosedLimitLift"
    })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "GoToLiftPercentage",
      id: 5,
      access: "O",
      conformance: "LF & PA_LF, [LF]",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "LiftPercent100thsValue", id: 0, type: "percent100ths", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "Ignored", id: 1, conformance: "X" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "GoToTiltValue",
      id: 7,
      access: "O",
      conformance: "[TL & ABS]",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({
      name: "TiltValue",
      id: 0,
      type: "uint16",
      conformance: "M",
      constraint: "installedOpenLimitTilt to installedClosedLimitTilt"
    })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "GoToTiltPercentage",
      id: 8,
      access: "O",
      conformance: "TL & PA_TL, [TL]",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "TiltPercent100thsValue", id: 0, type: "percent100ths", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "Ignored", id: 1, conformance: "X" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ConfigStatusBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "Operational", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "OnlineReserved", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "LiftMovementReversed", constraint: "2" }),
    (0, import_elements.FieldElement)({ name: "LiftPositionAware", constraint: "3" }),
    (0, import_elements.FieldElement)({ name: "TiltPositionAware", constraint: "4" }),
    (0, import_elements.FieldElement)({ name: "LiftEncoderControlled", constraint: "5" }),
    (0, import_elements.FieldElement)({ name: "TiltEncoderControlled", constraint: "6" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ModeBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "MotorDirectionReversed", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "CalibrationMode", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "MaintenanceMode", constraint: "2" }),
    (0, import_elements.FieldElement)({ name: "LedFeedback", constraint: "3" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "OperationalStatusBitmap", type: "map8" },
    (0, import_elements.FieldElement)({ name: "Global", type: "MovementStatus", constraint: "0 to 1" }),
    (0, import_elements.FieldElement)({ name: "Lift", type: "MovementStatus", constraint: "2 to 3" }),
    (0, import_elements.FieldElement)({ name: "Tilt", type: "MovementStatus", constraint: "4 to 5" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "SafetyStatusBitmap", type: "map16" },
    (0, import_elements.FieldElement)({ name: "RemoteLockout", constraint: "0" }),
    (0, import_elements.FieldElement)({ name: "TamperDetection", constraint: "1" }),
    (0, import_elements.FieldElement)({ name: "FailedCommunication", constraint: "2" }),
    (0, import_elements.FieldElement)({ name: "PositionFailure", constraint: "3" }),
    (0, import_elements.FieldElement)({ name: "ThermalProtection", constraint: "4" }),
    (0, import_elements.FieldElement)({ name: "ObstacleDetected", constraint: "5" }),
    (0, import_elements.FieldElement)({ name: "Power", constraint: "6" }),
    (0, import_elements.FieldElement)({ name: "StopInput", constraint: "7" }),
    (0, import_elements.FieldElement)({ name: "MotorJammed", constraint: "8" }),
    (0, import_elements.FieldElement)({ name: "HardwareFailure", constraint: "9" }),
    (0, import_elements.FieldElement)({ name: "ManualOperation", constraint: "10" }),
    (0, import_elements.FieldElement)({ name: "Protection", constraint: "11" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "TypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Rollershade", id: 0, conformance: "LF & !TL" }),
    (0, import_elements.FieldElement)({ name: "Rollershade2Motor", id: 1, conformance: "LF & !TL" }),
    (0, import_elements.FieldElement)({ name: "RollershadeExterior", id: 2, conformance: "LF & !TL" }),
    (0, import_elements.FieldElement)({ name: "RollershadeExterior2Motor", id: 3, conformance: "LF & !TL" }),
    (0, import_elements.FieldElement)({ name: "Drapery", id: 4, conformance: "LF & !TL" }),
    (0, import_elements.FieldElement)({ name: "Awning", id: 5, conformance: "LF & !TL" }),
    (0, import_elements.FieldElement)({ name: "Shutter", id: 6, conformance: "LF | TL" }),
    (0, import_elements.FieldElement)({ name: "TiltBlindTiltOnly", id: 7, conformance: "!LF & TL" }),
    (0, import_elements.FieldElement)({ name: "TiltBlindLift", id: 8, conformance: "LF & TL" }),
    (0, import_elements.FieldElement)({ name: "ProjectorScreen", id: 9, conformance: "LF & !TL" }),
    (0, import_elements.FieldElement)({ name: "Unknown", id: 255, conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "EndProductTypeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "RollerShade", id: 0, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "RomanShade", id: 1, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "BalloonShade", id: 2, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "WovenWood", id: 3, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "PleatedShade", id: 4, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "CellularShade", id: 5, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "LayeredShade", id: 6, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "LayeredShade2D", id: 7, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "SheerShade", id: 8, conformance: "LF & TL" }),
    (0, import_elements.FieldElement)({ name: "TiltOnlyInteriorBlind", id: 9, conformance: "TL" }),
    (0, import_elements.FieldElement)({ name: "InteriorBlind", id: 10, conformance: "LF & TL" }),
    (0, import_elements.FieldElement)({ name: "VerticalBlindStripCurtain", id: 11, conformance: "LF & TL" }),
    (0, import_elements.FieldElement)({ name: "InteriorVenetianBlind", id: 12, conformance: "LF & TL" }),
    (0, import_elements.FieldElement)({ name: "ExteriorVenetianBlind", id: 13, conformance: "LF & TL" }),
    (0, import_elements.FieldElement)({ name: "LateralLeftCurtain", id: 14, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "LateralRightCurtain", id: 15, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "CentralCurtain", id: 16, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "RollerShutter", id: 17, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "ExteriorVerticalScreen", id: 18, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "AwningTerracePatio", id: 19, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "AwningVerticalScreen", id: 20, conformance: "LF" }),
    (0, import_elements.FieldElement)({ name: "TiltOnlyPergola", id: 21, conformance: "LF | TL" }),
    (0, import_elements.FieldElement)({ name: "SwingingShutter", id: 22, conformance: "LF | TL" }),
    (0, import_elements.FieldElement)({ name: "SlidingShutter", id: 23, conformance: "LF | TL" }),
    (0, import_elements.FieldElement)({ name: "Unknown", id: 255, conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "MovementStatus", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Stopped", id: 0 }),
    (0, import_elements.FieldElement)({ name: "Opening", id: 1 }),
    (0, import_elements.FieldElement)({ name: "Closing", id: 2 })
  )
);
import_MatterDefinition.MatterDefinition.children.push(WindowCovering);
