// @matter/protocol/dist/cjs/action/request/Write.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Write_exports = {};
__export(Write_exports, {
  Write: () => Write
});
module.exports = __toCommonJS(Write_exports);
var import_Session = require("#session/Session.js");
var import_MalformedRequestError = require("./MalformedRequestError.js");
var import_Specifier = require("./Specifier.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function Write(optionsOrData, ...data) {
  let options;
  if ("kind" in optionsOrData) {
    data = [optionsOrData, ...data];
    options = {};
  } else {
    options = optionsOrData;
  }
  const { writes: writeRequests = [] } = options;
  const result = {
    timedRequest: !!options.timed || !!options.timeout,
    writeRequests,
    interactionModelRevision: options.interactionModelRevision ?? import_Session.FALLBACK_INTERACTIONMODEL_REVISION
  };
  for (const entry of data) {
    reifyData(entry);
  }
  if (!writeRequests.length) {
    throw new import_MalformedRequestError.MalformedRequestError(`Write action contains no attributes to write`);
  }
  return result;
  function reifyData(data2) {
    const cluster = import_Specifier.Specifier.clusterOf(data2);
    if (cluster === void 0) {
      throw new import_MalformedRequestError.MalformedRequestError(`Write action must specify a cluster`);
    }
    let { attributes } = data2;
    if (attributes === void 0) {
      throw new import_MalformedRequestError.MalformedRequestError(`Write action must specify an attribute`);
    }
    if (!Array.isArray(attributes)) {
      attributes = [attributes];
    }
    const { endpoint, value, version: dataVersion } = data2;
    const prototype = {
      path: {
        endpointId: endpoint !== void 0 ? import_Specifier.Specifier.endpointIdOf(data2) : void 0,
        clusterId: cluster.id,
        attributeId: void 0
      },
      dataVersion
    };
    for (const specifier of attributes) {
      const attribute = import_Specifier.Specifier.attributeFor(cluster, specifier);
      writeRequests.push({
        ...prototype,
        path: {
          ...prototype.path,
          attributeId: attribute.id
        },
        data: attribute.schema.encodeTlv(value, { forWriteInteraction: true })
      });
    }
  }
}
((Write2) => {
  function Attribute(data) {
    return {
      kind: "attribute",
      ...data
    };
  }
  Write2.Attribute = Attribute;
})(Write || (Write = {}));
