// @matter/node/dist/esm/storage/server/ServerEndpointStore.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ImplementationError } from "#general";
import { EndpointStore } from "#storage/EndpointStore.js";
import { DatasourceStore } from "./DatasourceStore.js";
const NUMBER_KEY = "__number__";
class ServerEndpointStore extends EndpointStore {
  #number;
  #childStorage;
  #childStores = {};
  get number() {
    return this.#number;
  }
  set number(number) {
    if (this.#number !== number) {
      this.#number = number;
    }
  }
  constructor(storage) {
    super(storage);
    this.#childStorage = storage.createContext("parts");
  }
  /**
   * Create a {@link Datasource.Store} for a behavior.
   */
  createStoreForBehavior(behaviorId) {
    const initialValues = this.consumeInitialValues(behaviorId);
    return DatasourceStore(this, behaviorId, initialValues);
  }
  /**
   * Invoke a function on this store and the stores of descendant parts.
   */
  visit(fn) {
    fn(this);
    for (const child of Object.values(this.#childStores)) {
      child.visit(fn);
    }
  }
  childStoreFor(endpoint) {
    if (!endpoint.lifecycle.hasId) {
      throw new ImplementationError("Cannot access endpoint storage because endpoint has no assigned ID");
    }
    return this.storeForPartId(endpoint.id);
  }
  storeForPartId(partId) {
    let store = this.#childStores[partId];
    if (store === void 0) {
      store = this.#childStores[partId] = new ServerEndpointStore(this.#childStorage.createContext(partId));
    }
    return store;
  }
  async saveNumber() {
    await this.storage.set(NUMBER_KEY, this.number);
  }
  async load() {
    await super.load();
    const number = await this.storage.get(NUMBER_KEY, -1);
    if (number !== -1) {
      this.#number = number;
    }
    await this.#loadSubparts();
  }
  /**
   * Erase the child storage for one part
   */
  async eraseChildStoreFor(endpoint) {
    const partId = endpoint.id;
    const store = this.#childStores[partId];
    await store.erase();
    delete this.#childStores[partId];
  }
  async #loadSubparts() {
    const knownParts = await this.#childStorage.contexts();
    for (const partId of knownParts) {
      await this.#loadKnownChildStores(partId);
    }
  }
  async #loadKnownChildStores(partId) {
    const endpointStore = new ServerEndpointStore(this.#childStorage.createContext(partId));
    this.#childStores[partId] = endpointStore;
    await endpointStore.load();
  }
}
export {
  ServerEndpointStore
};
