// @matter/types/dist/esm/globals/Namespace.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Namespace = /* @__PURE__ */ ((Namespace2) => {
  Namespace2[Namespace2["Closure"] = 1] = "Closure";
  Namespace2[Namespace2["CompassDirection"] = 2] = "CompassDirection";
  Namespace2[Namespace2["CompassLocation"] = 3] = "CompassLocation";
  Namespace2[Namespace2["Direction"] = 4] = "Direction";
  Namespace2[Namespace2["Level"] = 5] = "Level";
  Namespace2[Namespace2["Location"] = 6] = "Location";
  Namespace2[Namespace2["Number"] = 7] = "Number";
  Namespace2[Namespace2["Position"] = 8] = "Position";
  Namespace2[Namespace2["ElectricalMeasurement"] = 10] = "ElectricalMeasurement";
  Namespace2[Namespace2["Laundry"] = 14] = "Laundry";
  Namespace2[Namespace2["PowerSource"] = 15] = "PowerSource";
  Namespace2[Namespace2["AreaNamespace"] = 16] = "AreaNamespace";
  Namespace2[Namespace2["LandmarkNamespace"] = 17] = "LandmarkNamespace";
  Namespace2[Namespace2["RelativePosition"] = 18] = "RelativePosition";
  Namespace2[Namespace2["Refrigerator"] = 65] = "Refrigerator";
  Namespace2[Namespace2["RoomAirConditioner"] = 66] = "RoomAirConditioner";
  Namespace2[Namespace2["Switches"] = 67] = "Switches";
  return Namespace2;
})(Namespace || {});
export {
  Namespace
};
