// @matter/model/dist/esm/models/DeviceTypeModel.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { DeviceClassification } from "../common/DeviceClassification.js";
import { DeviceTypeElement } from "../elements/index.js";
import { Model } from "./Model.js";
import { RequirementModel } from "./RequirementModel.js";
class DeviceTypeModel extends Model {
  tag = DeviceTypeElement.Tag;
  get requirements() {
    return this.all(RequirementModel);
  }
  get revision() {
    return this?.get(RequirementModel, "Descriptor")?.get(RequirementModel, "DeviceTypeList")?.default[0].revision ?? 1;
  }
  get classification() {
    return this.hasLocalResource ? this.localResource.classification : DeviceClassification.Simple;
  }
  set classification(classification) {
    if (classification || this.hasLocalResource) {
      this.localResource.classification = classification;
    }
  }
  constructor(definition, ...children) {
    super(definition, ...children);
    if (!(definition instanceof Model)) {
      this.classification = definition.classification;
    }
  }
  toElement(omitResources = false, extra) {
    if (omitResources) {
      return super.toElement(omitResources, extra);
    }
    return super.toElement(omitResources, {
      classification: this.classification,
      ...extra
    });
  }
  get id() {
    return super.id;
  }
  set id(id) {
    super.id = id;
  }
  static Tag = DeviceTypeElement.Tag;
}
DeviceTypeModel.register();
export {
  DeviceTypeModel
};
