// @matter/protocol/dist/esm/fabric/TestFabric.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { CertificateAuthority } from "#certificate/CertificateAuthority.js";
import { ImplementationError, MockCrypto } from "#general";
import { FabricId, FabricIndex, VendorId } from "#types";
import { FabricAuthority } from "./FabricAuthority.js";
import { FabricManager } from "./FabricManager.js";
async function TestFabric(options = {}) {
  const authority = await TestFabric.Authority(options);
  return authority.createFabric();
}
((TestFabric2) => {
  async function Authority({ index, fabrics } = {}) {
    if (index === void 0) {
      if (fabrics) {
        index = fabrics.allocateFabricIndex();
      } else {
        index = 1;
      }
    }
    if (index < 1 || index > 254) {
      throw new ImplementationError("Test fabric indexes must be in the range 1-254");
    }
    if (!fabrics) {
      fabrics = new FabricManager(MockCrypto(index));
    }
    const authority = new FabricAuthority({
      ca: await CertificateAuthority.create(fabrics.crypto),
      config: {
        adminFabricLabel: `mock-fabric-${index}`,
        adminVendorId: VendorId(65521),
        fabricIndex: FabricIndex(index),
        fabricId: FabricId(1)
      },
      fabrics
    });
    return authority;
  }
  TestFabric2.Authority = Authority;
})(TestFabric || (TestFabric = {}));
export {
  TestFabric
};
