// @matter/protocol/dist/esm/action/client/ClientSubscriptionHandler.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { DecodedDataReport } from "#interaction/DecodedDataReport.js";
import { IncomingInteractionClientMessenger } from "#interaction/InteractionMessenger.js";
import { INTERACTION_PROTOCOL_ID, Status } from "#types";
import { Diagnostic, InternalError, Logger } from "@matter/general";
import { InputChunk } from "./InputChunk.js";
const logger = Logger.get("ClientSubscriptionHandler");
class ClientSubscriptionHandler {
  id = INTERACTION_PROTOCOL_ID;
  requiresSecureSession = true;
  #subscriptions;
  constructor(subscriptions) {
    this.#subscriptions = subscriptions;
  }
  async onNewExchange(exchange) {
    const messenger = new IncomingInteractionClientMessenger(exchange);
    const reports = messenger.readDataReports();
    const initialIteration = await reports.next();
    if (initialIteration.done) {
      throw new InternalError("Exchange initiated with no initial message");
    }
    const initialReport = initialIteration.value;
    const { subscriptionId } = initialReport;
    if (subscriptionId === void 0) {
      logger.debug("Ignoring unsolicited data report with no subscription ID");
      await sendInvalid(messenger, void 0);
      return;
    }
    const subscription = this.#subscriptions.get(subscriptionId);
    if (subscription === void 0) {
      logger.debug("Ignoring data report for unknown subscription ID", Diagnostic.strong(subscriptionId));
      await sendInvalid(messenger, subscriptionId);
      return;
    }
    if (!initialReport.attributeReports?.length && !initialReport.eventReports?.length) {
      subscription.timeoutAtMs = void 0;
      this.#subscriptions.resetTimer();
      await exchange.close();
      return;
    }
    try {
      subscription.isReading = true;
      if (subscription.request.updated) {
        await subscription.request.updated(processReports(initialReport, reports, messenger));
      } else {
        for await (const _chunk of reports) ;
      }
    } finally {
      subscription.isReading = false;
      subscription.timeoutAtMs = void 0;
      this.#subscriptions.resetTimer();
      await exchange.close();
    }
  }
  async close() {
  }
}
async function sendInvalid(messenger, subscriptionId) {
  await messenger.sendStatus(Status.InvalidSubscription, {
    multipleMessageInteraction: true,
    logContext: {
      subId: subscriptionId
    }
  });
  await messenger.close();
}
async function* processReports(initialReport, otherReports, messenger) {
  yield InputChunk(initialReport);
  const { subscriptionId } = initialReport;
  for await (const report of otherReports) {
    const decoded = DecodedDataReport(report);
    if (decoded.subscriptionId === void 0) {
      logger.debug(
        "Ignoring data report for incorrect subscription id",
        Diagnostic.strong(decoded.subscriptionId)
      );
      await sendInvalid(messenger, decoded.subscriptionId);
    }
    if (decoded.subscriptionId !== subscriptionId) {
      logger.debug(
        "Ignoring data report for incorrect subscription id",
        Diagnostic.strong(decoded.subscriptionId)
      );
      await sendInvalid(messenger, decoded.subscriptionId);
      continue;
    }
    yield InputChunk(report);
  }
}
export {
  ClientSubscriptionHandler
};
