// @matter/protocol/dist/esm/action/client/ClientSubscriptions.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  BasicSet,
  CanceledError,
  Diagnostic,
  Environmental,
  Logger,
  Time,
  TimeoutError
} from "#general";
const logger = Logger.get("ClientSubscriptions");
class ClientSubscriptions {
  #subscriptions = new BasicSet();
  #timeout;
  #nextTimeoutAt;
  static [Environmental.create](env) {
    const instance = new ClientSubscriptions();
    env.set(ClientSubscriptions, instance);
    return instance;
  }
  /**
   * Register an active subscription.
   */
  add(request, response) {
    const subscription = {
      ...response,
      request,
      close: () => this.#closeOne(subscription, new CanceledError()),
      timeoutAtMs: void 0,
      isClosed: false,
      isReading: true
    };
    this.#subscriptions.add(subscription);
    this.resetTimer();
    return subscription;
  }
  /**
   * Retrieve a subscription by ID.
   */
  get(id) {
    return this.#subscriptions.get("subscriptionId", id);
  }
  /**
   * Iterate over active subscriptions.
   */
  [Symbol.iterator]() {
    return this.#subscriptions[Symbol.iterator]();
  }
  /**
   * Terminate all subscriptions.
   */
  async close() {
    if (this.#timeout) {
      this.#timeout.stop();
      this.#timeout = void 0;
    }
    for (const subscription of this.#subscriptions) {
      subscription.close();
    }
  }
  /**
   * Restart the timeout timer for the current set of active subscriptions.
   */
  resetTimer() {
    const now = Time.nowMs();
    let nextTimeoutAt;
    for (const subscription of this.#subscriptions) {
      if (subscription.isReading) {
        continue;
      }
      let { timeoutAtMs } = subscription;
      if (timeoutAtMs === void 0) {
        timeoutAtMs = subscription.timeoutAtMs = timeoutFor(subscription);
      } else if (timeoutAtMs < now) {
        this.#timeOut(subscription);
        continue;
      }
      if (nextTimeoutAt === void 0 || nextTimeoutAt > timeoutAtMs) {
        nextTimeoutAt = timeoutAtMs;
      }
    }
    if (nextTimeoutAt === void 0) {
      this.#nextTimeoutAt = void 0;
      this.#timeout?.stop();
      return;
    }
    if (nextTimeoutAt !== this.#nextTimeoutAt) {
      this.#nextTimeoutAt = nextTimeoutAt;
      if (this.#timeout) {
        this.#timeout?.stop();
        this.#timeout.intervalMs = nextTimeoutAt - now;
      } else {
        this.#timeout = Time.getTimer("SubscriptionTimeout", nextTimeoutAt - now, this.resetTimer.bind(this));
      }
    }
  }
  #closeOne(subscription, cause) {
    if (subscription.isClosed) {
      return;
    }
    subscription.isClosed = true;
    this.#subscriptions.delete(subscription);
    try {
      subscription.request.closed?.(cause);
    } catch (e) {
      logger.error("Error canceling subscription", Diagnostic.strong(subscription.subscriptionId), e);
    }
  }
  #timeOut(registration) {
    logger.info(
      "Subscription",
      Diagnostic.strong(registration.subscriptionId),
      "timed out after",
      Diagnostic.strong(`${timeoutFor(registration)}ms`)
    );
    this.#closeOne(registration, new TimeoutError());
  }
}
function timeoutFor(subscription) {
  return subscription.maxInterval * 1e3 + (subscription.request.maxPeerResponseTime ?? 0);
}
export {
  ClientSubscriptions
};
