// @matter/general/dist/cjs/util/Set.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Set_exports = {};
__export(Set_exports, {
  BasicSet: () => BasicSet,
  MapOfIndexedSet: () => MapOfIndexedSet
});
module.exports = __toCommonJS(Set_exports);
var import_MatterError = require("#MatterError.js");
var import_Observable = require("./Observable.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class BasicSet {
  #entries = /* @__PURE__ */ new Set();
  #added;
  #deleted;
  #indices;
  #maps;
  constructor(...initialItems) {
    for (const item of initialItems) {
      this.add(item);
    }
  }
  [Symbol.iterator]() {
    return this.#entries[Symbol.iterator]();
  }
  get size() {
    return this.#entries.size;
  }
  map(mapper) {
    return [...this].map(mapper);
  }
  find(predicate) {
    for (const item of this) {
      if (predicate(item)) {
        return item;
      }
    }
  }
  filter(predicate) {
    const result = new Array();
    for (const item of this) {
      if (predicate(item)) {
        result.push(item);
      }
    }
    return result;
  }
  has(item) {
    return this.#entries.has(item);
  }
  add(item) {
    const created = this.create(item);
    if (this.#entries.has(item)) {
      return;
    }
    this.#entries.add(item);
    if (this.#indices) {
      for (const field in this.#indices) {
        const value = created[field];
        if (value === void 0) {
          continue;
        }
        const index = this.#indices[field];
        if (index === void 0 || index.has(value)) {
          continue;
        }
        index.set(value, created);
      }
    }
    this.#added?.emit(created);
  }
  get(field, value) {
    return this.#indexOf(field).get(value);
  }
  #indexOf(field) {
    if (!this.#indices) {
      this.#indices = {};
    }
    let index = this.#indices[field];
    if (index === void 0) {
      index = /* @__PURE__ */ new Map();
      for (const item of this) {
        const value = item[field];
        if (value === void 0 || index.has(value)) {
          continue;
        }
        index.set(value, item);
      }
      this.#indices[field] = index;
    }
    return index;
  }
  /**
   * Obtain key/value map using specific field as key.
   *
   * Note we use {@link This} to constrain usage to sets where {@link T} === {@link AddT} as required by {@link Map}.
   */
  mapOf(field) {
    if (!this.#maps) {
      this.#maps = {};
    }
    let map = this.#maps[field];
    if (map === void 0) {
      map = new MapOfIndexedSet(this, field, this.#indexOf(field));
      this.#maps[field] = map;
    }
    return map;
  }
  delete(itemOrField, value) {
    let item;
    if (value === void 0) {
      item = itemOrField;
    } else {
      item = this.get(itemOrField, value);
      if (item === void 0) {
        return false;
      }
    }
    if (!this.#entries.delete(item)) {
      return false;
    }
    if (this.#indices) {
      for (const field in this.#indices) {
        const value2 = item[field];
        if (value2 === void 0) {
          continue;
        }
        const index = this.#indices[field];
        if (index !== void 0 && index.get(value2) === item) {
          index.delete(value2);
        }
      }
    }
    this.#deleted?.emit(item);
    return true;
  }
  clear() {
    this.#entries.clear();
  }
  get added() {
    if (this.#added === void 0) {
      this.#added = (0, import_Observable.Observable)();
    }
    return this.#added;
  }
  get deleted() {
    if (this.#deleted === void 0) {
      this.#deleted = (0, import_Observable.Observable)();
    }
    return this.#deleted;
  }
  create(definition) {
    return definition;
  }
}
class MapOfIndexedSet {
  #set;
  #key;
  // This is an optimization for lookup
  #index;
  /**
   * Create a new map.
   *
   * @param set the backing data
   * @param key a property of {@link T} used as the key
   * @param index optional index that optimizes lookup by bypassing {@link IndexedSet#get}
   */
  constructor(set, key, index) {
    this.#set = set;
    this.#key = key;
    this.#index = index;
  }
  clear() {
    this.#set.clear();
  }
  delete(key) {
    const item = this.get(key);
    if (item) {
      return this.#set.delete(item);
    }
    return false;
  }
  forEach(callbackfn, thisArg) {
    if (thisArg) {
      callbackfn = callbackfn.bind(thisArg);
    }
    for (const [k, v] of this) {
      callbackfn(v, k, this);
    }
  }
  get(key) {
    if (this.#index) {
      return this.#index.get(key);
    }
    return this.#set.get(this.#key, key);
  }
  has(key) {
    return this.#index ? this.#index.has(key) : this.#set.get(this.#key, key) !== void 0;
  }
  set(key, value) {
    if (value[this.#key] !== key) {
      throw new MapOfIndexedSet.KeyValueMismatchError(
        `Cannot set key "${key}" because value property ${String(this.#key)} is "${value[this.#key]}"`
      );
    }
    if (this.has(key)) {
      if (this.get(key) === value) {
        return this;
      }
      this.#set.delete(this.#index ? this.#index.get(key) : this.#set.get(this.#key, key));
    }
    this.#set.add(value);
    return this;
  }
  get size() {
    return this.#set.size;
  }
  entries() {
    return this[Symbol.iterator]();
  }
  keys() {
    if (this.#index) {
      return this.#index.keys();
    }
    const keys = [...this.#set].map((item) => item[this.#key]).filter((key) => key !== void 0);
    return keys[Symbol.iterator]();
  }
  values() {
    if (this.#index) {
      return this.#index.values();
    }
    const values = [...this.#set].map((item) => item);
    return values[Symbol.iterator]();
  }
  *[Symbol.iterator]() {
    for (const item of this.#set) {
      const k = item[this.#key];
      if (k === void 0) {
        continue;
      }
      yield [k, item];
    }
  }
  [Symbol.toStringTag] = "Map";
}
((MapOfIndexedSet2) => {
  class KeyValueMismatchError extends import_MatterError.ImplementationError {
  }
  MapOfIndexedSet2.KeyValueMismatchError = KeyValueMismatchError;
})(MapOfIndexedSet || (MapOfIndexedSet = {}));
