// @matter/types/dist/esm/clusters/oven-mode.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MutableCluster } from "../cluster/mutation/MutableCluster.js";
import { BitFlag } from "../schema/BitmapSchema.js";
import { FixedAttribute, Attribute, Command } from "../cluster/Cluster.js";
import { TlvArray } from "../tlv/TlvArray.js";
import { TlvField, TlvOptionalField, TlvObject } from "../tlv/TlvObject.js";
import { TlvString } from "../tlv/TlvString.js";
import { TlvUInt8, TlvEnum } from "../tlv/TlvNumber.js";
import { TlvVendorId } from "../datatype/VendorId.js";
import { ModeBase } from "./mode-base.js";
import { ClusterRegistry } from "../cluster/ClusterRegistry.js";
var OvenMode;
((OvenMode2) => {
  let Feature;
  ((Feature2) => {
    Feature2["OnOff"] = "OnOff";
  })(Feature = OvenMode2.Feature || (OvenMode2.Feature = {}));
  let ModeTag;
  ((ModeTag2) => {
    ModeTag2[ModeTag2["Auto"] = 0] = "Auto";
    ModeTag2[ModeTag2["Quick"] = 1] = "Quick";
    ModeTag2[ModeTag2["Quiet"] = 2] = "Quiet";
    ModeTag2[ModeTag2["LowNoise"] = 3] = "LowNoise";
    ModeTag2[ModeTag2["LowEnergy"] = 4] = "LowEnergy";
    ModeTag2[ModeTag2["Vacation"] = 5] = "Vacation";
    ModeTag2[ModeTag2["Min"] = 6] = "Min";
    ModeTag2[ModeTag2["Max"] = 7] = "Max";
    ModeTag2[ModeTag2["Night"] = 8] = "Night";
    ModeTag2[ModeTag2["Day"] = 9] = "Day";
    ModeTag2[ModeTag2["Bake"] = 16384] = "Bake";
    ModeTag2[ModeTag2["Convection"] = 16385] = "Convection";
    ModeTag2[ModeTag2["Grill"] = 16386] = "Grill";
    ModeTag2[ModeTag2["Roast"] = 16387] = "Roast";
    ModeTag2[ModeTag2["Clean"] = 16388] = "Clean";
    ModeTag2[ModeTag2["ConvectionBake"] = 16389] = "ConvectionBake";
    ModeTag2[ModeTag2["ConvectionRoast"] = 16390] = "ConvectionRoast";
    ModeTag2[ModeTag2["Warming"] = 16391] = "Warming";
    ModeTag2[ModeTag2["Proofing"] = 16392] = "Proofing";
    ModeTag2[ModeTag2["Steam"] = 16393] = "Steam";
  })(ModeTag = OvenMode2.ModeTag || (OvenMode2.ModeTag = {}));
  OvenMode2.TlvModeTagStruct = TlvObject({
    /**
     * If the MfgCode field exists, the Value field shall be in the manufacturer-specific value range (see Section
     * 1.10.8, “Mode Namespace”).
     *
     * This field shall indicate the manufacturer’s VendorID and it shall determine the meaning of the Value field.
     *
     * The same manufacturer code and mode tag value in separate cluster instances are part of the same namespace
     * and have the same meaning. For example: a manufacturer tag meaning "pinch" can be used both in a cluster
     * whose purpose is to choose the amount of sugar, or in a cluster whose purpose is to choose the amount of
     * salt.
     *
     * @see {@link MatterSpecification.v141.Cluster} § 1.10.5.1.1
     */
    mfgCode: TlvOptionalField(0, TlvVendorId),
    /**
     * This field shall indicate the mode tag within a mode tag namespace which is either manufacturer specific or
     * standard.
     *
     * @see {@link MatterSpecification.v141.Cluster} § 1.10.5.1.2
     */
    value: TlvField(1, TlvEnum())
  });
  OvenMode2.TlvModeOption = TlvObject({
    label: TlvField(0, TlvString.bound({ maxLength: 64 })),
    mode: TlvField(1, TlvUInt8),
    modeTags: TlvField(2, TlvArray(OvenMode2.TlvModeTagStruct, { minLength: 1, maxLength: 8 }))
  });
  OvenMode2.Base = MutableCluster.Component({
    id: 73,
    name: "OvenMode",
    revision: 2,
    features: {
      /**
       * Dependency with the OnOff cluster
       */
      onOff: BitFlag(0)
    },
    attributes: {
      /**
       * At least one entry in the SupportedModes attribute shall include the Bake mode tag in the ModeTags field
       * list.
       *
       * @see {@link MatterSpecification.v141.Cluster} § 8.11.6.1
       */
      supportedModes: FixedAttribute(
        0,
        TlvArray(OvenMode2.TlvModeOption, { minLength: 2, maxLength: 255 }),
        { default: [] }
      ),
      /**
       * @see {@link MatterSpecification.v141.Cluster} § 8.11.6
       */
      currentMode: Attribute(1, TlvUInt8, { persistent: true })
    },
    commands: {
      /**
       * This command is used to change device modes.
       *
       * On receipt of this command the device shall respond with a ChangeToModeResponse command.
       *
       * @see {@link MatterSpecification.v141.Cluster} § 1.10.7.1
       */
      changeToMode: Command(0, ModeBase.TlvChangeToModeRequest, 1, ModeBase.TlvChangeToModeResponse)
    },
    /**
     * This metadata controls which OvenModeCluster elements matter.js activates for specific feature combinations.
     */
    extensions: MutableCluster.Extensions({ flags: { onOff: true }, component: false })
  });
  OvenMode2.ClusterInstance = MutableCluster(OvenMode2.Base);
  OvenMode2.Cluster = OvenMode2.ClusterInstance;
  OvenMode2.Complete = OvenMode2.Cluster;
})(OvenMode || (OvenMode = {}));
const OvenModeCluster = OvenMode.Cluster;
ClusterRegistry.register(OvenMode.Complete);
export {
  OvenMode,
  OvenModeCluster
};
