// @matter/model/dist/cjs/models/DeviceTypeModel.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DeviceTypeModel_exports = {};
__export(DeviceTypeModel_exports, {
  DeviceTypeModel: () => DeviceTypeModel
});
module.exports = __toCommonJS(DeviceTypeModel_exports);
var import_DeviceClassification = require("../common/DeviceClassification.js");
var import_elements = require("../elements/index.js");
var import_Model = require("./Model.js");
var import_RequirementModel = require("./RequirementModel.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class DeviceTypeModel extends import_Model.Model {
  tag = import_elements.DeviceTypeElement.Tag;
  get requirements() {
    return this.all(import_RequirementModel.RequirementModel);
  }
  get revision() {
    return this?.get(import_RequirementModel.RequirementModel, "Descriptor")?.get(import_RequirementModel.RequirementModel, "DeviceTypeList")?.default[0].revision ?? 1;
  }
  get classification() {
    return this.hasLocalResource ? this.localResource.classification : import_DeviceClassification.DeviceClassification.Simple;
  }
  set classification(classification) {
    if (classification || this.hasLocalResource) {
      this.localResource.classification = classification;
    }
  }
  constructor(definition, ...children) {
    super(definition, ...children);
    if (!(definition instanceof import_Model.Model)) {
      this.classification = definition.classification;
    }
  }
  toElement(omitResources = false, extra) {
    if (omitResources) {
      return super.toElement(omitResources, extra);
    }
    return super.toElement(omitResources, {
      classification: this.classification,
      ...extra
    });
  }
  get id() {
    return super.id;
  }
  set id(id) {
    super.id = id;
  }
  static Tag = import_elements.DeviceTypeElement.Tag;
}
DeviceTypeModel.register();
