// @matter/node/dist/cjs/node/client/ClientNodeInteraction.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClientNodeInteraction_exports = {};
__export(ClientNodeInteraction_exports, {
  ClientNodeInteraction: () => ClientNodeInteraction
});
module.exports = __toCommonJS(ClientNodeInteraction_exports);
var import_EndpointInitializer = require("#endpoint/properties/EndpointInitializer.js");
var import_protocol = require("#protocol");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class ClientNodeInteraction {
  #node;
  constructor(node) {
    this.#node = node;
  }
  async *read(request, context) {
    request = this.structure.injectVersionFilters(request);
    const interaction = await this.#connect();
    const response = interaction.read(request, context);
    yield* this.structure.mutate(request, response);
  }
  async subscribe(request, context) {
    const intermediateRequest = {
      ...this.structure.injectVersionFilters(request),
      updated: async (data) => {
        const result = this.structure.mutate(request, data);
        if (request.updated) {
          await request.updated(result);
        } else {
          for await (const _chunk of result) ;
        }
      },
      closed(cause) {
        request.closed?.(cause);
      }
    };
    const interaction = await this.#connect();
    return interaction.subscribe(intermediateRequest, context);
  }
  async write(request, context) {
    return (await this.#connect()).write(request, context);
  }
  async *invoke(request, context) {
    yield* (await this.#connect()).invoke(request, context);
  }
  async #connect() {
    if (!this.#node.lifecycle.isOnline) {
      await this.#node.start();
    }
    return this.#node.env.get(import_protocol.ClientInteraction);
  }
  get structure() {
    return this.#node.env.get(import_EndpointInitializer.EndpointInitializer).structure;
  }
}
