// @matter/protocol/dist/esm/action/server/AttributeSubscriptionResponse.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { InternalError } from "#general";
import { AttributeReadResponse } from "./AttributeReadResponse.js";
class AttributeSubscriptionResponse extends AttributeReadResponse {
  #filter;
  #currentEndpointFilter;
  #currentClusterFilter;
  constructor(node, session, filter) {
    super(node, session);
    this.#filter = filter;
  }
  get filter() {
    return this.#filter;
  }
  /** Guarded accessor for this.#currentEndpointFilter.  This should never be undefined */
  get currentEndpointFilter() {
    if (!this.#currentEndpointFilter) {
      throw new InternalError("currentEndpointFilter is not set. Should never happen");
    }
    return this.#currentEndpointFilter;
  }
  /** Guarded accessor for this.#currentCLusterFilter.  This should never be undefined */
  get currentClusterFilter() {
    if (!this.#currentClusterFilter) {
      throw new InternalError("currentClusterFilter is not set. Should never happen");
    }
    return this.#currentClusterFilter;
  }
  addConcrete(path) {
    const { endpointId, clusterId, attributeId } = path;
    if (this.#filter[endpointId]?.[clusterId]?.has(attributeId) === void 0) {
      return;
    }
    super.addConcrete(path);
  }
  *readEndpointForWildcard(endpoint, path) {
    this.#currentEndpointFilter = this.#filter[endpoint.id];
    if (this.#currentEndpointFilter === void 0) {
      return;
    }
    yield* super.readEndpointForWildcard(endpoint, path);
  }
  readClusterForWildcard(cluster, path) {
    this.#currentClusterFilter = this.currentEndpointFilter[cluster.type.id];
    if (this.#currentClusterFilter === void 0) {
      return;
    }
    super.readClusterForWildcard(cluster, path);
  }
  readAttributeForWildcard(attribute, path) {
    if (!this.currentClusterFilter.has(attribute.id)) {
      return;
    }
    super.readAttributeForWildcard(attribute, path);
  }
  addStatus() {
  }
}
export {
  AttributeSubscriptionResponse
};
