// @matter/node/dist/cjs/node/ClientNode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClientNode_exports = {};
__export(ClientNode_exports, {
  ClientNode: () => ClientNode
});
module.exports = __toCommonJS(ClientNode_exports);
var import_CommissioningClient = require("#behavior/system/commissioning/CommissioningClient.js");
var import_ClientNetworkRuntime = require("#behavior/system/network/ClientNetworkRuntime.js");
var import_NetworkClient = require("#behavior/system/network/NetworkClient.js");
var import_EndpointInitializer = require("#endpoint/properties/EndpointInitializer.js");
var import_general = require("#general");
var import_ClientNodeStore = require("#storage/client/ClientNodeStore.js");
var import_RemoteWriter = require("#storage/client/RemoteWriter.js");
var import_ServerNodeStore = require("#storage/server/ServerNodeStore.js");
var import_model = require("@matter/model");
var import_ClientEndpointInitializer = require("./client/ClientEndpointInitializer.js");
var import_ClientNodeInteraction = require("./client/ClientNodeInteraction.js");
var import_Node = require("./Node.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class ClientNode extends import_Node.Node {
  #matter;
  #interaction;
  constructor(options) {
    const opts = {
      ...options,
      number: 0,
      type: ClientNode.RootEndpoint
    };
    super(opts);
    this.env.set(import_Node.Node, this);
    this.env.set(ClientNode, this);
    this.#matter = options.matter ?? import_model.Matter;
  }
  /**
   * Model of Matter semantics understood by this node.
   *
   * Matter elements missing from this model will not support all functionality.
   */
  get matter() {
    return this.#matter;
  }
  initialize() {
    const store = this.env.get(import_ServerNodeStore.ServerNodeStore).clientStores.storeForNode(this);
    this.env.set(import_ClientNodeStore.ClientNodeStore, store);
    const initializer = new import_ClientEndpointInitializer.ClientEndpointInitializer(this);
    this.env.set(import_EndpointInitializer.EndpointInitializer, initializer);
    store.write = (0, import_RemoteWriter.RemoteWriter)(this, initializer.structure);
    initializer.structure.loadCache();
    return super.initialize();
  }
  get owner() {
    return super.owner;
  }
  set owner(owner) {
    super.owner = owner;
  }
  /**
   * Add this node to a fabric.
   */
  async commission(options) {
    await this.act("commission", (agent) => agent.commissioning.commission(options));
  }
  /**
   * Remove this node from the fabric (if commissioned) and locally.
   */
  async delete() {
    if (this.lifecycle.isCommissioned) {
      this.statusUpdate("decommissioning");
      await this.act("decommission", (agent) => agent.commissioning.decommission());
    }
    await super.delete();
  }
  /**
   * Force-remove the node without first decommissioning.
   *
   * If the node is still available you should use {@link delete} to remove it from the fabric.
   */
  async erase() {
    await this.lifecycle.mutex.produce(this.eraseWithMutex.bind(this));
  }
  async eraseWithMutex() {
    await this.cancelWithMutex();
    await this.resetWithMutex();
  }
  createRuntime() {
    return new import_ClientNetworkRuntime.ClientNetworkRuntime(this);
  }
  async prepareRuntimeShutdown() {
  }
  get container() {
    return this.owner?.nodes;
  }
  act(actorOrPurpose, actor) {
    if (this.construction.status === import_general.Lifecycle.Status.Inactive) {
      this.construction.start();
    }
    if (this.construction.status === import_general.Lifecycle.Status.Initializing) {
      return this.construction.then(() => super.act(actorOrPurpose, actor));
    }
    return super.act(actorOrPurpose, actor);
  }
  get interaction() {
    if (this.#interaction === void 0) {
      this.#interaction = new import_ClientNodeInteraction.ClientNodeInteraction(this);
    }
    return this.#interaction;
  }
}
((ClientNode2) => {
  ClientNode2.RootEndpoint = import_Node.Node.CommonRootEndpoint.with(import_CommissioningClient.CommissioningClient, import_NetworkClient.NetworkClient);
})(ClientNode || (ClientNode = {}));
