// @matter/node/dist/cjs/storage/client/RemoteWriteParticipant.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RemoteWriteParticipant_exports = {};
__export(RemoteWriteParticipant_exports, {
  RemoteWriteParticipant: () => RemoteWriteParticipant
});
module.exports = __toCommonJS(RemoteWriteParticipant_exports);
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class RemoteWriteParticipant {
  #request = [];
  #store;
  /**
   * There is one participant for each transaction/client node pair.  We therefore use the store as the role.
   */
  get role() {
    return this.#store;
  }
  /**
   * Add an attribute update to the write request.
   */
  set(endpointNumber, behaviorId, values) {
    this.#request.push({
      number: endpointNumber,
      behaviorId,
      values
    });
  }
  async commit2() {
    if (!this.#request.length) {
      return;
    }
    const request = this.#request;
    this.#request = [];
    await this.#store.write(request);
  }
  rollback() {
    this.#request = [];
  }
  toString() {
    return `writer#${this.#store.id}`;
  }
  constructor(store) {
    this.#store = store;
  }
}
