// @matter/protocol/dist/cjs/groups/Groups.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Groups_exports = {};
__export(Groups_exports, {
  Groups: () => Groups
});
module.exports = __toCommonJS(Groups_exports);
var import_Subject = require("#action/server/Subject.js");
var import_general = require("#general");
var import_types = require("#types");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class Groups {
  #fabric;
  #keySets;
  /** Operational variant of the group key map attribute from Group Key Management cluster, maps group Ids to key sets. */
  #groupKeyIdMap = new import_general.BasicMap();
  /** Operational variant of the group table, maps group Ids to a list of enabled endpoints. */
  endpointMap = /* @__PURE__ */ new Map();
  constructor(fabric, keySets) {
    this.#fabric = fabric;
    this.#keySets = keySets;
  }
  /** Operative lookup of the group key sets by a group id and to react on added removed groups. */
  get idMap() {
    return this.#groupKeyIdMap;
  }
  /** Updates the group key id map when changed in Group Key Management Cluster. Only changes are taken over. */
  set idMap(map) {
    for (const [groupId, keySetId] of map.entries()) {
      this.#groupKeyIdMap.set(groupId, keySetId);
    }
    for (const groupId of this.#groupKeyIdMap.keys()) {
      if (!map.has(groupId)) {
        this.#groupKeyIdMap.delete(groupId);
      }
    }
  }
  subjectForGroup(id, keySetId) {
    return import_Subject.Subject.Group({
      id,
      hasValidMapping: this.idMap.get(id) === keySetId,
      endpoints: this.endpointMap.get(id) ?? []
    });
  }
  /** Returns the multicast address for a given group id for this fabric. */
  multicastAddress(groupId) {
    import_types.GroupId.assertGroupId(groupId);
    const writer = new import_general.DataWriter();
    writer.writeUInt16(65333);
    writer.writeUInt16(64);
    writer.writeUInt8(253);
    writer.writeUInt64(this.#fabric.fabricId);
    writer.writeUInt8(0);
    writer.writeUInt16((0, import_types.GroupId)(groupId));
    return (0, import_general.ipv6BytesToString)(writer.toByteArray());
  }
  /**
   * Returns the current operational group key for a given group id and returns the keys, start time and
   * their session IDs.
   */
  currentKeyForId(groupId) {
    const keySetId = this.#groupKeyIdMap.get(groupId);
    if (keySetId === void 0) {
      throw new import_general.ImplementationError(`No group key set found for groupId ${groupId}.`);
    }
    return { ...this.#keySets.currentKeyForId(keySetId), keySetId };
  }
}
