// @matter/node/dist/esm/behaviors/groups/GroupsServer.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { GroupKeyManagementServer } from "#behaviors/group-key-management";
import { IdentifyBehavior } from "#behaviors/identify";
import { Groups } from "#clusters/groups";
import { RootEndpoint } from "#endpoints/root";
import { InternalError, Logger } from "#general";
import { AccessLevel } from "#model";
import {
  Command,
  StatusCode,
  StatusResponseError,
  TlvField,
  TlvGroupId,
  TlvNoResponse,
  TlvObject,
  TlvString
} from "#types";
import { GroupsBehavior } from "./GroupsBehavior.js";
const logger = Logger.get("GroupsServer");
Groups.Cluster.commands = {
  ...Groups.Cluster.commands,
  addGroup: Command(
    0,
    TlvObject({
      groupId: TlvField(0, TlvGroupId),
      groupName: TlvField(1, TlvString)
    }),
    0,
    Groups.TlvAddGroupResponse,
    { invokeAcl: AccessLevel.Manage }
  ),
  addGroupIfIdentifying: Command(
    5,
    TlvObject({
      groupId: TlvField(0, TlvGroupId),
      groupName: TlvField(1, TlvString)
    }),
    5,
    TlvNoResponse,
    { invokeAcl: AccessLevel.Manage }
  )
};
const GroupsBase = GroupsBehavior.with(Groups.Feature.GroupNames);
class GroupsServer extends GroupsBase {
  initialize() {
    this.state.nameSupport.groupNames = this.features.groupNames;
  }
  // We need to search the root here ourselves because we cannot include ServerNode because else we generate a
  // circular dependency
  #rootEndpoint() {
    const rootEndpoint = this.endpoint.ownerOfType(RootEndpoint);
    if (rootEndpoint === void 0) {
      throw new InternalError("RootEndpoint not found");
    }
    return rootEndpoint;
  }
  #actOnGroupKeyManagement(act) {
    return this.#rootEndpoint().act((agent) => act(agent.get(GroupKeyManagementServer)));
  }
  async addGroup({ groupId, groupName }) {
    const fabric = this.session.associatedFabric;
    if (groupId < 1) {
      return { status: StatusCode.ConstraintError, groupId };
    }
    if (groupName.length > 16) {
      return { status: StatusCode.ConstraintError, groupId };
    }
    if (!fabric.groups.groupKeyIdMap.has(groupId)) {
      return { status: StatusCode.UnsupportedAccess, groupId };
    }
    const endpointNumber = this.endpoint.number;
    try {
      await this.#actOnGroupKeyManagement(
        (gkm) => gkm.addEndpointForGroup(fabric, groupId, endpointNumber, groupName)
      );
    } catch (error) {
      logger.error(error);
      StatusResponseError.accept(error);
      return { status: error.code, groupId };
    }
    return { status: StatusCode.Success, groupId };
  }
  viewGroup({ groupId }) {
    const fabric = this.session.associatedFabric;
    if (groupId < 1) {
      return { status: StatusCode.ConstraintError, groupId, groupName: "" };
    }
    const fabricIndex = fabric.fabricIndex;
    const endpointNumber = this.endpoint.number;
    const { groupTable } = this.#rootEndpoint().stateOf(GroupKeyManagementServer);
    const groupEntry = groupTable.find((entry) => entry.groupId === groupId && entry.fabricIndex === fabricIndex);
    if (groupEntry === void 0 || !groupEntry.endpoints.includes(endpointNumber)) {
      return { status: StatusCode.NotFound, groupId, groupName: "" };
    }
    return { status: StatusCode.Success, groupId, groupName: groupEntry.groupName ?? "" };
  }
  async getGroupMembership({
    groupList
  }) {
    const fabric = this.session.associatedFabric;
    const fabricIndex = fabric.fabricIndex;
    const endpointNumber = this.endpoint.number;
    const { groupTable } = this.#rootEndpoint().stateOf(GroupKeyManagementServer);
    const endpointGroups = groupTable.filter(
      (entry) => entry.endpoints.includes(endpointNumber) && entry.fabricIndex === fabricIndex
    );
    const fabricGroupsList = endpointGroups.map((entry) => entry.groupId);
    if (groupList.length === 0) {
      return { capacity: 254 - fabricGroupsList.length, groupList: fabricGroupsList };
    }
    const filteredGroupsList = groupList.filter((groupId) => fabricGroupsList.includes(groupId));
    return { capacity: 254 - fabricGroupsList.length, groupList: filteredGroupsList };
  }
  async removeGroup({ groupId }) {
    if (groupId < 1) {
      return { status: StatusCode.ConstraintError, groupId };
    }
    try {
      if (await this.#actOnGroupKeyManagement(
        (gkm) => gkm.removeEndpoint(this.session.associatedFabric, this.endpoint.number, groupId)
      )) {
        return { status: StatusCode.Success, groupId };
      }
      return { status: StatusCode.NotFound, groupId };
    } catch (error) {
      StatusResponseError.accept(error);
      return { status: error.code, groupId };
    }
  }
  // TODO ScenesManagement cluster is also affected by this command
  async removeAllGroups() {
    try {
      await this.#actOnGroupKeyManagement(
        (gkm) => gkm.removeEndpoint(this.session.associatedFabric, this.endpoint.number)
      );
    } catch (error) {
      StatusResponseError.accept(error);
      throw error;
    }
  }
  async addGroupIfIdentifying({ groupId, groupName }) {
    if (this.endpoint.stateOf(IdentifyBehavior).identifyTime > 0) {
      const { status } = await this.addGroup({ groupId, groupName });
      if (status !== StatusCode.Success) {
        throw new StatusResponseError(`Failed to add group ${groupId}`, status);
      }
    }
  }
}
export {
  GroupsServer
};
