// @matter/protocol/dist/cjs/certificate/kinds/definitions/asn.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var asn_exports = {};
__export(asn_exports, {
  FabricId_Matter: () => FabricId_Matter,
  FirmwareSigningId_Matter: () => FirmwareSigningId_Matter,
  IcacId_Matter: () => IcacId_Matter,
  NocCat_Matter: () => NocCat_Matter,
  NodeId_Matter: () => NodeId_Matter,
  ProductId_Matter: () => ProductId_Matter,
  RcacId_Matter: () => RcacId_Matter,
  VendorId_Matter: () => VendorId_Matter,
  jsToMatterDate: () => jsToMatterDate,
  matterToJsDate: () => matterToJsDate
});
module.exports = __toCommonJS(asn_exports);
var import_general = require("#general");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const YEAR_S = 365 * 24 * 60 * 60;
const EPOCH_OFFSET_S = 10957 * 24 * 60 * 60;
function matterToJsDate(date) {
  return date === 0 ? import_general.X520.NON_WELL_DEFINED_DATE : new Date((date + EPOCH_OFFSET_S) * 1e3);
}
function jsToMatterDate(date, addYears = 0) {
  return date.getTime() === import_general.X520.NON_WELL_DEFINED_DATE.getTime() ? 0 : Math.floor(date.getTime() / 1e3) - EPOCH_OFFSET_S + addYears * YEAR_S;
}
function intTo16Chars(value) {
  const byteArray = new Uint8Array(8);
  const dataView = import_general.Bytes.dataViewOf(byteArray);
  dataView.setBigUint64(0, typeof value === "bigint" ? value : BigInt(value));
  return import_general.Bytes.toHex(byteArray).toUpperCase();
}
function uInt16To8Chars(value) {
  const byteArray = new Uint8Array(4);
  const dataView = import_general.Bytes.dataViewOf(byteArray);
  dataView.setUint32(0, value);
  return import_general.Bytes.toHex(byteArray).toUpperCase();
}
function uInt16To4Chars(value) {
  const byteArray = new Uint8Array(2);
  const dataView = import_general.Bytes.dataViewOf(byteArray);
  dataView.setUint16(0, value);
  return import_general.Bytes.toHex(byteArray).toUpperCase();
}
const GenericMatterOpCertObject = (id, valueConverter) => (value) => [
  (0, import_general.DerObject)(`2b0601040182a27c01${id.toString(16).padStart(2, "0")}`, {
    value: (valueConverter ?? intTo16Chars)(value)
  })
];
const GenericMatterAttCertObject = (id, valueConverter) => (value) => [
  (0, import_general.DerObject)(`2b0601040182a27c02${id.toString(16).padStart(2, "0")}`, {
    value: (valueConverter ?? intTo16Chars)(value)
  })
];
const NodeId_Matter = GenericMatterOpCertObject(1);
const FirmwareSigningId_Matter = GenericMatterOpCertObject(2);
const IcacId_Matter = GenericMatterOpCertObject(3);
const RcacId_Matter = GenericMatterOpCertObject(4);
const FabricId_Matter = GenericMatterOpCertObject(5);
const NocCat_Matter = GenericMatterOpCertObject(6, uInt16To8Chars);
const VendorId_Matter = GenericMatterAttCertObject(1, uInt16To4Chars);
const ProductId_Matter = GenericMatterAttCertObject(2, uInt16To4Chars);
