// @matter/protocol/dist/cjs/certificate/kinds/definitions/base.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var base_exports = {};
__export(base_exports, {
  ExtensionKeyUsageBitmap: () => ExtensionKeyUsageBitmap,
  ExtensionKeyUsageSchema: () => ExtensionKeyUsageSchema
});
module.exports = __toCommonJS(base_exports);
var import_types = require("#types");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ExtensionKeyUsageBitmap = {
  digitalSignature: (0, import_types.BitFlag)(0),
  nonRepudiation: (0, import_types.BitFlag)(1),
  keyEncipherment: (0, import_types.BitFlag)(2),
  dataEncipherment: (0, import_types.BitFlag)(3),
  keyAgreement: (0, import_types.BitFlag)(4),
  keyCertSign: (0, import_types.BitFlag)(5),
  cRLSign: (0, import_types.BitFlag)(6),
  encipherOnly: (0, import_types.BitFlag)(7),
  decipherOnly: (0, import_types.BitFlag)(8)
};
const ExtensionKeyUsageSchema = (0, import_types.BitmapSchema)(ExtensionKeyUsageBitmap);
