// @matter/protocol/dist/cjs/certificate/kinds/definitions/certification-declaration.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var certification_declaration_exports = {};
__export(certification_declaration_exports, {
  CertificationDeclaration: () => CertificationDeclaration
});
module.exports = __toCommonJS(certification_declaration_exports);
var import_types = require("#types");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var CertificationDeclaration;
((CertificationDeclaration2) => {
  CertificationDeclaration2.TlvDc = (0, import_types.TlvObject)({
    formatVersion: (0, import_types.TlvField)(0, import_types.TlvUInt16),
    vendorId: (0, import_types.TlvField)(1, import_types.TlvVendorId),
    produceIdArray: (0, import_types.TlvField)(2, (0, import_types.TlvArray)(import_types.TlvUInt16, { minLength: 1, maxLength: 100 })),
    deviceTypeId: (0, import_types.TlvField)(3, import_types.TlvUInt32),
    certificateId: (0, import_types.TlvField)(4, import_types.TlvString.bound({ length: 19 })),
    securityLevel: (0, import_types.TlvField)(5, import_types.TlvUInt8),
    securityInformation: (0, import_types.TlvField)(6, import_types.TlvUInt16),
    versionNumber: (0, import_types.TlvField)(7, import_types.TlvUInt16),
    certificationType: (0, import_types.TlvField)(8, import_types.TlvUInt8),
    dacOriginVendorId: (0, import_types.TlvOptionalField)(9, import_types.TlvVendorId),
    dacOriginProductId: (0, import_types.TlvOptionalField)(10, import_types.TlvUInt16),
    authorizedPaaList: (0, import_types.TlvOptionalField)(
      11,
      (0, import_types.TlvArray)(import_types.TlvByteString.bound({ length: 20 }), { minLength: 1, maxLength: 10 })
    )
  });
})(CertificationDeclaration || (CertificationDeclaration = {}));
