// @matter/types/dist/cjs/clusters/oven-mode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var oven_mode_exports = {};
__export(oven_mode_exports, {
  OvenMode: () => OvenMode,
  OvenModeCluster: () => OvenModeCluster
});
module.exports = __toCommonJS(oven_mode_exports);
var import_MutableCluster = require("../cluster/mutation/MutableCluster.js");
var import_BitmapSchema = require("../schema/BitmapSchema.js");
var import_Cluster = require("../cluster/Cluster.js");
var import_TlvArray = require("../tlv/TlvArray.js");
var import_TlvObject = require("../tlv/TlvObject.js");
var import_TlvString = require("../tlv/TlvString.js");
var import_TlvNumber = require("../tlv/TlvNumber.js");
var import_VendorId = require("../datatype/VendorId.js");
var import_mode_base = require("./mode-base.js");
var import_ClusterRegistry = require("../cluster/ClusterRegistry.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var OvenMode;
((OvenMode2) => {
  let Feature;
  ((Feature2) => {
    Feature2["OnOff"] = "OnOff";
  })(Feature = OvenMode2.Feature || (OvenMode2.Feature = {}));
  let ModeTag;
  ((ModeTag2) => {
    ModeTag2[ModeTag2["Auto"] = 0] = "Auto";
    ModeTag2[ModeTag2["Quick"] = 1] = "Quick";
    ModeTag2[ModeTag2["Quiet"] = 2] = "Quiet";
    ModeTag2[ModeTag2["LowNoise"] = 3] = "LowNoise";
    ModeTag2[ModeTag2["LowEnergy"] = 4] = "LowEnergy";
    ModeTag2[ModeTag2["Vacation"] = 5] = "Vacation";
    ModeTag2[ModeTag2["Min"] = 6] = "Min";
    ModeTag2[ModeTag2["Max"] = 7] = "Max";
    ModeTag2[ModeTag2["Night"] = 8] = "Night";
    ModeTag2[ModeTag2["Day"] = 9] = "Day";
    ModeTag2[ModeTag2["Bake"] = 16384] = "Bake";
    ModeTag2[ModeTag2["Convection"] = 16385] = "Convection";
    ModeTag2[ModeTag2["Grill"] = 16386] = "Grill";
    ModeTag2[ModeTag2["Roast"] = 16387] = "Roast";
    ModeTag2[ModeTag2["Clean"] = 16388] = "Clean";
    ModeTag2[ModeTag2["ConvectionBake"] = 16389] = "ConvectionBake";
    ModeTag2[ModeTag2["ConvectionRoast"] = 16390] = "ConvectionRoast";
    ModeTag2[ModeTag2["Warming"] = 16391] = "Warming";
    ModeTag2[ModeTag2["Proofing"] = 16392] = "Proofing";
    ModeTag2[ModeTag2["Steam"] = 16393] = "Steam";
  })(ModeTag = OvenMode2.ModeTag || (OvenMode2.ModeTag = {}));
  OvenMode2.TlvModeTagStruct = (0, import_TlvObject.TlvObject)({
    /**
     * If the MfgCode field exists, the Value field shall be in the manufacturer-specific value range (see Section
     * 1.10.8, “Mode Namespace”).
     *
     * This field shall indicate the manufacturer’s VendorID and it shall determine the meaning of the Value field.
     *
     * The same manufacturer code and mode tag value in separate cluster instances are part of the same namespace
     * and have the same meaning. For example: a manufacturer tag meaning "pinch" can be used both in a cluster
     * whose purpose is to choose the amount of sugar, or in a cluster whose purpose is to choose the amount of
     * salt.
     *
     * @see {@link MatterSpecification.v141.Cluster} § 1.10.5.1.1
     */
    mfgCode: (0, import_TlvObject.TlvOptionalField)(0, import_VendorId.TlvVendorId),
    /**
     * This field shall indicate the mode tag within a mode tag namespace which is either manufacturer specific or
     * standard.
     *
     * @see {@link MatterSpecification.v141.Cluster} § 1.10.5.1.2
     */
    value: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvEnum)())
  });
  OvenMode2.TlvModeOption = (0, import_TlvObject.TlvObject)({
    label: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvString.bound({ maxLength: 64 })),
    mode: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8),
    modeTags: (0, import_TlvObject.TlvField)(2, (0, import_TlvArray.TlvArray)(OvenMode2.TlvModeTagStruct, { minLength: 1, maxLength: 8 }))
  });
  OvenMode2.Base = import_MutableCluster.MutableCluster.Component({
    id: 73,
    name: "OvenMode",
    revision: 2,
    features: {
      /**
       * Dependency with the OnOff cluster
       */
      onOff: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * At least one entry in the SupportedModes attribute shall include the Bake mode tag in the ModeTags field
       * list.
       *
       * @see {@link MatterSpecification.v141.Cluster} § 8.11.6.1
       */
      supportedModes: (0, import_Cluster.FixedAttribute)(
        0,
        (0, import_TlvArray.TlvArray)(OvenMode2.TlvModeOption, { minLength: 2, maxLength: 255 }),
        { default: [] }
      ),
      /**
       * @see {@link MatterSpecification.v141.Cluster} § 8.11.6
       */
      currentMode: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt8, { persistent: true })
    },
    commands: {
      /**
       * This command is used to change device modes.
       *
       * On receipt of this command the device shall respond with a ChangeToModeResponse command.
       *
       * @see {@link MatterSpecification.v141.Cluster} § 1.10.7.1
       */
      changeToMode: (0, import_Cluster.Command)(0, import_mode_base.ModeBase.TlvChangeToModeRequest, 1, import_mode_base.ModeBase.TlvChangeToModeResponse)
    },
    /**
     * This metadata controls which OvenModeCluster elements matter.js activates for specific feature combinations.
     */
    extensions: import_MutableCluster.MutableCluster.Extensions({ flags: { onOff: true }, component: false })
  });
  OvenMode2.ClusterInstance = (0, import_MutableCluster.MutableCluster)(OvenMode2.Base);
  OvenMode2.Cluster = OvenMode2.ClusterInstance;
  OvenMode2.Complete = OvenMode2.Cluster;
})(OvenMode || (OvenMode = {}));
const OvenModeCluster = OvenMode.Cluster;
import_ClusterRegistry.ClusterRegistry.register(OvenMode.Complete);
