// @matter/types/dist/cjs/protocol/definitions/secure-channel.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var secure_channel_exports = {};
__export(secure_channel_exports, {
  SECURE_CHANNEL_PROTOCOL_ID: () => SECURE_CHANNEL_PROTOCOL_ID,
  SecureChannelStatusCode: () => SecureChannelStatusCode,
  SecureMessageType: () => SecureMessageType
});
module.exports = __toCommonJS(secure_channel_exports);
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const SECURE_CHANNEL_PROTOCOL_ID = 0;
var SecureMessageType = /* @__PURE__ */ ((SecureMessageType2) => {
  SecureMessageType2[SecureMessageType2["MsgCounterSyncReq"] = 0] = "MsgCounterSyncReq";
  SecureMessageType2[SecureMessageType2["MsgCounterSyncRsp"] = 1] = "MsgCounterSyncRsp";
  SecureMessageType2[SecureMessageType2["StandaloneAck"] = 16] = "StandaloneAck";
  SecureMessageType2[SecureMessageType2["PbkdfParamRequest"] = 32] = "PbkdfParamRequest";
  SecureMessageType2[SecureMessageType2["PbkdfParamResponse"] = 33] = "PbkdfParamResponse";
  SecureMessageType2[SecureMessageType2["PasePake1"] = 34] = "PasePake1";
  SecureMessageType2[SecureMessageType2["PasePake2"] = 35] = "PasePake2";
  SecureMessageType2[SecureMessageType2["PasePake3"] = 36] = "PasePake3";
  SecureMessageType2[SecureMessageType2["Sigma1"] = 48] = "Sigma1";
  SecureMessageType2[SecureMessageType2["Sigma2"] = 49] = "Sigma2";
  SecureMessageType2[SecureMessageType2["Sigma3"] = 50] = "Sigma3";
  SecureMessageType2[SecureMessageType2["Sigma2Resume"] = 51] = "Sigma2Resume";
  SecureMessageType2[SecureMessageType2["StatusReport"] = 64] = "StatusReport";
  SecureMessageType2[SecureMessageType2["IcdCheckInMessage"] = 80] = "IcdCheckInMessage";
  return SecureMessageType2;
})(SecureMessageType || {});
var SecureChannelStatusCode = /* @__PURE__ */ ((SecureChannelStatusCode2) => {
  SecureChannelStatusCode2[SecureChannelStatusCode2["Success"] = 0] = "Success";
  SecureChannelStatusCode2[SecureChannelStatusCode2["NoSharedTrustRoots"] = 1] = "NoSharedTrustRoots";
  SecureChannelStatusCode2[SecureChannelStatusCode2["InvalidParam"] = 2] = "InvalidParam";
  SecureChannelStatusCode2[SecureChannelStatusCode2["CloseSession"] = 3] = "CloseSession";
  SecureChannelStatusCode2[SecureChannelStatusCode2["Busy"] = 4] = "Busy";
  SecureChannelStatusCode2[SecureChannelStatusCode2["GeneralError"] = 65535] = "GeneralError";
  return SecureChannelStatusCode2;
})(SecureChannelStatusCode || {});
((SecureMessageType2) => {
  function isStandaloneAck(protocolId, messageType) {
    return protocolId === SECURE_CHANNEL_PROTOCOL_ID && messageType === 16 /* StandaloneAck */;
  }
  SecureMessageType2.isStandaloneAck = isStandaloneAck;
})(SecureMessageType || (SecureMessageType = {}));
