// @matter/protocol/dist/cjs/action/client/ClientSubscriptionHandler.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClientSubscriptionHandler_exports = {};
__export(ClientSubscriptionHandler_exports, {
  ClientSubscriptionHandler: () => ClientSubscriptionHandler
});
module.exports = __toCommonJS(ClientSubscriptionHandler_exports);
var import_DecodedDataReport = require("#interaction/DecodedDataReport.js");
var import_InteractionMessenger = require("#interaction/InteractionMessenger.js");
var import_types = require("#types");
var import_general = require("@matter/general");
var import_InputChunk = require("./InputChunk.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("ClientSubscriptionHandler");
class ClientSubscriptionHandler {
  id = import_types.INTERACTION_PROTOCOL_ID;
  requiresSecureSession = true;
  #subscriptions;
  constructor(subscriptions) {
    this.#subscriptions = subscriptions;
  }
  async onNewExchange(exchange) {
    const messenger = new import_InteractionMessenger.IncomingInteractionClientMessenger(exchange);
    const reports = messenger.readDataReports();
    const initialIteration = await reports.next();
    if (initialIteration.done) {
      throw new import_general.InternalError("Exchange initiated with no initial message");
    }
    const initialReport = initialIteration.value;
    const { subscriptionId } = initialReport;
    if (subscriptionId === void 0) {
      logger.debug("Ignoring unsolicited data report with no subscription ID");
      await sendInvalid(messenger, void 0);
      return;
    }
    const subscription = this.#subscriptions.get(subscriptionId);
    if (subscription === void 0) {
      logger.debug("Ignoring data report for unknown subscription ID", import_general.Diagnostic.strong(subscriptionId));
      await sendInvalid(messenger, subscriptionId);
      return;
    }
    if (!initialReport.attributeReports?.length && !initialReport.eventReports?.length) {
      subscription.timeoutAtMs = void 0;
      this.#subscriptions.resetTimer();
      await exchange.close();
      return;
    }
    try {
      subscription.isReading = true;
      if (subscription.request.updated) {
        await subscription.request.updated(processReports(initialReport, reports, messenger));
      } else {
        for await (const _chunk of reports) ;
      }
    } finally {
      subscription.isReading = false;
      subscription.timeoutAtMs = void 0;
      this.#subscriptions.resetTimer();
      await exchange.close();
    }
  }
  async close() {
  }
}
async function sendInvalid(messenger, subscriptionId) {
  await messenger.sendStatus(import_types.Status.InvalidSubscription, {
    multipleMessageInteraction: true,
    logContext: {
      subId: subscriptionId
    }
  });
  await messenger.close();
}
async function* processReports(initialReport, otherReports, messenger) {
  yield (0, import_InputChunk.InputChunk)(initialReport);
  const { subscriptionId } = initialReport;
  for await (const report of otherReports) {
    const decoded = (0, import_DecodedDataReport.DecodedDataReport)(report);
    if (decoded.subscriptionId === void 0) {
      logger.debug(
        "Ignoring data report for incorrect subscription id",
        import_general.Diagnostic.strong(decoded.subscriptionId)
      );
      await sendInvalid(messenger, decoded.subscriptionId);
    }
    if (decoded.subscriptionId !== subscriptionId) {
      logger.debug(
        "Ignoring data report for incorrect subscription id",
        import_general.Diagnostic.strong(decoded.subscriptionId)
      );
      await sendInvalid(messenger, decoded.subscriptionId);
      continue;
    }
    yield (0, import_InputChunk.InputChunk)(report);
  }
}
