// @matter/protocol/dist/cjs/action/client/ClientSubscriptions.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClientSubscriptions_exports = {};
__export(ClientSubscriptions_exports, {
  ClientSubscriptions: () => ClientSubscriptions
});
module.exports = __toCommonJS(ClientSubscriptions_exports);
var import_general = require("#general");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("ClientSubscriptions");
class ClientSubscriptions {
  #subscriptions = new import_general.BasicSet();
  #timeout;
  #nextTimeoutAt;
  static [import_general.Environmental.create](env) {
    const instance = new ClientSubscriptions();
    env.set(ClientSubscriptions, instance);
    return instance;
  }
  /**
   * Register an active subscription.
   */
  add(request, response) {
    const subscription = {
      ...response,
      request,
      close: () => this.#closeOne(subscription, new import_general.CanceledError()),
      timeoutAtMs: void 0,
      isClosed: false,
      isReading: true
    };
    this.#subscriptions.add(subscription);
    this.resetTimer();
    return subscription;
  }
  /**
   * Retrieve a subscription by ID.
   */
  get(id) {
    return this.#subscriptions.get("subscriptionId", id);
  }
  /**
   * Iterate over active subscriptions.
   */
  [Symbol.iterator]() {
    return this.#subscriptions[Symbol.iterator]();
  }
  /**
   * Terminate all subscriptions.
   */
  async close() {
    if (this.#timeout) {
      this.#timeout.stop();
      this.#timeout = void 0;
    }
    for (const subscription of this.#subscriptions) {
      subscription.close();
    }
  }
  /**
   * Restart the timeout timer for the current set of active subscriptions.
   */
  resetTimer() {
    const now = import_general.Time.nowMs();
    let nextTimeoutAt;
    for (const subscription of this.#subscriptions) {
      if (subscription.isReading) {
        continue;
      }
      let { timeoutAtMs } = subscription;
      if (timeoutAtMs === void 0) {
        timeoutAtMs = subscription.timeoutAtMs = timeoutFor(subscription);
      } else if (timeoutAtMs < now) {
        this.#timeOut(subscription);
        continue;
      }
      if (nextTimeoutAt === void 0 || nextTimeoutAt > timeoutAtMs) {
        nextTimeoutAt = timeoutAtMs;
      }
    }
    if (nextTimeoutAt === void 0) {
      this.#nextTimeoutAt = void 0;
      this.#timeout?.stop();
      return;
    }
    if (nextTimeoutAt !== this.#nextTimeoutAt) {
      this.#nextTimeoutAt = nextTimeoutAt;
      if (this.#timeout) {
        this.#timeout?.stop();
        this.#timeout.intervalMs = nextTimeoutAt - now;
      } else {
        this.#timeout = import_general.Time.getTimer("SubscriptionTimeout", nextTimeoutAt - now, this.resetTimer.bind(this));
      }
    }
  }
  #closeOne(subscription, cause) {
    if (subscription.isClosed) {
      return;
    }
    subscription.isClosed = true;
    this.#subscriptions.delete(subscription);
    try {
      subscription.request.closed?.(cause);
    } catch (e) {
      logger.error("Error canceling subscription", import_general.Diagnostic.strong(subscription.subscriptionId), e);
    }
  }
  #timeOut(registration) {
    logger.info(
      "Subscription",
      import_general.Diagnostic.strong(registration.subscriptionId),
      "timed out after",
      import_general.Diagnostic.strong(`${timeoutFor(registration)}ms`)
    );
    this.#closeOne(registration, new import_general.TimeoutError());
  }
}
function timeoutFor(subscription) {
  return subscription.maxInterval * 1e3 + (subscription.request.maxPeerResponseTime ?? 0);
}
